{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File C:/Users/CedAI/Desktop/LLM/Data/ConceptsofBiology-WEB.pdf exists.\n"
     ]
    }
   ],
   "source": [
    "# Download PDF file\n",
    "import os\n",
    "import requests\n",
    "\n",
    "# Get PDF document\n",
    "pdf_path = \"C:/Users/CedAI/Desktop/LLM/Data/ConceptsofBiology-WEB.pdf\"\n",
    "\n",
    "# Download PDF if it doesn't already exist\n",
    "if not os.path.exists(pdf_path):\n",
    "  print(\"Download the File\")\n",
    "\n",
    "else:\n",
    "  print(f\"File {pdf_path} exists.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: tqdm in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (4.66.4)\n",
      "Requirement already satisfied: colorama in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tqdm) (0.4.6)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\CedAI\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import fitz\n",
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def text_formatter(text:str) -> str:\n",
    "    cleaned_text = text.replace(\"\\n\", \" \").strip()\n",
    "    return cleaned_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open PDF and get lines/pages\n",
    "# Note: this only focuses on text, rather than images/figures etc\n",
    "def open_and_read_pdf(pdf_path: str) -> list[dict]:\n",
    "    \"\"\"\n",
    "    Opens a PDF file, reads its text content page by page, and collects statistics.\n",
    "\n",
    "    Parameters:\n",
    "        pdf_path (str): The file path to the PDF document to be opened and read.\n",
    "\n",
    "    Returns:\n",
    "        list[dict]: A list of dictionaries, each containing the page number\n",
    "        (adjusted), character count, word count, sentence count, token count, and the extracted text\n",
    "        for each page.\n",
    "    \"\"\"\n",
    "    doc = fitz.open(pdf_path)  # open a document\n",
    "    pages_and_texts = []\n",
    "    for page_number, page in tqdm(enumerate(doc)):  # iterate the document pages\n",
    "        text = page.get_text()  # get plain text encoded as UTF-8\n",
    "        text = text_formatter(text)\n",
    "        pages_and_texts.append({\"page_number\": page_number ,  # adjust page numbers since our PDF starts on page 19\n",
    "                                \"page_char_count\": len(text),\n",
    "                                \"page_word_count\": len(text.split(\" \")),\n",
    "                                \"page_sentence_count_raw\": len(text.split(\". \")),\n",
    "                                \"page_token_count\": len(text) / 4,  # 1 token = ~4 chars, see: https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them\n",
    "                                \"text\": text})\n",
    "    return pages_and_texts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "615it [00:01, 535.57it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'page_number': 18,\n",
       "  'page_char_count': 1542,\n",
       "  'page_word_count': 255,\n",
       "  'page_sentence_count_raw': 10,\n",
       "  'page_token_count': 385.5,\n",
       "  'text': 'INTRODUCTION CHAPTER 1 Introduction to Biology 1.1 Themes and Concepts of Biology 1.2 The Process of Science Viewed from space, Earth (Figure 1.1) offers few clues about the diversity of life forms that reside there. The first forms of life on Earth are thought to have been microorganisms that existed for billions of years before plants and animals appeared. The mammals, birds, and flowers so familiar to us are all relatively recent, originating 130 to 200 million years ago. Humans have inhabited this planet for only the last 2.5 million years, and only in the last 300,000 years have humans started looking like we do today. 1.1 Themes and Concepts of Biology LEARNING OBJECTIVES By the end of this section, you will be able to: • Identify and describe the properties of life • Describe the levels of organization among living things • List examples of different sub disciplines in biology Biology is the science that studies life. What exactly is life? This may sound like a silly question with an obvious answer, but it is not easy to define life. For example, a branch of biology called virology studies viruses, which exhibit some of the characteristics of living entities but lack others. It turns out that although viruses can attack living organisms, cause diseases, and even reproduce, FIGURE 1.1 This NASA image is a composite of several satellite-based views of Earth. To make the whole-Earth image, NASA scientists combine observations of different parts of the planet. (credit: modification of work by NASA) CHAPTER OUTLINE'},\n",
       " {'page_number': 19,\n",
       "  'page_char_count': 2234,\n",
       "  'page_word_count': 337,\n",
       "  'page_sentence_count_raw': 16,\n",
       "  'page_token_count': 558.5,\n",
       "  'text': 'they do not meet the criteria that biologists use to define life. From its earliest beginnings, biology has wrestled with four questions: What are the shared properties that make something “alive”? How do those various living things function? When faced with the remarkable diversity of life, how do we organize the different kinds of organisms so that we can better understand them? And, finally—what biologists ultimately seek to understand—how did this diversity arise and how is it continuing? As new organisms are discovered every day, biologists continue to seek answers to these and other questions. Properties of Life All groups of living organisms share several key characteristics or functions: order, sensitivity or response to stimuli, reproduction, adaptation, growth and development, regulation/homeostasis, energy processing, and evolution. When viewed together, these eight characteristics serve to define life. Order Organisms are highly organized structures that consist of one or more cells. Even very simple, single-celled organisms are remarkably complex. Inside each cell, atoms make up molecules. These in turn make up cell components or organelles. Multicellular organisms, which may consist of millions of individual cells, have an advantage over single-celled organisms in that their cells can be specialized to perform specific functions, and even sacrificed in certain situations for the good of the organism as a whole. How these specialized cells come together to form organs such as the heart, lung, or skin in organisms like the toad shown in Figure 1.2 will be discussed later. FIGURE 1.2 A toad represents a highly organized structure consisting of cells, tissues, organs, and organ systems. (credit: \"Ivengo(RUS)\"/Wikimedia Commons) Sensitivity or Response to Stimuli Organisms respond to diverse stimuli. For example, plants can bend toward a source of light or respond to touch (Figure 1.3). Even tiny bacteria can move toward or away from chemicals (a process called chemotaxis) or light (phototaxis). Movement toward a stimulus is considered a positive response, while movement away from a stimulus is considered a negative response. 6 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 20,\n",
       "  'page_char_count': 1988,\n",
       "  'page_word_count': 306,\n",
       "  'page_sentence_count_raw': 16,\n",
       "  'page_token_count': 497.0,\n",
       "  'text': 'FIGURE 1.3 The leaves of this sensitive plant (Mimosa pudica) will instantly droop and fold when touched. After a few minutes, the plant returns to its normal state. (credit: Alex Lomas) LINK TO LEARNING Watch this video (http://openstax.org/l/thigmonasty) to see how the sensitive plant responds to a touch stimulus. Reproduction Single-celled organisms reproduce by first duplicating their DNA, which is the genetic material, and then dividing it equally as the cell prepares to divide to form two new cells. Many multicellular organisms (those made up of more than one cell) produce specialized reproductive cells that will form new individuals. When reproduction occurs, DNA containing genes is passed along to an organism’s offspring. These genes are the reason that the offspring will belong to the same species and will have characteristics similar to the parent, such as fur color and blood type. Adaptation All living organisms exhibit a “fit” to their environment. Biologists refer to this fit as adaptation and it is a consequence of evolution by natural selection, which operates in every lineage of reproducing organisms. Examples of adaptations are diverse and unique, from heat-resistant Archaea that live in boiling hot springs to the tongue length of a nectar-feeding moth that matches the size of the flower from which it feeds. Adaptations enhance the reproductive potential of the individual exhibiting them, including their ability to survive to reproduce. Adaptations are not constant. As an environment changes, natural selection causes the characteristics of the individuals in a population to track those changes. Growth and Development Organisms grow and develop according to specific instructions coded for by their genes. These genes provide instructions that will direct cellular growth and development, ensuring that a species’ young (Figure 1.4) will grow up to exhibit many of the same characteristics as its parents. 1.1 • Themes and Concepts of Biology 7'},\n",
       " {'page_number': 21,\n",
       "  'page_char_count': 2092,\n",
       "  'page_word_count': 309,\n",
       "  'page_sentence_count_raw': 14,\n",
       "  'page_token_count': 523.0,\n",
       "  'text': 'FIGURE 1.4 Although no two look alike, these kittens have inherited genes from both parents and share many of the same characteristics. (credit: Pieter & Renée Lanser) Regulation/Homeostasis Even the smallest organisms are complex and require multiple regulatory mechanisms to coordinate internal functions, such as the transport of nutrients, response to stimuli, and coping with environmental stresses. Homeostasis (literally, “steady state”) refers to the relatively stable internal environment required to maintain life. For example, organ systems such as the digestive or circulatory systems perform specific functions like carrying oxygen throughout the body, removing wastes, delivering nutrients to every cell, and cooling the body. To function properly, cells require appropriate conditions such as proper temperature, pH, and concentrations of diverse chemicals. These conditions may, however, change from one moment to the next. Organisms are able to maintain homeostatic internal conditions within a narrow range almost constantly, despite environmental changes, by activation of regulatory mechanisms. For example, many organisms regulate their body temperature in a process known as thermoregulation. Organisms that live in cold climates, such as the polar bear (Figure 1.5), have body structures that help them withstand low temperatures and conserve body heat. In hot climates, organisms have methods (such as perspiration in humans or panting in dogs) that help them to shed excess body heat. FIGURE 1.5 Polar bears and other mammals living in ice-covered regions maintain their body temperature by generating heat and reducing heat loss through thick fur and a dense layer of fat under their skin. (credit: \"longhorndave\"/Flickr) Energy Processing All organisms (such as the California condor shown in Figure 1.6) use a source of energy for their metabolic activities. Some organisms capture energy from the Sun and convert it into chemical energy in food; others use chemical energy from molecules they take in. 8 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 22,\n",
       "  'page_char_count': 1477,\n",
       "  'page_word_count': 234,\n",
       "  'page_sentence_count_raw': 15,\n",
       "  'page_token_count': 369.25,\n",
       "  'text': 'FIGURE 1.6 A lot of energy is required for a California condor to fly. Chemical energy derived from food is used to power flight. California condors are an endangered species; scientists have strived to place a wing tag on each bird to help them identify and locate each individual bird. (credit: Pacific Southwest Region U.S. Fish and Wildlife) Evolution The diversity of life on Earth is a result of mutations, or random changes in hereditary material over time. These mutations allow the possibility for organisms to adapt to a changing environment. An organism that evolves characteristics fit for the environment will have greater reproductive success, subject to the forces of natural selection. Levels of Organization of Living Things Living things are highly organized and structured, following a hierarchy on a scale from small to large. The atom is the smallest and most fundamental unit of matter that retains the properties of an element. It consists of a nucleus surrounded by electrons. Atoms form molecules. A molecule is a chemical structure consisting of at least two atoms held together by a chemical bond. Many molecules that are biologically important are macromolecules, large molecules that are typically formed by combining smaller units called monomers. An example of a macromolecule is deoxyribonucleic acid (DNA) (Figure 1.7), which contains the instructions for the functioning of the organism that contains it. 1.1 • Themes and Concepts of Biology 9'},\n",
       " {'page_number': 23,\n",
       "  'page_char_count': 1982,\n",
       "  'page_word_count': 302,\n",
       "  'page_sentence_count_raw': 18,\n",
       "  'page_token_count': 495.5,\n",
       "  'text': 'FIGURE 1.7 A molecule, like this large DNA molecule, is composed of atoms. (credit: \"Brian0918\"/Wikimedia Commons) LINK TO LEARNING To see an animation of this DNA molecule, click here (http://openstax.org/l/rotating_DNA2). Some cells contain aggregates of macromolecules surrounded by membranes; these are called organelles. Organelles are small structures that exist within cells and perform specialized functions. All living things are made of cells; the cell itself is the smallest fundamental unit of structure and function in living organisms. (This requirement is why viruses are not considered living: they are not made of cells. To make new viruses, they have to invade and hijack a living cell; only then can they obtain the materials they need to reproduce.) Some organisms consist of a single cell and others are multicellular. Cells are classified as prokaryotic or eukaryotic. Prokaryotes are single- celled organisms that lack organelles surrounded by a membrane and do not have nuclei surrounded by nuclear membranes; in contrast, the cells of eukaryotes do have membrane-bound organelles and nuclei. In most multicellular organisms, cells combine to make tissues, which are groups of similar cells carrying out the same function. Organs are collections of tissues grouped together based on a common function. Organs are present not only in animals but also in plants. An organ system is a higher level of organization that consists of functionally related organs. For example vertebrate animals have many organ systems, such as the circulatory system that transports blood throughout the body and to and from the lungs; it includes organs such as the heart and blood vessels. Organisms are individual living entities. For example, each tree in a forest is an organism. Single-celled prokaryotes and single-celled eukaryotes are also considered organisms and are typically referred to as microorganisms. 10 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 24,\n",
       "  'page_char_count': 791,\n",
       "  'page_word_count': 113,\n",
       "  'page_sentence_count_raw': 4,\n",
       "  'page_token_count': 197.75,\n",
       "  'text': 'VISUAL CONNECTION FIGURE 1.8 From an atom to the entire Earth, biology examines all aspects of life. (credit \"molecule\": modification of work by Jane Whitney; credit \"organelles\": modification of work by Louisa Howard; credit \"cells\": modification of work by Bruce Wetzel, Harry Schaefer, National Cancer Institute; credit \"tissue\": modification of work by \"Kilbad\"/Wikimedia Commons; credit \"organs\": modification of work by Mariana Ruiz Villareal, Joaquim Alves Gaspar; credit \"organisms\": modification of work by Peter Dutton; credit \"ecosystem\": modification of work by \"gigi4791\"/Flickr; credit \"biosphere\": modification of work by NASA) Which of the following statements is false? a. Tissues exist within organs which exist within organ systems. 1.1 • Themes and Concepts of Biology 11'},\n",
       " {'page_number': 25,\n",
       "  'page_char_count': 3620,\n",
       "  'page_word_count': 564,\n",
       "  'page_sentence_count_raw': 37,\n",
       "  'page_token_count': 905.0,\n",
       "  'text': 'b. Communities exist within populations which exist within ecosystems. c. Organelles exist within cells which exist within tissues. d. Communities exist within ecosystems which exist in the biosphere. All the individuals of a species living within a specific area are collectively called a population. For example, a forest may include many white pine trees. All of these pine trees represent the population of white pine trees in this forest. Different populations may live in the same specific area. For example, the forest with the pine trees includes populations of flowering plants and also insects and microbial populations. A community is the set of populations inhabiting a particular area. For instance, all of the trees, flowers, insects, and other populations in a forest form the forest’s community. The forest itself is an ecosystem. An ecosystem consists of all the living things in a particular area together with the abiotic, or non-living, parts of that environment such as nitrogen in the soil or rainwater. At the highest level of organization (Figure 1.8), the biosphere is the collection of all ecosystems, and it represents the zones of life on Earth. It includes land, water, and portions of the atmosphere. The Diversity of Life The science of biology is very broad in scope because there is a tremendous diversity of life on Earth. The source of this diversity is evolution, the process of gradual change during which new species arise from older species. Evolutionary biologists study the evolution of living things in everything from the microscopic world to ecosystems. In the 18th century, a scientist named Carl Linnaeus first proposed organizing the known species of organisms into a hierarchical taxonomy. In this system, species that are most similar to each other are put together within a grouping known as a genus. Furthermore, similar genera (the plural of genus) are put together within a family. This grouping continues until all organisms are collected together into groups at the highest level. The current taxonomic system now has eight levels in its hierarchy, from lowest to highest, they are: species, genus, family, order, class, phylum, kingdom, domain. Thus species are grouped within genera, genera are grouped within families, families are grouped within orders, and so on (Figure 1.9). FIGURE 1.9 This diagram shows the levels of taxonomic hierarchy for a dog, from the broadest category—domain—to the most specific—species. The highest level, domain, is a relatively new addition to the system since the 1970s. Scientists now recognize three domains of life, the Eukarya, the Archaea, and the Bacteria. The domain Eukarya contains organisms that have cells with nuclei. It includes the kingdoms of fungi, plants, animals, and several kingdoms of protists. The Archaea, are single-celled organisms without nuclei and include many extremophiles that live in harsh environments like hot springs. The Bacteria are another quite different group of single-celled organisms without nuclei (Figure 1.10). Both the Archaea and the Bacteria are prokaryotes, an informal name for cells without nuclei. The recognition in the 1970s that certain “bacteria,” now known as the Archaea, were as different genetically and biochemically from other bacterial cells as they were from eukaryotes, motivated the recommendation to divide life into three domains. This dramatic change in our knowledge of the tree of life demonstrates that classifications are not permanent and will change when new information becomes available. 12 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 26,\n",
       "  'page_char_count': 2963,\n",
       "  'page_word_count': 459,\n",
       "  'page_sentence_count_raw': 23,\n",
       "  'page_token_count': 740.75,\n",
       "  'text': 'In addition to the hierarchical taxonomic system, Linnaeus was the first to name organisms using two unique names, now called the binomial naming system. Before Linnaeus, the use of common names to refer to organisms caused confusion because there were regional differences in these common names. Binomial names consist of the genus name (which is capitalized) and the species name (all lower-case). Both names are set in italics when they are printed. Every species is given a unique binomial which is recognized the world over, so that a scientist in any location can know which organism is being referred to. For example, the North American blue jay is known uniquely as Cyanocitta cristata. Our own species is Homo sapiens. FIGURE 1.10 These images represent different domains. The scanning electron micrograph shows (a) bacterial cells belong to the domain Bacteria, while the (b) extremophiles, seen all together as colored mats in this hot spring, belong to domain Archaea. Both the (c) sunflower and (d) lion are part of domain Eukarya. (credit a: modification of work by Rocky Mountain Laboratories, NIAID, NIH; credit b: modification of work by Steve Jurvetson; credit c: modification of work by Michael Arrighi; credit d: modification of work by Frank Vassen) EVOLUTION CONNECTION Carl Woese and the Phylogenetic Tree The evolutionary relationships of various life forms on Earth can be summarized in a phylogenetic tree. A phylogenetic tree is a diagram showing the evolutionary relationships among biological species based on similarities and differences in genetic or physical traits or both. A phylogenetic tree is composed of branch points, or nodes, and branches. The internal nodes represent ancestors and are points in evolution when, based on scientific evidence, an ancestor is thought to have diverged to form two new species. The length of each branch can be considered as estimates of relative time. In the past, biologists grouped living organisms into five kingdoms: animals, plants, fungi, protists, and bacteria. The pioneering work of American microbiologist Carl Woese in the early 1970s has shown, however, that life on Earth has evolved along three lineages, now called domains—Bacteria, Archaea, and Eukarya. Woese proposed the domain as a new taxonomic level and Archaea as a new domain, to reflect the new phylogenetic tree (Figure 1.11). Many organisms belonging to the Archaea domain live under extreme conditions and are called extremophiles. To construct his tree, Woese used genetic relationships rather than similarities based on morphology (shape). Various genes were used in phylogenetic studies. Woese’s tree was constructed from comparative sequencing of the genes that are universally distributed, found in some slightly altered form in every organism, conserved (meaning that these genes have remained only slightly changed throughout evolution), and of an appropriate length. 1.1 • Themes and Concepts of Biology 13'},\n",
       " {'page_number': 27,\n",
       "  'page_char_count': 1611,\n",
       "  'page_word_count': 238,\n",
       "  'page_sentence_count_raw': 12,\n",
       "  'page_token_count': 402.75,\n",
       "  'text': 'FIGURE 1.11 This phylogenetic tree was constructed by microbiologist Carl Woese using genetic relationships. The tree shows the separation of living organisms into three domains: Bacteria, Archaea, and Eukarya. Bacteria and Archaea are organisms without a nucleus or other organelles surrounded by a membrane and, therefore, are prokaryotes. (credit: modification of work by Eric Gaba) Branches of Biological Study The scope of biology is broad and therefore contains many branches and sub disciplines. Biologists may pursue one of those sub disciplines and work in a more focused field. For instance, molecular biology studies biological processes at the molecular level, including interactions among molecules such as DNA, RNA, and proteins, as well as the way they are regulated. Microbiology is the study of the structure and function of microorganisms. It is quite a broad branch itself, and depending on the subject of study, there are also microbial physiologists, ecologists, and geneticists, among others. Another field of biological study, neurobiology, studies the biology of the nervous system, and although it is considered a branch of biology, it is also recognized as an interdisciplinary field of study known as neuroscience. Because of its interdisciplinary nature, this sub discipline studies different functions of the nervous system using molecular, cellular, developmental, medical, and computational approaches. FIGURE 1.12 Researchers work on excavating dinosaur fossils at a site in Castellón, Spain. (credit: Mario Modesto) 14 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 28,\n",
       "  'page_char_count': 3040,\n",
       "  'page_word_count': 471,\n",
       "  'page_sentence_count_raw': 27,\n",
       "  'page_token_count': 760.0,\n",
       "  'text': 'Paleontology, another branch of biology, uses fossils to study life’s history (Figure 1.12). Zoology and botany are the study of animals and plants, respectively. Biologists can also specialize as biotechnologists, ecologists, or physiologists, to name just a few areas. Biotechnologists apply the knowledge of biology to create useful products. Ecologists study the interactions of organisms in their environments. Physiologists study the workings of cells, tissues and organs. This is just a small sample of the many fields that biologists can pursue. From our own bodies to the world we live in, discoveries in biology can affect us in very direct and important ways. We depend on these discoveries for our health, our food sources, and the benefits provided by our ecosystem. Because of this, knowledge of biology can benefit us in making decisions in our day-to-day lives. The development of technology in the twentieth century that continues today, particularly the technology to describe and manipulate the genetic material, DNA, has transformed biology. This transformation will allow biologists to continue to understand the history of life in greater detail, how the human body works, our human origins, and how humans can survive as a species on this planet despite the stresses caused by our increasing numbers. Biologists continue to decipher huge mysteries about life suggesting that we have only begun to understand life on the planet, its history, and our relationship to it. For this and other reasons, the knowledge of biology gained through this textbook and other printed and electronic media should be a benefit in whichever field you enter. CAREER CONNECTION Forensic Scientist Forensic science is the application of science to answer questions related to the law. Biologists as well as chemists and biochemists can be forensic scientists. Forensic scientists provide scientific evidence for use in courts, and their job involves examining trace material associated with crimes. Interest in forensic science has increased in the last few years, possibly because of popular television shows that feature forensic scientists on the job. Also, the development of molecular techniques and the establishment of DNA databases have updated the types of work that forensic scientists can do. Their job activities are primarily related to crimes against people such as murder, rape, and assault. Their work involves analyzing samples such as hair, blood, and other body fluids and also processing DNA (Figure 1.13) found in many different environments and materials. Forensic scientists also analyze other biological evidence left at crime scenes, such as insect parts or pollen grains. Students who want to pursue careers in forensic science will most likely be required to take chemistry and biology courses as well as some intensive math courses. FIGURE 1.13 This forensic scientist works in a DNA extraction room at the U.S. Army Criminal Investigation Laboratory. (credit: U.S. Army 1.1 • Themes and Concepts of Biology 15'},\n",
       " {'page_number': 29,\n",
       "  'page_char_count': 3526,\n",
       "  'page_word_count': 550,\n",
       "  'page_sentence_count_raw': 19,\n",
       "  'page_token_count': 881.5,\n",
       "  'text': 'CID Command Public Affairs) Scientific Ethics Scientists must ensure that their efforts do not cause undue damage to humans, animals, or the environment. They also must ensure that their research and communications are free of bias and that they properly balance financial, legal, safety, replicability, and other considerations. Bioethics is an important and continually evolving field, in which researchers collaborate with other thinkers and organizations. They work to define guidelines for current practice, and also continually consider new developments and emerging technologies in order to form answers for the years and decades to come. Unfortunately, the emergence of bioethics as a field came after a number of clearly unethical practices, where biologists did not treat research subjects with dignity and in some cases did them harm. In the 1932 Tuskegee syphilis study, 399 African American men were diagnosed with syphilis but were never informed that they had the disease, leaving them to live with and pass on the illness to others. Doctors even withheld proven medications because the goal of the study was to understand the impact of untreated syphilis on Black men. While the decisions made in the Tuskegee study are unjustifiable, some decisions are genuinely difficult to make. For example, bioethicists may examine the implications of gene editing technologies, including the ability to create organisms that may displace others in the environment, as well as the ability to “design” human beings. In that effort, ethicists will likely seek to balance the positive outcomes -- such as improved therapies or prevention of certain illnesses -- with negative outcomes. Bioethics are not simple, and often leave scientists balancing benefits with harm. In this text and course, you will discuss medical discoveries that, at their core, have what many consider an ethical lapse. In 1951, Henrietta Lacks, a 30-year-old African American woman, was diagnosed with cervical cancer at Johns Hopkins Hospital. Unique characteristics of her illnesses gave her cells the ability to divide continuously, essentially making them “immortal.” Without her knowledge or permission, researchers took samples of her cells and with them created the immortal HeLa cell line. These cells have contributed to major medical discoveries, including the polio vaccine and work related to cancer, AIDS, cell aging, and even very recently in COVID-19 research. For the most part, Lacks has not been credited for her role in those discoveries, and her family has not benefited from the billions of dollars in pharmaceutical profits obtained partly through the use of her cells. Today, harvesting tissue or organs from a dying patient without consent is not only considered unethical but also illegal, regardless of whether such an act could save other patients’ lives. Part of the role of ethics in scientific research is to examine similar issues before, during, and after research or practice takes place, as well as to adhere to established professional principles and consider the dignity and safety of all organisms involved or affected by the work. 1.2 The Process of Science LEARNING OBJECTIVES By the end of this section, you will be able to: • Identify the shared characteristics of the natural sciences • Understand the process of scientific inquiry • Compare inductive reasoning with deductive reasoning • Describe the goals of basic science and applied science 16 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 30,\n",
       "  'page_char_count': 2344,\n",
       "  'page_word_count': 355,\n",
       "  'page_sentence_count_raw': 17,\n",
       "  'page_token_count': 586.0,\n",
       "  'text': 'FIGURE 1.14 Formerly called blue-green algae, the (a) cyanobacteria seen through a light microscope are some of Earth’s oldest life forms. These (b) stromatolites along the shores of Lake Thetis in Western Australia are ancient structures formed by the layering of cyanobacteria in shallow waters. (credit a: modification of work by NASA; scale-bar data from Matt Russell; credit b: modification of work by Ruth Ellison) Like geology, physics, and chemistry, biology is a science that gathers knowledge about the natural world. Specifically, biology is the study of life. The discoveries of biology are made by a community of researchers who work individually and together using agreed-on methods. In this sense, biology, like all sciences is a social enterprise like politics or the arts. The methods of science include careful observation, record keeping, logical and mathematical reasoning, experimentation, and submitting conclusions to the scrutiny of others. Science also requires considerable imagination and creativity; a well-designed experiment is commonly described as elegant, or beautiful. Like politics, science has considerable practical implications and some science is dedicated to practical applications, such as the prevention of disease (see Figure 1.15). Other science proceeds largely motivated by curiosity. Whatever its goal, there is no doubt that science, including biology, has transformed human existence and will continue to do so. FIGURE 1.15 Biologists may choose to study Escherichia coli (E. coli), a bacterium that is a normal resident of our digestive tracts but which is also sometimes responsible for disease outbreaks. In this micrograph, the bacterium is visualized using a scanning electron microscope and digital colorization. (credit: Eric Erbe; digital colorization by Christopher Pooley, USDA-ARS) The Nature of Science Biology is a science, but what exactly is science? What does the study of biology share with other scientific disciplines? Science (from the Latin scientia, meaning \"knowledge\") can be defined as knowledge about the natural world. Science is a very specific way of learning, or knowing, about the world. The history of the past 500 years demonstrates that science is a very powerful way of knowing about the world; it is largely responsible for the 1.2 • The Process of Science 17'},\n",
       " {'page_number': 31,\n",
       "  'page_char_count': 3264,\n",
       "  'page_word_count': 490,\n",
       "  'page_sentence_count_raw': 25,\n",
       "  'page_token_count': 816.0,\n",
       "  'text': 'technological revolutions that have taken place during this time. There are however, areas of knowledge and human experience that the methods of science cannot be applied to. These include such things as answering purely moral questions, aesthetic questions, or what can be generally categorized as spiritual questions. Science cannot investigate these areas because they are outside the realm of material phenomena, the phenomena of matter and energy, and cannot be observed and measured. The scientific method is a method of research with defined steps that include experiments and careful observation. The steps of the scientific method will be examined in detail later, but one of the most important aspects of this method is the testing of hypotheses. A hypothesis is a suggested explanation for an event, which can be tested. Hypotheses, or tentative explanations, are generally produced within the context of a scientific theory. A generally accepted scientific theory is thoroughly tested and confirmed explanation for a set of observations or phenomena. Scientific theory is the foundation of scientific knowledge. In addition, in many scientific disciplines (less so in biology) there are scientific laws, often expressed in mathematical formulas, which describe how elements of nature will behave under certain specific conditions. There is not an evolution of hypotheses through theories to laws as if they represented some increase in certainty about the world. Hypotheses are the day-to-day material that scientists work with and they are developed within the context of theories. Laws are concise descriptions of parts of the world that are amenable to formulaic or mathematical description. Natural Sciences What would you expect to see in a museum of natural sciences? Frogs? Plants? Dinosaur skeletons? Exhibits about how the brain functions? A planetarium? Gems and minerals? Or maybe all of the above? Science includes such diverse fields as astronomy, biology, computer sciences, geology, logic, physics, chemistry, and mathematics (Figure 1.16). However, those fields of science related to the physical world and its phenomena and processes are considered natural sciences. Thus, a museum of natural sciences might contain any of the items listed above. FIGURE 1.16 Some fields of science include astronomy, biology, computer science, geology, logic, physics, chemistry, and mathematics. (credit: \"Image Editor\"/Flickr) There is no complete agreement when it comes to defining what the natural sciences include. For some experts, the natural sciences are astronomy, biology, chemistry, earth science, and physics. Other scholars choose to divide natural sciences into life sciences, which study living things and include biology, and physical sciences, which study nonliving matter and include astronomy, physics, and chemistry. Some disciplines such as biophysics and biochemistry build on two sciences and are interdisciplinary. Scientific Inquiry One thing is common to all forms of science: an ultimate goal “to know.” Curiosity and inquiry are the driving forces for the development of science. Scientists seek to understand the world and the way it operates. Two methods of 18 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 32,\n",
       "  'page_char_count': 3180,\n",
       "  'page_word_count': 490,\n",
       "  'page_sentence_count_raw': 28,\n",
       "  'page_token_count': 795.0,\n",
       "  'text': 'logical thinking are used: inductive reasoning and deductive reasoning. Inductive reasoning is a form of logical thinking that uses related observations to arrive at a general conclusion. This type of reasoning is common in descriptive science. A life scientist such as a biologist makes observations and records them. These data can be qualitative (descriptive) or quantitative (consisting of numbers), and the raw data can be supplemented with drawings, pictures, photos, or videos. From many observations, the scientist can infer conclusions (inductions) based on evidence. Inductive reasoning involves formulating generalizations inferred from careful observation and the analysis of a large amount of data. Brain studies often work this way. Many brains are observed while people are doing a task. The part of the brain that lights up, indicating activity, is then demonstrated to be the part controlling the response to that task. Deductive reasoning or deduction is the type of logic used in hypothesis-based science. In deductive reasoning, the pattern of thinking moves in the opposite direction as compared to inductive reasoning. Deductive reasoning is a form of logical thinking that uses a general principle or law to predict specific results. From those general principles, a scientist can deduce and predict the specific results that would be valid as long as the general principles are valid. For example, a prediction would be that if the climate is becoming warmer in a region, the distribution of plants and animals should change. Comparisons have been made between distributions in the past and the present, and the many changes that have been found are consistent with a warming climate. Finding the change in distribution is evidence that the climate change conclusion is a valid one. Both types of logical thinking are related to the two main pathways of scientific study: descriptive science and hypothesis-based science. Descriptive (or discovery) science aims to observe, explore, and discover, while hypothesis-based science begins with a specific question or problem and a potential answer or solution that can be tested. The boundary between these two forms of study is often blurred, because most scientific endeavors combine both approaches. Observations lead to questions, questions lead to forming a hypothesis as a possible answer to those questions, and then the hypothesis is tested. Thus, descriptive science and hypothesis-based science are in continuous dialogue. Hypothesis Testing Biologists study the living world by posing questions about it and seeking science-based responses. This approach is common to other sciences as well and is often referred to as the scientific method. The scientific method was used even in ancient times, but it was first documented by England’s Sir Francis Bacon (1561–1626) (Figure 1.17), who set up inductive methods for scientific inquiry. The scientific method is not exclusively used by biologists but can be applied to almost anything as a logical problem-solving method. FIGURE 1.17 Sir Francis Bacon is credited with being the first to document the scientific method. 1.2 • The Process of Science 19'},\n",
       " {'page_number': 33,\n",
       "  'page_char_count': 4145,\n",
       "  'page_word_count': 706,\n",
       "  'page_sentence_count_raw': 39,\n",
       "  'page_token_count': 1036.25,\n",
       "  'text': 'The scientific process typically starts with an observation (often a problem to be solved) that leads to a question. Let’s think about a simple problem that starts with an observation and apply the scientific method to solve the problem. One Monday morning, a student arrives at class and quickly discovers that the classroom is too warm. That is an observation that also describes a problem: the classroom is too warm. The student then asks a question: “Why is the classroom so warm?” Recall that a hypothesis is a suggested explanation that can be tested. To solve a problem, several hypotheses may be proposed. For example, one hypothesis might be, “The classroom is warm because no one turned on the air conditioning.” But there could be other responses to the question, and therefore other hypotheses may be proposed. A second hypothesis might be, “The classroom is warm because there is a power failure, and so the air conditioning doesn’t work.” Once a hypothesis has been selected, a prediction may be made. A prediction is similar to a hypothesis but it typically has the format “If . . . then . . . .” For example, the prediction for the first hypothesis might be, “If the student turns on the air conditioning, then the classroom will no longer be too warm.” A hypothesis must be testable to ensure that it is valid. For example, a hypothesis that depends on what a bear thinks is not testable, because it can never be known what a bear thinks. It should also be falsifiable, meaning that it can be disproven by experimental results. An example of an unfalsifiable hypothesis is “Botticelli’s Birth of Venus is beautiful.” There is no experiment that might show this statement to be false. To test a hypothesis, a researcher will conduct one or more experiments designed to eliminate one or more of the hypotheses. This is important. A hypothesis can be disproven, or eliminated, but it can never be proven. Science does not deal in proofs like mathematics. If an experiment fails to disprove a hypothesis, then we find support for that explanation, but this is not to say that down the road a better explanation will not be found, or a more carefully designed experiment will be found to falsify the hypothesis. Each experiment will have one or more variables and one or more controls. A variable is any part of the experiment that can vary or change during the experiment. A control is a part of the experiment that does not change. Look for the variables and controls in the example that follows. As a simple example, an experiment might be conducted to test the hypothesis that phosphate limits the growth of algae in freshwater ponds. A series of artificial ponds are filled with water and half of them are treated by adding phosphate each week, while the other half are treated by adding a salt that is known not to be used by algae. The variable here is the phosphate (or lack of phosphate), the experimental or treatment cases are the ponds with added phosphate and the control ponds are those with something inert added, such as the salt. Just adding something is also a control against the possibility that adding extra matter to the pond has an effect. If the treated ponds show lesser growth of algae, then we have found support for our hypothesis. If they do not, then we reject our hypothesis. Be aware that rejecting one hypothesis does not determine whether or not the other hypotheses can be accepted; it simply eliminates one hypothesis that is not valid (Figure 1.18). Using the scientific method, the hypotheses that are inconsistent with experimental data are rejected. In recent years a new approach of testing hypotheses has developed as a result of an exponential growth of data deposited in various databases. Using computer algorithms and statistical analyses of data in databases, a new field of so-called \"data research\" (also referred to as \"in silico\" research) provides new methods of data analyses and their interpretation. This will increase the demand for specialists in both biology and computer science, a promising career opportunity. 20 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 34,\n",
       "  'page_char_count': 1582,\n",
       "  'page_word_count': 256,\n",
       "  'page_sentence_count_raw': 21,\n",
       "  'page_token_count': 395.5,\n",
       "  'text': 'VISUAL CONNECTION FIGURE 1.18 The scientific method is a series of defined steps that include experiments and careful observation. If a hypothesis is not supported by data, a new hypothesis can be proposed. In the example below, the scientific method is used to solve an everyday problem. Which part in the example below is the hypothesis? Which is the prediction? Based on the results of the experiment, is the hypothesis supported? If it is not supported, propose some alternative hypotheses. 1. My toaster doesn’t toast my bread. 2. Why doesn’t my toaster work? 3. There is something wrong with the electrical outlet. 4. If something is wrong with the outlet, my coffeemaker also won’t work when plugged into it. 5. I plug my coffeemaker into the outlet. 6. My coffeemaker works. In practice, the scientific method is not as rigid and structured as it might at first appear. Sometimes an experiment leads to conclusions that favor a change in approach; often, an experiment brings entirely new scientific questions to the puzzle. Many times, science does not operate in a linear fashion; instead, scientists continually draw inferences and make generalizations, finding patterns as their research proceeds. Scientific reasoning is more complex than the scientific method alone suggests. Basic and Applied Science The scientific community has been debating for the last few decades about the value of different types of science. Is it valuable to pursue science for the sake of simply gaining knowledge, or does scientific knowledge only have worth 1.2 • The Process of Science 21'},\n",
       " {'page_number': 35,\n",
       "  'page_char_count': 3687,\n",
       "  'page_word_count': 586,\n",
       "  'page_sentence_count_raw': 25,\n",
       "  'page_token_count': 921.75,\n",
       "  'text': 'if we can apply it to solving a specific problem or bettering our lives? This question focuses on the differences between two types of science: basic science and applied science. Basic science or “pure” science seeks to expand knowledge regardless of the short-term application of that knowledge. It is not focused on developing a product or a service of immediate public or commercial value. The immediate goal of basic science is knowledge for knowledge’s sake, though this does not mean that in the end it may not result in an application. In contrast, applied science or “technology,” aims to use science to solve real-world problems, making it possible, for example, to improve a crop yield, find a cure for a particular disease, or save animals threatened by a natural disaster. In applied science, the problem is usually defined for the researcher. Some individuals may perceive applied science as “useful” and basic science as “useless.” A question these people might pose to a scientist advocating knowledge acquisition would be, “What for?” A careful look at the history of science, however, reveals that basic knowledge has resulted in many remarkable applications of great value. Many scientists think that a basic understanding of science is necessary before an application is developed; therefore, applied science relies on the results generated through basic science. Other scientists think that it is time to move on from basic science and instead to find solutions to actual problems. Both approaches are valid. It is true that there are problems that demand immediate attention; however, few solutions would be found without the help of the knowledge generated through basic science. One example of how basic and applied science can work together to solve practical problems occurred after the discovery of DNA structure led to an understanding of the molecular mechanisms governing DNA replication. Strands of DNA, unique in every human, are found in our cells, where they provide the instructions necessary for life. During DNA replication, new copies of DNA are made, shortly before a cell divides to form new cells. Understanding the mechanisms of DNA replication enabled scientists to develop laboratory techniques that are now used to identify genetic diseases, pinpoint individuals who were at a crime scene, and determine paternity. Without basic science, it is unlikely that applied science could exist. Another example of the link between basic and applied research is the Human Genome Project, a study in which each human chromosome was analyzed and mapped to determine the precise sequence of DNA subunits and the exact location of each gene. (The gene is the basic unit of heredity represented by a specific DNA segment that codes for a functional molecule.) Other organisms have also been studied as part of this project to gain a better understanding of human chromosomes. The Human Genome Project (Figure 1.19) relied on basic research carried out with non-human organisms and, later, with the human genome. An important end goal eventually became using the data for applied research seeking cures for genetically related diseases. FIGURE 1.19 The Human Genome Project was a 13-year collaborative effort among researchers working in several different fields of science. The project was completed in 2003. (credit: the U.S. Department of Energy Genome Programs) While research efforts in both basic science and applied science are usually carefully planned, it is important to note that some discoveries are made by serendipity, that is, by means of a fortunate accident or a lucky surprise. 22 1 • Introduction to Biology Access for free at openstax.org'},\n",
       " {'page_number': 36,\n",
       "  'page_char_count': 3092,\n",
       "  'page_word_count': 467,\n",
       "  'page_sentence_count_raw': 22,\n",
       "  'page_token_count': 773.0,\n",
       "  'text': \"Penicillin was discovered when biologist Alexander Fleming accidentally left a petri dish of Staphylococcus bacteria open. An unwanted mold grew, killing the bacteria. The mold turned out to be Penicillium, and a new critically important antibiotic was discovered. In a similar manner, Percy Lavon Julian was an established medicinal chemist working on a way to mass produce compounds with which to manufacture important drugs. He was focused on using soybean oil in the production of progesterone (a hormone important in the menstrual cycle and pregnancy), but it wasn't until water accidentally leaked into a large soybean oil storage tank that he found his method. Immediately recognizing the resulting substance as stigmasterol, a primary ingredient in progesterone and similar drugs, he began the process of replicating and industrializing the process in a manner that has helped millions of people. Even in the highly organized world of science, luck—when combined with an observant, curious mind focused on the types of reasoning discussed above—can lead to unexpected breakthroughs. Reporting Scientific Work Whether scientific research is basic science or applied science, scientists must share their findings for other researchers to expand and build upon their discoveries. Communication and collaboration within and between sub disciplines of science are key to the advancement of knowledge in science. For this reason, an important aspect of a scientist’s work is disseminating results and communicating with peers. Scientists can share results by presenting them at a scientific meeting or conference, but this approach can reach only the limited few who are present. Instead, most scientists present their results in peer-reviewed articles that are published in scientific journals. Peer-reviewed articles are scientific papers that are reviewed, usually anonymously by a scientist’s colleagues, or peers. These colleagues are qualified individuals, often experts in the same research area, who judge whether or not the scientist’s work is suitable for publication. The process of peer review helps to ensure that the research described in a scientific paper or grant proposal is original, significant, logical, and thorough. Grant proposals, which are requests for research funding, are also subject to peer review. Scientists publish their work so other scientists can reproduce their experiments under similar or different conditions to expand on the findings. There are many journals and the popular press that do not use a peer-review system. A large number of online open-access journals, journals with articles available without cost, are now available many of which use rigorous peer-review systems, but some of which do not. Results of any studies published in these forums without peer review are not reliable and should not form the basis for other scientific work. In one exception, journals may allow a researcher to cite a personal communication from another researcher about unpublished results with the cited author’s permission. 1.2 • The Process of Science 23\"},\n",
       " {'page_number': 37,\n",
       "  'page_char_count': 4190,\n",
       "  'page_word_count': 629,\n",
       "  'page_sentence_count_raw': 9,\n",
       "  'page_token_count': 1047.5,\n",
       "  'text': 'Key Terms applied science a form of science that solves real- world problems atom a basic unit of matter that cannot be broken down by normal chemical reactions basic science science that seeks to expand knowledge regardless of the short-term application of that knowledge biology the study of life biosphere a collection of all ecosystems on Earth cell the smallest fundamental unit of structure and function in living things community a set of populations inhabiting a particular area control a part of an experiment that does not change during the experiment deductive reasoning a form of logical thinking that uses a general statement to predict specific results descriptive science a form of science that aims to observe, explore, and find things out ecosystem all living things in a particular area together with the abiotic, nonliving parts of that environment eukaryote an organism with cells that have nuclei and membrane-bound organelles evolution the process of gradual change in a population that can also lead to new species arising from older species falsifiable able to be disproven by experimental results homeostasis the ability of an organism to maintain constant internal conditions hypothesis a suggested explanation for an event, which can be tested hypothesis-based science a form of science that begins with a specific explanation that is then tested inductive reasoning a form of logical thinking that uses related observations to arrive at a general conclusion life science a field of science, such as biology, that studies living things macromolecule a large molecule typically formed by the joining of smaller molecules molecule a chemical structure consisting of at least two atoms held together by a chemical bond natural science a field of science that studies the physical world, its phenomena, and processes organ a structure formed of tissues operating together to perform a common function organ system the higher level of organization that consists of functionally related organs organelle a membrane-bound compartment or sac within a cell organism an individual living entity peer-reviewed article a scientific report that is reviewed by a scientist’s colleagues before publication phylogenetic tree a diagram showing the evolutionary relationships among biological species based on similarities and differences in genetic or physical traits or both physical science a field of science, such as astronomy, physics, and chemistry, that studies nonliving matter population all individuals within a species living within a specific area prokaryote a unicellular organism that lacks a nucleus or any other membrane-bound organelle science knowledge that covers general truths or the operation of general laws, especially when acquired and tested by the scientific method scientific law a description, often in the form of a mathematical formula, for the behavior of some aspect of nature under certain specific conditions scientific method a method of research with defined steps that include experiments and careful observation scientific theory a thoroughly tested and confirmed explanation for observations or phenomena tissue a group of similar cells carrying out the same function variable a part of an experiment that can vary or change Chapter Summary 1.1 Themes and Concepts of Biology Biology is the science of life. All living organisms share several key properties such as order, sensitivity or response to stimuli, reproduction, adaptation, growth and development, regulation, homeostasis, and energy processing. Living things are highly organized following a hierarchy that includes atoms, molecules, organelles, cells, tissues, organs, and organ systems. Organisms, in turn, are grouped as populations, communities, ecosystems, and the biosphere. Evolution is the source of the tremendous biological diversity on Earth today. A diagram called a phylogenetic tree can be used to show evolutionary relationships among organisms. Biology is very broad and includes many branches and sub disciplines. Examples include molecular biology, microbiology, neurobiology, zoology, and botany, among others. 24 1 • Key Terms Access for free at openstax.org'},\n",
       " {'page_number': 38,\n",
       "  'page_char_count': 3202,\n",
       "  'page_word_count': 488,\n",
       "  'page_sentence_count_raw': 61,\n",
       "  'page_token_count': 800.5,\n",
       "  'text': '1.2 The Process of Science Biology is the science that studies living organisms and their interactions with one another and their environments. Science attempts to describe and understand the nature of the universe in whole or in part. Science has many fields; those fields related to the physical world and its phenomena are considered natural sciences. A hypothesis is a tentative explanation for an observation. A generally accepted scientific theory is thoroughly tested and confirmed explanation for a set of observations or phenomena. A scientific law is a description, often in the form of a mathematical formula, of the behavior of an aspect of nature under certain circumstances. Two types of logical reasoning are used in science. Inductive reasoning uses results to produce general scientific principles. Deductive reasoning is a form of logical thinking that predicts results by applying general principles. The common thread throughout scientific research is the use of the scientific method. Scientists present their results in peer-reviewed scientific papers published in scientific journals. Science can be basic or applied. The main goal of basic science is to expand knowledge without any expectation of short-term practical application of that knowledge. The primary goal of applied research, however, is to solve practical problems. Visual Connection Questions 1. Figure 1.8 Which of the following statements is false? A. Tissues exist within organs which exist within organ systems. B. Communities exist within populations which exist within ecosystems. C. Organelles exist within cells which exist within tissues. D. Communities exist within ecosystems which exist in the biosphere. 2. Figure 1.18 In the example below, the scientific method is used to solve an everyday problem. Which part in the example below is the hypothesis? Which is the prediction? Based on the results of the experiment, is the hypothesis supported? If it is not supported, propose some alternative hypotheses. 1. My toaster doesn’t toast my bread. 2. Why doesn’t my toaster work? 3. There is something wrong with the electrical outlet. 4. If something is wrong with the outlet, my coffeemaker also won’t work when plugged into it. 5. I plug my coffeemaker into the outlet. 6. My coffeemaker works. Review Questions 3. The smallest unit of biological structure that meets the functional requirements of “living” is the ________. a. organ b. organelle c. cell d. macromolecule 4. Which of the following sequences represents the hierarchy of biological organization from the most complex to the least complex level? a. organelle, tissue, biosphere, ecosystem, population b. organ, organism, tissue, organelle, molecule c. organism, community, biosphere, molecule, tissue, organ d. biosphere, ecosystem, community, population, organism 5. A suggested and testable explanation for an event is called a ________. a. hypothesis b. variable c. theory d. control 6. The type of logical thinking that uses related observations to arrive at a general conclusion is called ________. a. deductive reasoning b. the scientific method c. hypothesis-based science d. inductive reasoning 1 • Visual Connection Questions 25'},\n",
       " {'page_number': 39,\n",
       "  'page_char_count': 284,\n",
       "  'page_word_count': 48,\n",
       "  'page_sentence_count_raw': 5,\n",
       "  'page_token_count': 71.0,\n",
       "  'text': 'Critical Thinking Questions 7. Using examples, explain how biology can be studied from a microscopic approach to a global approach. 8. Give an example of how applied science has had a direct effect on your daily life. 26 1 • Critical Thinking Questions Access for free at openstax.org'},\n",
       " {'page_number': 40,\n",
       "  'page_char_count': 1272,\n",
       "  'page_word_count': 192,\n",
       "  'page_sentence_count_raw': 9,\n",
       "  'page_token_count': 318.0,\n",
       "  'text': 'INTRODUCTION CHAPTER 2 Chemistry of Life 2.1 The Building Blocks of Molecules 2.2 Water 2.3 Biological Molecules The elements carbon, hydrogen, nitrogen, oxygen, sulfur, and phosphorus are the key building blocks of the chemicals found in living things. They form the carbohydrates, nucleic acids, proteins, and lipids (all of which will be defined later in this chapter) that are the fundamental molecular components of all organisms. In this chapter, we will discuss these important building blocks and learn how the unique properties of the atoms of different elements affect their interactions with other atoms to form the molecules of life. Food provides an organism with nutrients—the matter it needs to survive. Many of these critical nutrients come in the form of biological macromolecules, or large molecules necessary for life. These macromolecules are built from different combinations of smaller organic molecules. What specific types of biological macromolecules do living things require? How are these molecules formed? What functions do they serve? In this chapter, we will explore these questions. FIGURE 2.1 Foods such as bread, fruit, and cheese are rich sources of biological macromolecules. (credit: modification of work by Bengt Nyman) CHAPTER OUTLINE'},\n",
       " {'page_number': 41,\n",
       "  'page_char_count': 2841,\n",
       "  'page_word_count': 487,\n",
       "  'page_sentence_count_raw': 27,\n",
       "  'page_token_count': 710.25,\n",
       "  'text': '2.1 The Building Blocks of Molecules LEARNING OBJECTIVES By the end of this section, you will be able to: • Describe matter and elements • Describe the interrelationship between protons, neutrons, and electrons, and the ways in which electrons can be donated or shared between atoms At its most fundamental level, life is made up of matter. Matter occupies space and has mass. All matter is composed of elements, substances that cannot be broken down or transformed chemically into other substances. Each element is made of atoms, each with a constant number of protons and unique properties. A total of 118 elements have been defined; however, only 92 occur naturally, and fewer than 30 are found in living cells. The remaining 26 elements are unstable and, therefore, do not exist for very long or are theoretical and have yet to be detected. Each element is designated by its chemical symbol (such as H, N, O, C, and Na), and possesses unique properties. These unique properties allow elements to combine and to bond with each other in specific ways. Atoms An atom is the smallest component of an element that retains all of the chemical properties of that element. For example, one hydrogen atom has all of the properties of the element hydrogen, such as it exists as a gas at room temperature, and it bonds with oxygen to create a water molecule. Hydrogen atoms cannot be broken down into anything smaller while still retaining the properties of hydrogen. If a hydrogen atom were broken down into subatomic particles, it would no longer have the properties of hydrogen. At the most basic level, all organisms are made of a combination of elements. They contain atoms that combine together to form molecules. In multicellular organisms, such as animals, molecules can interact to form cells that combine to form tissues, which make up organs. These combinations continue until entire multicellular organisms are formed. All atoms contain protons, electrons, and neutrons (Figure 2.2). The most common isotope of hydrogen (H) is the only exception and is made of one proton and one electron with no neutrons. A proton is a positively charged particle that resides in the nucleus (the core of the atom) of an atom and has a mass of 1 and a charge of +1. An electron is a negatively charged particle that travels in the space around the nucleus. In other words, it resides outside of the nucleus. It has a negligible mass and has a charge of –1. FIGURE 2.2 Atoms are made up of protons and neutrons located within the nucleus, and electrons surrounding the nucleus. Neutrons, like protons, reside in the nucleus of an atom. They have a mass of 1 and no charge. The positive (protons) and negative (electrons) charges balance each other in a neutral atom, which has a net zero charge. 28 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 42,\n",
       "  'page_char_count': 2466,\n",
       "  'page_word_count': 408,\n",
       "  'page_sentence_count_raw': 25,\n",
       "  'page_token_count': 616.5,\n",
       "  'text': 'Because protons and neutrons each have a mass of 1, the mass of an atom is equal to the number of protons and neutrons of that atom. The number of electrons does not factor into the overall mass, because their mass is so small. As stated earlier, each element has its own unique properties. Each contains a different number of protons and neutrons, giving it its own atomic number and mass number. The atomic number of an element is equal to the number of protons that element contains. The mass number, or atomic mass, is the number of protons plus the number of neutrons of that element. Therefore, it is possible to determine the number of neutrons by subtracting the atomic number from the mass number. These numbers provide information about the elements and how they will react when combined. Different elements have different melting and boiling points, and are in different states (liquid, solid, or gas) at room temperature. They also combine in different ways. Some form specific types of bonds, whereas others do not. How they combine is based on the number of electrons present. Because of these characteristics, the elements are arranged into the periodic table of elements, a chart of the elements that includes the atomic number and relative atomic mass of each element. The periodic table also provides key information about the properties of elements (Figure 2.2)—often indicated by color-coding. The arrangement of the table also shows how the electrons in each element are organized and provides important details about how atoms will react with each other to form molecules. Isotopes are different forms of the same element that have the same number of protons, but a different number of neutrons. Some elements, such as carbon, potassium, and uranium, have naturally occurring isotopes. Carbon-12, the most common isotope of carbon, contains six protons and six neutrons. Therefore, it has a mass number of 12 (six protons and six neutrons) and an atomic number of 6 (which makes it carbon). Carbon-14 contains six protons and eight neutrons. Therefore, it has a mass number of 14 (six protons and eight neutrons) and an atomic number of 6, meaning it is still the element carbon. These two alternate forms of carbon are isotopes. Some isotopes are unstable and will lose protons, other subatomic particles, or energy to form more stable elements. These are called radioactive isotopes or radioisotopes. 2.1 • The Building Blocks of Molecules 29'},\n",
       " {'page_number': 43,\n",
       "  'page_char_count': 1809,\n",
       "  'page_word_count': 307,\n",
       "  'page_sentence_count_raw': 15,\n",
       "  'page_token_count': 452.25,\n",
       "  'text': 'VISUAL CONNECTION FIGURE 2.3 Arranged in columns and rows based on the characteristics of the elements, the periodic table provides key information about the elements and how they might interact with each other to form molecules. Most periodic tables provide a key or legend to the information they contain. How many neutrons do (K) potassium-39 and potassium-40 have, respectively? EVOLUTION CONNECTION Carbon Dating Carbon-14 (14C) is a naturally occurring radioisotope that is created in the atmosphere by cosmic rays. This is a continuous process, so more 14C is always being created. As a living organism develops, the relative level of 14C in its body is equal to the concentration of 14C in the atmosphere. When an organism dies, it is no longer ingesting 14C, so the ratio will decline. 14C decays to 14N by a process called beta decay; it gives off energy in this slow process. After approximately 5,730 years, only one-half of the starting concentration of 14C will have been converted to 14N. The time it takes for half of the original concentration of an isotope to decay to its more stable form is called its half- life. Because the half-life of 14C is long, it is used to age formerly living objects, such as fossils. Using the ratio of the 14C concentration found in an object to the amount of 14C detected in the atmosphere, the amount of the isotope that has not yet decayed can be determined. Based on this amount, the age of the fossil can be calculated to about 50,000 years (Figure 2.4). Isotopes with longer half-lives, such as potassium-40, are used to calculate the ages of older fossils. Through the use of carbon dating, scientists can reconstruct the ecology and biogeography of organisms living within the past 50,000 years. 30 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 44,\n",
       "  'page_char_count': 3038,\n",
       "  'page_word_count': 515,\n",
       "  'page_sentence_count_raw': 24,\n",
       "  'page_token_count': 759.5,\n",
       "  'text': 'FIGURE 2.4 The age of remains that contain carbon and are less than about 50,000 years old, such as this pygmy mammoth, can be determined using carbon dating. (credit: Bill Faulkner/NPS) LINK TO LEARNING To learn more about atoms and isotopes, and how you can tell one isotope from another, visit this site (http://openstax.org/l/isotopes) and run the simulation. Chemical Bonds How elements interact with one another depends on how their electrons are arranged and how many openings for electrons exist at the outermost region where electrons are present in an atom. Electrons exist at energy levels that form shells around the nucleus. The closest shell can hold up to two electrons. The closest shell to the nucleus is always filled first, before any other shell can be filled. Hydrogen has one electron; therefore, it has only one spot occupied within the lowest shell. Helium has two electrons; therefore, it can completely fill the lowest shell with its two electrons. If you look at the periodic table, you will see that hydrogen and helium are the only two elements in the first row. This is because they only have electrons in their first shell. Hydrogen and helium are the only two elements that have the lowest shell and no other shells. The second and third energy levels can hold up to eight electrons. The eight electrons are arranged in four pairs and one position in each pair is filled with an electron before any pairs are completed. Looking at the periodic table again (Figure 2.3), you will notice that there are seven rows. These rows correspond to the number of shells that the elements within that row have. The elements within a particular row have increasing numbers of electrons as the columns proceed from left to right. Although each element has the same number of shells, not all of the shells are completely filled with electrons. If you look at the second row of the periodic table, you will find lithium (Li), beryllium (Be), boron (B), carbon (C), nitrogen (N), oxygen (O), fluorine (F), and neon (Ne). These all have electrons that occupy only the first and second shells. Lithium has only one electron in its outermost shell, beryllium has two electrons, boron has three, and so on, until the entire shell is filled with eight electrons, as is the case with neon. Not all elements have enough electrons to fill their outermost shells, but an atom is at its most stable when all of the electron positions in the outermost shell are filled. Because of these vacancies in the outermost shells, we see the formation of chemical bonds, or interactions between two or more of the same or different elements that result in the formation of molecules. To achieve greater stability, atoms will tend to completely fill their outer shells and will bond with other elements to accomplish this goal by sharing electrons, accepting electrons from another atom, or donating electrons to another atom. Because the outermost shells of the elements with low atomic numbers (up to 2.1 • The Building Blocks of Molecules 31'},\n",
       " {'page_number': 45,\n",
       "  'page_char_count': 3224,\n",
       "  'page_word_count': 562,\n",
       "  'page_sentence_count_raw': 32,\n",
       "  'page_token_count': 806.0,\n",
       "  'text': 'calcium, with atomic number 20) can hold eight electrons, this is referred to as the octet rule. An element can donate, accept, or share electrons with other elements to fill its outer shell and satisfy the octet rule. When an atom does not contain equal numbers of protons and electrons, it is called an ion. Because the number of electrons does not equal the number of protons, each ion has a net charge. Positive ions are formed by losing electrons and are called cations. Negative ions are formed by gaining electrons and are called anions. For example, sodium only has one electron in its outermost shell. It takes less energy for sodium to donate that one electron than it does to accept seven more electrons to fill the outer shell. If sodium loses an electron, it now has 11 protons and only 10 electrons, leaving it with an overall charge of +1. It is now called a sodium ion. The chlorine atom has seven electrons in its outer shell. Again, it is more energy-efficient for chlorine to gain one electron than to lose seven. Therefore, it tends to gain an electron to create an ion with 17 protons and 18 electrons, giving it a net negative (–1) charge. It is now called a chloride ion. This movement of electrons from one element to another is referred to as electron transfer. As Figure 2.5 illustrates, a sodium atom (Na) only has one electron in its outermost shell, whereas a chlorine atom (Cl) has seven electrons in its outermost shell. A sodium atom will donate its one electron to empty its shell, and a chlorine atom will accept that electron to fill its shell, becoming chloride. Both ions now satisfy the octet rule and have complete outermost shells. Because the number of electrons is no longer equal to the number of protons, each is now an ion and has a +1 (sodium) or –1 (chloride) charge. FIGURE 2.5 Elements tend to fill their outermost shells with electrons. To do this, they can either donate or accept electrons from other elements. Ionic Bonds There are four types of bonds or interactions: ionic, covalent, hydrogen bonds, and van der Waals interactions. Ionic and covalent bonds are strong interactions that require a larger energy input to break apart. When an element donates an electron from its outer shell, as in the sodium atom example above, a positive ion is formed. The element accepting the electron is now negatively charged. Because positive and negative charges attract, these ions stay together and form an ionic bond, or a bond between ions. The elements bond together with the electron from one element staying predominantly with the other element. When Na+ and Cl– ions combine to produce NaCl, an electron from a sodium atom stays with the other seven from the chlorine atom, and the sodium and chloride ions attract each other in a lattice of ions with a net zero charge. Covalent Bonds Another type of strong chemical bond between two or more atoms is a covalent bond. These bonds form when an electron is shared between two elements and are the strongest and most common form of chemical bond in living organisms. Covalent bonds form between the elements that make up the biological molecules in our cells. Unlike 32 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 46,\n",
       "  'page_char_count': 3835,\n",
       "  'page_word_count': 629,\n",
       "  'page_sentence_count_raw': 36,\n",
       "  'page_token_count': 958.75,\n",
       "  'text': 'ionic bonds, covalent bonds do not dissociate in water. The hydrogen and oxygen atoms that combine to form water molecules are bound together by covalent bonds. The electron from the hydrogen atom divides its time between the outer shell of the hydrogen atom and the incomplete outer shell of the oxygen atom. To completely fill the outer shell of an oxygen atom, two electrons from two hydrogen atoms are needed, hence the subscript “2” in H2O. The electrons are shared between the atoms, dividing their time between them to “fill” the outer shell of each. This sharing is a lower energy state for all of the atoms involved than if they existed without their outer shells filled. There are two types of covalent bonds: polar and nonpolar. Nonpolar covalent bonds form between two atoms of the same element or between different elements that share the electrons equally. For example, an oxygen atom can bond with another oxygen atom to fill their outer shells. This association is nonpolar because the electrons will be equally distributed between each oxygen atom. Two covalent bonds form between the two oxygen atoms because oxygen requires two shared electrons to fill its outermost shell. Nitrogen atoms will form three covalent bonds (also called triple covalent) between two atoms of nitrogen because each nitrogen atom needs three electrons to fill its outermost shell. Another example of a nonpolar covalent bond is found in the methane (CH4) molecule. The carbon atom has four electrons in its outermost shell and needs four more to fill it. It gets these four from four hydrogen atoms, each atom providing one. These elements all share the electrons equally, creating four nonpolar covalent bonds (Figure 2.6). In a polar covalent bond, the electrons shared by the atoms spend more time closer to one nucleus than to the other nucleus. Because of the unequal distribution of electrons between the different nuclei, a slightly positive (δ+) or slightly negative (δ–) charge develops. The covalent bonds between hydrogen and oxygen atoms in water are polar covalent bonds. The shared electrons spend more time near the oxygen nucleus, giving it a small negative charge, than they spend near the hydrogen nuclei, giving these molecules a small positive charge. FIGURE 2.6 The water molecule (left) depicts a polar bond with a slightly positive charge on the hydrogen atoms and a slightly negative charge on the oxygen. Examples of nonpolar bonds include methane (middle) and oxygen (right). Hydrogen Bonds Ionic and covalent bonds are strong bonds that require considerable energy to break. However, not all bonds between elements are ionic or covalent bonds. Weaker bonds can also form. These are attractions that occur between positive and negative charges that do not require much energy to break. Two weak bonds that occur frequently are hydrogen bonds and van der Waals interactions. These bonds give rise to the unique properties of water and the unique structures of DNA and proteins. When polar covalent bonds containing a hydrogen atom form, the hydrogen atom in that bond has a slightly positive charge. This is because the shared electron is pulled more strongly toward the other element and away from the hydrogen nucleus. Because the hydrogen atom is slightly positive (δ+), it will be attracted to neighboring negative partial charges (δ–). When this happens, a weak interaction occurs between the δ+ charge of the hydrogen atom of one molecule and the δ– charge of the other molecule. This interaction is called a hydrogen bond. This type of bond is common; for example, the liquid nature of water is caused by the hydrogen bonds between water molecules (Figure 2.7). Hydrogen bonds give water the unique properties that sustain life. If it were not for hydrogen bonding, 2.1 • The Building Blocks of Molecules 33'},\n",
       " {'page_number': 47,\n",
       "  'page_char_count': 3000,\n",
       "  'page_word_count': 465,\n",
       "  'page_sentence_count_raw': 22,\n",
       "  'page_token_count': 750.0,\n",
       "  'text': 'water would be a gas rather than a liquid at room temperature. FIGURE 2.7 Hydrogen bonds form between slightly positive (δ+) and slightly negative (δ–) charges of polar covalent molecules, such as water. Hydrogen bonds can form between different molecules and they do not always have to include a water molecule. Hydrogen atoms in polar bonds within any molecule can form bonds with other adjacent molecules. For example, hydrogen bonds hold together two long strands of DNA to give the DNA molecule its characteristic double-stranded structure. Hydrogen bonds are also responsible for some of the three-dimensional structure of proteins. van der Waals Interactions Like hydrogen bonds, van der Waals interactions are weak attractions or interactions between molecules. They occur between polar, covalently bound, atoms in different molecules. Some of these weak attractions are caused by temporary partial charges formed when electrons move around a nucleus. These weak interactions between molecules are important in biological systems. CAREER CONNECTION Radiography Technician Have you or anyone you know ever had a magnetic resonance imaging (MRI) scan, a mammogram, or an X-ray? These tests produce images of your soft tissues and organs (as with an MRI or mammogram) or your bones (as happens in an X-ray) by using either radiowaves or special isotopes (radiolabeled or fluorescently labeled) that are ingested or injected into the body. These tests provide data for disease diagnoses by creating images of your organs or skeletal system. MRI imaging works by subjecting hydrogen nuclei, which are abundant in the water in soft tissues, to fluctuating magnetic fields, which cause them to emit their own magnetic field. This signal is then read by sensors in the machine and interpreted by a computer to form a detailed image. Some radiography technologists and technicians specialize in computed tomography, MRI, and mammography. They produce films or images of the body that help medical professionals examine and diagnose. Radiologists work directly with patients, explaining machinery, preparing them for exams, and ensuring that their body or body parts are positioned correctly to produce the needed images. Physicians or radiologists then analyze the test results. Radiography technicians can work in hospitals, doctors’ offices, or specialized imaging centers. Training to become a radiography technician happens at hospitals, colleges, and universities that offer certificates, associate’s degrees, or bachelor’s degrees in radiography. 2.2 Water LEARNING OBJECTIVES By the end of this section, you will be able to: • Describe the properties of water that are critical to maintaining life Do you ever wonder why scientists spend time looking for water on other planets? It is because water is essential to life; even minute traces of it on another planet can indicate that life could or did exist on that planet. Water is one of 34 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 48,\n",
       "  'page_char_count': 3361,\n",
       "  'page_word_count': 551,\n",
       "  'page_sentence_count_raw': 32,\n",
       "  'page_token_count': 840.25,\n",
       "  'text': 'the more abundant molecules in living cells and the one most critical to life as we know it. Approximately 60–70 percent of your body is made up of water. Without it, life simply would not exist. Water Is Polar The hydrogen and oxygen atoms within water molecules form polar covalent bonds. The shared electrons spend more time associated with the oxygen atom than they do with hydrogen atoms. There is no overall charge to a water molecule, but there is a slight positive charge on each hydrogen atom and a slight negative charge on the oxygen atom. Because of these charges, the slightly positive hydrogen atoms repel each other and form the unique shape seen in Figure 2.7. Each water molecule attracts other water molecules because of the positive and negative charges in the different parts of the molecule. Water also attracts other polar molecules (such as sugars), forming hydrogen bonds. When a substance readily forms hydrogen bonds with water, it can dissolve in water and is referred to as hydrophilic (“water-loving”). Hydrogen bonds are not readily formed with nonpolar substances like oils and fats (Figure 2.8). These nonpolar compounds are hydrophobic (“water-fearing”) and will not dissolve in water. FIGURE 2.8 As this macroscopic image of oil and water show, oil is a nonpolar compound and, hence, will not dissolve in water. Oil and water do not mix. (credit: Gautam Dogra) Water Stabilizes Temperature The hydrogen bonds in water allow it to absorb and release heat energy more slowly than many other substances. Temperature is a measure of the motion (kinetic energy) of molecules. As the motion increases, energy is higher and thus temperature is higher. Water absorbs a great deal of energy before its temperature rises. Increased energy disrupts the hydrogen bonds between water molecules. Because these bonds can be created and disrupted rapidly, water absorbs an increase in energy and temperature changes only minimally. This means that water moderates temperature changes within organisms and in their environments. As energy input continues, the balance between hydrogen-bond formation and destruction swings toward the destruction side. More bonds are broken than are formed. This process results in the release of individual water molecules at the surface of the liquid (such as a body of water, the leaves of a plant, or the skin of an organism) in a process called evaporation. Evaporation of sweat, which is 90 percent water, allows for cooling of an organism, because breaking hydrogen bonds requires an input of energy and takes heat away from the body. Conversely, as molecular motion decreases and temperatures drop, less energy is present to break the hydrogen bonds between water molecules. These bonds remain intact and begin to form a rigid, lattice-like structure (e.g., ice) (Figure 2.9a). When frozen, ice is less dense than liquid water (the molecules are farther apart). This means that ice floats on the surface of a body of water (Figure 2.9b). In lakes, ponds, and oceans, ice will form on the surface of the water, creating an insulating barrier to protect the animal and plant life beneath from freezing in the water. If this did not happen, plants and animals living in water would freeze in a block of ice and could not move freely, making life in cold temperatures difficult or impossible. 2.2 • Water 35'},\n",
       " {'page_number': 49,\n",
       "  'page_char_count': 1625,\n",
       "  'page_word_count': 268,\n",
       "  'page_sentence_count_raw': 14,\n",
       "  'page_token_count': 406.25,\n",
       "  'text': \"FIGURE 2.9 (a) The lattice structure of ice makes it less dense than the freely flowing molecules of liquid water. Ice's lower density enables it to (b) float on water. (credit a: modification of work by Jane Whitney; credit b: modification of work by Carlos Ponte) LINK TO LEARNING Click here (http://openstax.org/l/ice_lattice) to see a 3-D animation of the structure of an ice lattice. Water Is an Excellent Solvent Because water is polar, with slight positive and negative charges, ionic compounds and polar molecules can readily dissolve in it. Water is, therefore, what is referred to as a solvent—a substance capable of dissolving another substance. The charged particles will form hydrogen bonds with a surrounding layer of water molecules. This is referred to as a sphere of hydration and serves to keep the particles separated or dispersed in the water. In the case of table salt (NaCl) mixed in water (Figure 2.10), the sodium and chloride ions separate, or dissociate, in the water, and spheres of hydration are formed around the ions. A positively charged sodium ion is surrounded by the partially negative charges of oxygen atoms in water molecules. A negatively charged chloride ion is surrounded by the partially positive charges of hydrogen atoms in water molecules. These spheres of hydration are also referred to as hydration shells. The polarity of the water molecule makes it an effective solvent and is important in its many roles in living systems. FIGURE 2.10 When table salt (NaCl) is mixed in water, spheres of hydration form around the ions. 36 2 • Chemistry of Life Access for free at openstax.org\"},\n",
       " {'page_number': 50,\n",
       "  'page_char_count': 3354,\n",
       "  'page_word_count': 596,\n",
       "  'page_sentence_count_raw': 29,\n",
       "  'page_token_count': 838.5,\n",
       "  'text': 'Water Is Cohesive Have you ever filled up a glass of water to the very top and then slowly added a few more drops? Before it overflows, the water actually forms a dome-like shape above the rim of the glass. This water can stay above the glass because of the property of cohesion. In cohesion, water molecules are attracted to each other (because of hydrogen bonding), keeping the molecules together at the liquid-air (gas) interface, although there is no more room in the glass. Cohesion gives rise to surface tension, the capacity of a substance to withstand rupture when placed under tension or stress. When you drop a small scrap of paper onto a droplet of water, the paper floats on top of the water droplet, although the object is denser (heavier) than the water. This occurs because of the surface tension that is created by the water molecules. Cohesion and surface tension keep the water molecules intact and the item floating on the top. It is even possible to “float” a steel needle on top of a glass of water if you place it gently, without breaking the surface tension (Figure 2.11). FIGURE 2.11 The weight of a needle on top of water pulls the surface tension downward; at the same time, the surface tension of the water is pulling it up, suspending the needle on the surface of the water and keeping it from sinking. Notice the indentation in the water around the needle. (credit: Cory Zanker) These cohesive forces are also related to the water’s property of adhesion, or the attraction between water molecules and other molecules. This is observed when water “climbs” up a straw placed in a glass of water. You will notice that the water appears to be higher on the sides of the straw than in the middle. This is because the water molecules are attracted to the straw and therefore adhere to it. Cohesive and adhesive forces are important for sustaining life. For example, because of these forces, water can flow up from the roots to the tops of plants to feed the plant. LINK TO LEARNING To learn more about water, visit the U.S. Geological Survey Water Science for Schools: All About Water! website. (http://openstax.org/l/about_water) Buffers, pH, Acids, and Bases The pH of a solution is a measure of its acidity or bascicity. You have probably used litmus paper, paper that has been treated with a natural water-soluble dye so it can be used as a pH indicator, to test how much acid or base (basicity) exists in a solution. You might have even used some to make sure the water in an outdoor swimming pool is properly treated. In both cases, this pH test measures the amount of hydrogen ions that exists in a given solution. High concentrations of hydrogen ions yield a low pH, whereas low levels of hydrogen ions result in a high pH. The overall concentration of hydrogen ions is inversely related to its pH and can be measured on the pH scale (Figure 2.12). Therefore, the more hydrogen ions present, the lower the pH; conversely, the fewer hydrogen ions, the higher the pH. The pH scale ranges from 0 to 14. A change of one unit on the pH scale represents a change in the concentration of hydrogen ions by a factor of 10, a change in two units represents a change in the concentration of hydrogen ions by a factor of 100. Thus, small changes in pH represent large changes in the concentrations of hydrogen ions. Pure 2.2 • Water 37'},\n",
       " {'page_number': 51,\n",
       "  'page_char_count': 2548,\n",
       "  'page_word_count': 450,\n",
       "  'page_sentence_count_raw': 27,\n",
       "  'page_token_count': 637.0,\n",
       "  'text': 'water is neutral. It is neither acidic nor basic, and has a pH of 7.0. Anything below 7.0 (ranging from 0.0 to 6.9) is acidic, and anything above 7.0 (from 7.1 to 14.0) is alkaline. The blood in your veins is slightly alkaline (pH = 7.4). The environment in your stomach is highly acidic (pH = 1 to 2). Orange juice is mildly acidic (pH = approximately 3.5), whereas baking soda is basic (pH = 9.0). FIGURE 2.12 The pH scale measures the amount of hydrogen ions (H+) in a substance. (credit: modification of work by Edward Stevens) Acids are substances that provide hydrogen ions (H+) and lower pH, whereas bases provide hydroxide ions (OH–) and raise pH. The stronger the acid, the more readily it donates H+. For example, hydrochloric acid and lemon juice are very acidic and readily give up H+ when added to water. Conversely, bases are those substances that readily donate OH–. The OH– ions combine with H+ to produce water, which raises a substance’s pH. Sodium hydroxide and many household cleaners are very alkaline and give up OH– rapidly when placed in water, thereby raising the pH. Most cells in our bodies operate within a very narrow window of the pH scale, typically ranging only from 7.2 to 7.6. If the pH of the body is outside of this range, the respiratory system malfunctions, as do other organs in the body. Cells no longer function properly, and proteins will break down. Deviation outside of the pH range can induce coma or even cause death. So how is it that we can ingest or inhale acidic or basic substances and not die? Buffers are the key. Buffers readily absorb excess H+ or OH–, keeping the pH of the body carefully maintained in the aforementioned narrow range. Carbon dioxide is part of a prominent buffer system in the human body; it keeps the pH within the proper range. This buffer system involves carbonic acid (H2CO3) and bicarbonate (HCO3–) anion. If too much H+ enters the body, bicarbonate will combine with the H+ to create carbonic acid and limit the decrease in pH. Likewise, if too much OH– is introduced into the system, carbonic acid will rapidly dissociate into bicarbonate and H+ ions. The H+ ions can combine with the OH– ions, limiting the increase in pH. While carbonic acid is an important product in this reaction, its presence is fleeting because the carbonic acid is released from the body as carbon dioxide gas each time we breathe. Without this buffer system, the pH in our bodies would fluctuate too much and we would fail to survive. 38 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 52,\n",
       "  'page_char_count': 2480,\n",
       "  'page_word_count': 397,\n",
       "  'page_sentence_count_raw': 19,\n",
       "  'page_token_count': 620.0,\n",
       "  'text': '2.3 Biological Molecules LEARNING OBJECTIVES By the end of this section, you will be able to: • Describe the ways in which carbon is critical to life • Explain the impact of slight changes in amino acids on organisms • Describe the four major types of biological molecules • Understand the functions of the four major types of molecules The large molecules necessary for life that are built from smaller organic molecules are called biological macromolecules. There are four major classes of biological macromolecules (carbohydrates, lipids, proteins, and nucleic acids), and each is an important component of the cell and performs a wide array of functions. Combined, these molecules make up the majority of a cell’s dry mass. Biological macromolecules are organic, meaning they contain carbon and are bound to hydrogen, and may contain oxygen, nitrogen, and additional minor elements. Carbon It is often said that life is “carbon-based.” This means that carbon atoms, bonded to other carbon atoms or other elements, form the fundamental components of many, if not most, of the molecules found uniquely in living things. Other elements play important roles in biological molecules, but carbon certainly qualifies as the “foundation” element for molecules in living things. It is the bonding properties of carbon atoms that are responsible for its important role. Carbon Bonding Carbon contains four electrons in its outer shell. Therefore, it can form four covalent bonds with other atoms or molecules. The simplest organic carbon molecule is methane (CH4), in which four hydrogen atoms bind to a carbon atom (Figure 2.13). FIGURE 2.13 Carbon can form four covalent bonds to create an organic molecule. The simplest carbon molecule is methane (CH4), depicted here. However, structures that are more complex are made using carbon. Any of the hydrogen atoms can be replaced with another carbon atom covalently bonded to the first carbon atom. In this way, long and branching chains of carbon compounds can be made (Figure 2.14a). The carbon atoms may bond with atoms of other elements, such as nitrogen, oxygen, and phosphorus (Figure 2.14b). The molecules may also form rings, which themselves can link with other rings (Figure 2.14c). This diversity of molecular forms accounts for the diversity of functions of the biological macromolecules and is based to a large degree on the ability of carbon to form multiple bonds with itself and other atoms. 2.3 • Biological Molecules 39'},\n",
       " {'page_number': 53,\n",
       "  'page_char_count': 2785,\n",
       "  'page_word_count': 431,\n",
       "  'page_sentence_count_raw': 26,\n",
       "  'page_token_count': 696.25,\n",
       "  'text': 'FIGURE 2.14 These examples show three molecules (found in living organisms) that contain carbon atoms bonded in various ways to other carbon atoms and the atoms of other elements. (a) This molecule of stearic acid has a long chain of carbon atoms. (b) Glycine, a component of proteins, contains carbon, nitrogen, oxygen, and hydrogen atoms. (c) Glucose, a sugar, has a ring of carbon atoms and one oxygen atom. Carbohydrates Carbohydrates are macromolecules with which most consumers are somewhat familiar. To lose weight, some individuals adhere to “low-carb” diets. Athletes, in contrast, often “carb-load” before important competitions to ensure that they have sufficient energy to compete at a high level. Carbohydrates are, in fact, an essential part of our diet; grains, fruits, and vegetables are all natural sources of carbohydrates. Carbohydrates provide energy to the body, particularly through glucose, a simple sugar. Carbohydrates also have other important functions in humans, animals, and plants. Carbohydrates can be represented by the formula (CH2O)n, where n is the number of carbon atoms in the molecule. In other words, the ratio of carbon to hydrogen to oxygen is 1:2:1 in carbohydrate molecules. Carbohydrates are classified into three subtypes: monosaccharides, disaccharides, and polysaccharides. Monosaccharides (mono- = “one”; sacchar- = “sweet”) are simple sugars, the most common of which is glucose. In monosaccharides, the number of carbon atoms usually ranges from three to six. Most monosaccharide names end with the suffix -ose. Depending on the number of carbon atoms in the sugar, they may be known as trioses (three carbon atoms), pentoses (five carbon atoms), and hexoses (six carbon atoms). Monosaccharides may exist as a linear chain or as ring-shaped molecules; in aqueous solutions, they are usually found in the ring form. The chemical formula for glucose is C6H12O6. In most living species, glucose is an important source of energy. During cellular respiration, energy is released from glucose, and that energy is used to help make adenosine triphosphate (ATP). Plants synthesize glucose using carbon dioxide and water by the process of photosynthesis, and the glucose, in turn, is used for the energy requirements of the plant. The excess synthesized glucose is often stored as starch that is broken down by other organisms that feed on plants. Galactose (part of lactose, or milk sugar) and fructose (found in fruit) are other common monosaccharides. Although glucose, galactose, and fructose all have the same chemical formula (C6H12O6), they differ structurally and chemically (and are known as isomers) because of differing arrangements of atoms in the carbon chain (Figure 2.15). 40 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 54,\n",
       "  'page_char_count': 3394,\n",
       "  'page_word_count': 532,\n",
       "  'page_sentence_count_raw': 33,\n",
       "  'page_token_count': 848.5,\n",
       "  'text': 'FIGURE 2.15 Glucose, galactose, and fructose are isomeric monosaccharides, meaning that they have the same chemical formula but slightly different structures. Disaccharides (di- = “two”) form when two monosaccharides undergo a dehydration reaction (a reaction in which the removal of a water molecule occurs). During this process, the hydroxyl group (–OH) of one monosaccharide combines with a hydrogen atom of another monosaccharide, releasing a molecule of water (H2O) and forming a covalent bond between atoms in the two sugar molecules. Common disaccharides include lactose, maltose, and sucrose. Lactose is a disaccharide consisting of the monomers glucose and galactose. It is found naturally in milk. Maltose, or malt sugar, is a disaccharide formed from a dehydration reaction between two glucose molecules. The most common disaccharide is sucrose, or table sugar, which is composed of the monomers glucose and fructose. A long chain of monosaccharides linked by covalent bonds is known as a polysaccharide (poly- = “many”). The chain may be branched or unbranched, and it may contain different types of monosaccharides. Polysaccharides may be very large molecules. Starch, glycogen, cellulose, and chitin are examples of polysaccharides. Starch is the stored form of sugars in plants and is made up of amylose and amylopectin (both polymers of glucose). Plants are able to synthesize glucose, and the excess glucose is stored as starch in different plant parts, including roots and seeds. The starch that is consumed by animals is broken down into smaller molecules, such as glucose. The cells can then absorb the glucose. Glycogen is the storage form of glucose in humans and other vertebrates, and is made up of monomers of glucose. Glycogen is the animal equivalent of starch and is a highly branched molecule usually stored in liver and muscle cells. Whenever glucose levels decrease, glycogen is broken down to release glucose. Cellulose is one of the most abundant natural biopolymers. The cell walls of plants are mostly made of cellulose, which provides structural support to the cell. Wood and paper are mostly cellulosic in nature. Cellulose is made up of glucose monomers that are linked by bonds between particular carbon atoms in the glucose molecule. Every other glucose monomer in cellulose is flipped over and packed tightly as extended long chains. This gives cellulose its rigidity and high tensile strength—which is so important to plant cells. Cellulose passing through our digestive system is called dietary fiber. While the glucose-glucose bonds in cellulose cannot be broken down by human digestive enzymes, herbivores such as cows, buffalos, and horses are able to digest grass that is rich in cellulose and use it as a food source. In these animals, certain species of bacteria reside in the digestive system of herbivores and secrete the enzyme cellulase. The appendix also contains bacteria that break down cellulose, giving it an important role in the digestive systems of some ruminants. Cellulases can break down cellulose into glucose monomers that can be used as an energy source by the animal. Carbohydrates serve other functions in different animals. Arthropods, such as insects, spiders, and crabs, have an outer skeleton, called the exoskeleton, which protects their internal body parts. This exoskeleton is made of the 2.3 • Biological Molecules 41'},\n",
       " {'page_number': 55,\n",
       "  'page_char_count': 2472,\n",
       "  'page_word_count': 363,\n",
       "  'page_sentence_count_raw': 22,\n",
       "  'page_token_count': 618.0,\n",
       "  'text': 'biological macromolecule chitin, which is a nitrogenous carbohydrate. It is made of repeating units of a modified sugar containing nitrogen. Thus, through differences in molecular structure, carbohydrates are able to serve the very different functions of energy storage (starch and glycogen) and structural support and protection (cellulose and chitin) (Figure 2.16). FIGURE 2.16 Although their structures and functions differ, all polysaccharide carbohydrates are made up of monosaccharides and have the chemical formula (CH2O)n. CAREER CONNECTION Registered Dietitian Obesity is a worldwide health concern, and many diseases, such as diabetes and heart disease, are becoming more prevalent because of obesity. This is one of the reasons why registered dietitians are increasingly sought after for advice. Registered dietitians help plan food and nutrition programs for individuals in various settings. They often work with patients in health-care facilities, designing nutrition plans to prevent and treat diseases. For example, dietitians may teach a patient with diabetes how to manage blood-sugar levels by eating the correct types and amounts of carbohydrates. Dietitians may also work in nursing homes, schools, and private practices. To become a registered dietitian, one needs to earn at least a bachelor’s degree in dietetics, nutrition, food technology, or a related field. In addition, registered dietitians must complete a supervised internship program and pass a national exam. Those who pursue careers in dietetics take courses in nutrition, chemistry, biochemistry, biology, microbiology, and human physiology. Dietitians must become experts in the chemistry and functions of food (proteins, carbohydrates, and fats). Lipids Lipids include a diverse group of compounds that are united by a common feature. Lipids are hydrophobic (“water- fearing”), or insoluble in water, because they are nonpolar molecules. This is because they are hydrocarbons that include only nonpolar carbon-carbon or carbon-hydrogen bonds. Lipids perform many different functions in a cell. Cells store energy for long-term use in the form of lipids called fats. Lipids also provide insulation from the environment for plants and animals (Figure 2.17). For example, they help keep aquatic birds and mammals dry because of their water-repelling nature. Lipids are also the building blocks of many hormones and are an important 42 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 56,\n",
       "  'page_char_count': 1231,\n",
       "  'page_word_count': 202,\n",
       "  'page_sentence_count_raw': 12,\n",
       "  'page_token_count': 307.75,\n",
       "  'text': 'constituent of the plasma membrane. Lipids include fats, oils, waxes, phospholipids, and steroids. FIGURE 2.17 Hydrophobic lipids in the fur of aquatic mammals, such as this river otter, protect them from the elements. (credit: Ken Bosma) A fat molecule, such as a triglyceride, consists of two main components—glycerol and fatty acids. Glycerol is an organic compound with three carbon atoms, five hydrogen atoms, and three hydroxyl (–OH) groups. Fatty acids have a long chain of hydrocarbons to which an acidic carboxyl group is attached, hence the name “fatty acid.” The number of carbons in the fatty acid may range from 4 to 36; most common are those containing 12–18 carbons. In a fat molecule, a fatty acid is attached to each of the three oxygen atoms in the –OH groups of the glycerol molecule with a covalent bond (Figure 2.18). FIGURE 2.18 Lipids include fats, such as triglycerides, which are made up of fatty acids and glycerol, phospholipids, and steroids. During this covalent bond formation, three water molecules are released. The three fatty acids in the fat may be similar or dissimilar. These fats are also called triglycerides because they have three fatty acids. Some fatty acids 2.3 • Biological Molecules 43'},\n",
       " {'page_number': 57,\n",
       "  'page_char_count': 3434,\n",
       "  'page_word_count': 557,\n",
       "  'page_sentence_count_raw': 32,\n",
       "  'page_token_count': 858.5,\n",
       "  'text': 'have common names that specify their origin. For example, palmitic acid, a saturated fatty acid, is derived from the palm tree. Arachidic acid is derived from Arachis hypogaea, the scientific name for peanuts. Fatty acids may be saturated or unsaturated. In a fatty acid chain, if there are only single bonds between neighboring carbons in the hydrocarbon chain, the fatty acid is saturated. Saturated fatty acids are saturated with hydrogen; in other words, the number of hydrogen atoms attached to the carbon skeleton is maximized. When the hydrocarbon chain contains a double bond, the fatty acid is an unsaturated fatty acid. Most unsaturated fats are liquid at room temperature and are called oils. If there is one double bond in the molecule, then it is known as a monounsaturated fat (e.g., olive oil), and if there is more than one double bond, then it is known as a polyunsaturated fat (e.g., canola oil). Saturated fats tend to get packed tightly and are solid at room temperature. Animal fats with stearic acid and palmitic acid contained in meat, and the fat with butyric acid contained in butter, are examples of saturated fats. Mammals store fats in specialized cells called adipocytes, where globules of fat occupy most of the cell. In plants, fat or oil is stored in seeds and is used as a source of energy during embryonic development. Unsaturated fats or oils are usually of plant origin and contain unsaturated fatty acids. The double bond causes a bend or a “kink” that prevents the fatty acids from packing tightly, keeping them liquid at room temperature. Olive oil, corn oil, canola oil, and cod liver oil are examples of unsaturated fats. Unsaturated fats help to improve blood cholesterol levels, whereas saturated fats might contribute to plaque formation in the arteries, which increases the risk of a heart attack. In the food industry, oils are artificially hydrogenated to make them semi-solid, leading to less spoilage and increased shelf life. Simply speaking, hydrogen gas is bubbled through oils to solidify them. During this hydrogenation process, double bonds of the cis-conformation in the hydrocarbon chain may be converted to double bonds in the trans-conformation. This forms a trans-fat from a cis-fat. The orientation of the double bonds affects the chemical properties of the fat (Figure 2.19). FIGURE 2.19 During the hydrogenation process, the orientation around the double bonds is changed, making a trans-fat from a cis-fat. This changes the chemical properties of the molecule. Margarine, some types of peanut butter, and shortening are examples of artificially hydrogenated trans-fats. Recent studies have shown that an increase in trans-fats in the human diet may lead to an increase in levels of low-density lipoprotein (LDL), or “bad” cholesterol, which, in turn, may lead to plaque deposition in the arteries, resulting in heart disease. Many fast food restaurants have recently eliminated the use of trans-fats, and U.S. food labels are now required to list their trans-fat content. Essential fatty acids are fatty acids that are required but not synthesized by the human body. Consequently, they must be supplemented through the diet. Omega-3 fatty acids fall into this category and are one of only two known essential fatty acids for humans (the other being omega-6 fatty acids). They are a type of polyunsaturated fat and 44 2 • Chemistry of Life Access for free at openstax.org'},\n",
       " {'page_number': 58,\n",
       "  'page_char_count': 4163,\n",
       "  'page_word_count': 669,\n",
       "  'page_sentence_count_raw': 43,\n",
       "  'page_token_count': 1040.75,\n",
       "  'text': 'are called omega-3 fatty acids because the third carbon from the end of the fatty acid participates in a double bond. Salmon, trout, and tuna are good sources of omega-3 fatty acids. Omega-3 fatty acids are important in brain function and normal growth and development. They may also prevent heart disease and reduce the risk of cancer. Like carbohydrates, fats have received a lot of bad publicity. It is true that eating an excess of fried foods and other “fatty” foods leads to weight gain. However, fats do have important functions. Fats serve as long-term energy storage. They also provide insulation for the body. Therefore, “healthy” unsaturated fats in moderate amounts should be consumed on a regular basis. Phospholipids are the major constituent of the plasma membrane. Like fats, they are composed of fatty acid chains attached to a glycerol or similar backbone. Instead of three fatty acids attached, however, there are two fatty acids and the third carbon of the glycerol backbone is bound to a phosphate group. The phosphate group is modified by the addition of an alcohol. A phospholipid has both hydrophobic and hydrophilic regions. The fatty acid chains are hydrophobic and exclude themselves from water, whereas the phosphate is hydrophilic and interacts with water. Cells are surrounded by a membrane, which has a bilayer of phospholipids. The fatty acids of phospholipids face inside, away from water, whereas the phosphate group can face either the outside environment or the inside of the cell, which are both aqueous. Steroids and Waxes Unlike the phospholipids and fats discussed earlier, steroids have a ring structure. Although they do not resemble other lipids, they are grouped with them because they are also hydrophobic. All steroids have four, linked carbon rings and several of them, like cholesterol, have a short tail. Cholesterol is a steroid. Cholesterol is mainly synthesized in the liver and is the precursor of many steroid hormones, such as testosterone and estradiol. It is also the precursor of vitamins E and K. Cholesterol is the precursor of bile salts, which help in the breakdown of fats and their subsequent absorption by cells. Although cholesterol is often spoken of in negative terms, it is necessary for the proper functioning of the body. It is a key component of the plasma membranes of animal cells. Waxes are made up of a hydrocarbon chain with an alcohol (–OH) group and a fatty acid. Examples of animal waxes include beeswax and lanolin. Plants also have waxes, such as the coating on their leaves, that helps prevent them from drying out. LINK TO LEARNING For an additional perspective on lipids, watch this video about types of fat (http://openstax.org/l/lipids). Proteins Proteins are one of the most abundant organic molecules in living systems and have the most diverse range of functions of all macromolecules. Proteins may be structural, regulatory, contractile, or protective; they may serve in transport, storage, or membranes; or they may be toxins or enzymes. Each cell in a living system may contain thousands of different proteins, each with a unique function. Their structures, like their functions, vary greatly. They are all, however, polymers of amino acids, arranged in a linear sequence. The functions of proteins are very diverse because there are 20 different chemically distinct amino acids that form long chains, and the amino acids can be in any order. For example, proteins can function as enzymes or hormones. Enzymes, which are produced by living cells, are catalysts in biochemical reactions (like digestion) and are usually proteins. Each enzyme is specific for the substrate (a reactant that binds to an enzyme) upon which it acts. Enzymes can function to break molecular bonds, to rearrange bonds, or to form new bonds. An example of an enzyme is salivary amylase, which breaks down amylose, a component of starch. Hormones are chemical signaling molecules, usually proteins or steroids, secreted by an endocrine gland or group of endocrine cells that act to control or regulate specific physiological processes, including growth, development, 2.3 • Biological Molecules 45'},\n",
       " {'page_number': 59,\n",
       "  'page_char_count': 2149,\n",
       "  'page_word_count': 363,\n",
       "  'page_sentence_count_raw': 20,\n",
       "  'page_token_count': 537.25,\n",
       "  'text': 'metabolism, and reproduction. For example, insulin is a protein hormone that maintains blood glucose levels. Proteins have different shapes and molecular weights; some proteins are globular in shape whereas others are fibrous in nature. For example, hemoglobin is a globular protein, but collagen, found in our skin, is a fibrous protein. Protein shape is critical to its function. Changes in temperature, pH, and exposure to chemicals may lead to permanent changes in the shape of the protein, leading to a loss of function or denaturation (to be discussed in more detail later). All proteins are made up of different arrangements of the same 20 kinds of amino acids. Amino acids are the monomers that make up proteins. Each amino acid has the same fundamental structure, which consists of a central carbon atom bonded to an amino group (–NH2), a carboxyl group (–COOH), and a hydrogen atom. Every amino acid also has another variable atom or group of atoms bonded to the central carbon atom known as the R group. The R group is the only difference in structure between the 20 amino acids; otherwise, the amino acids are identical (Figure 2.20). FIGURE 2.20 Amino acids are made up of a central carbon bonded to an amino group (–NH2), a carboxyl group (–COOH), and a hydrogen atom. The central carbon’s fourth bond varies among the different amino acids, as seen in these examples of alanine, valine, lysine, and aspartic acid. The chemical nature of the R group determines the chemical nature of the amino acid within its protein (that is, whether it is acidic, basic, polar, or nonpolar). The sequence and number of amino acids ultimately determine a protein’s shape, size, and function. Each amino acid is attached to another amino acid by a covalent bond, known as a peptide bond, which is formed by a dehydration reaction. The carboxyl group of one amino acid and the amino group of a second amino acid combine, releasing a water molecule. The resulting bond is the peptide bond. The products formed by such a linkage are called polypeptides. While the terms polypeptide and protein are 46 2 • Chemistry of Life Access for free at openstax.org'}]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pages_and_texts = open_and_read_pdf(pdf_path=pdf_path)\n",
    "pages_and_texts = pages_and_texts[18:60]\n",
    "pages_and_texts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting pandas\n",
      "  Using cached pandas-2.2.2-cp312-cp312-win_amd64.whl.metadata (19 kB)\n",
      "Requirement already satisfied: numpy>=1.26.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas) (1.26.4)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pandas) (2.9.0.post0)\n",
      "Collecting pytz>=2020.1 (from pandas)\n",
      "  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)\n",
      "Collecting tzdata>=2022.7 (from pandas)\n",
      "  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n",
      "Using cached pandas-2.2.2-cp312-cp312-win_amd64.whl (11.5 MB)\n",
      "Using cached pytz-2024.1-py2.py3-none-any.whl (505 kB)\n",
      "Using cached tzdata-2024.1-py2.py3-none-any.whl (345 kB)\n",
      "Installing collected packages: pytz, tzdata, pandas\n",
      "Successfully installed pandas-2.2.2 pytz-2024.1 tzdata-2024.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_number</th>\n",
       "      <th>page_char_count</th>\n",
       "      <th>page_word_count</th>\n",
       "      <th>page_sentence_count_raw</th>\n",
       "      <th>page_token_count</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>1542</td>\n",
       "      <td>255</td>\n",
       "      <td>10</td>\n",
       "      <td>385.50</td>\n",
       "      <td>INTRODUCTION CHAPTER 1 Introduction to Biology...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>2234</td>\n",
       "      <td>337</td>\n",
       "      <td>16</td>\n",
       "      <td>558.50</td>\n",
       "      <td>they do not meet the criteria that biologists ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20</td>\n",
       "      <td>1988</td>\n",
       "      <td>306</td>\n",
       "      <td>16</td>\n",
       "      <td>497.00</td>\n",
       "      <td>FIGURE 1.3 The leaves of this sensitive plant ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21</td>\n",
       "      <td>2092</td>\n",
       "      <td>309</td>\n",
       "      <td>14</td>\n",
       "      <td>523.00</td>\n",
       "      <td>FIGURE 1.4 Although no two look alike, these k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>1477</td>\n",
       "      <td>234</td>\n",
       "      <td>15</td>\n",
       "      <td>369.25</td>\n",
       "      <td>FIGURE 1.6 A lot of energy is required for a C...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   page_number  page_char_count  page_word_count  page_sentence_count_raw  \\\n",
       "0           18             1542              255                       10   \n",
       "1           19             2234              337                       16   \n",
       "2           20             1988              306                       16   \n",
       "3           21             2092              309                       14   \n",
       "4           22             1477              234                       15   \n",
       "\n",
       "   page_token_count                                               text  \n",
       "0            385.50  INTRODUCTION CHAPTER 1 Introduction to Biology...  \n",
       "1            558.50  they do not meet the criteria that biologists ...  \n",
       "2            497.00  FIGURE 1.3 The leaves of this sensitive plant ...  \n",
       "3            523.00  FIGURE 1.4 Although no two look alike, these k...  \n",
       "4            369.25  FIGURE 1.6 A lot of energy is required for a C...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame(pages_and_texts)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pip in c:\\program files\\windowsapps\\pythonsoftwarefoundation.python.3.12_3.12.1008.0_x64__qbz5n2kfra8p0\\lib\\site-packages (24.0)\n",
      "Requirement already satisfied: setuptools in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (69.5.1)\n",
      "Collecting wheel\n",
      "  Using cached wheel-0.43.0-py3-none-any.whl.metadata (2.2 kB)\n",
      "Using cached wheel-0.43.0-py3-none-any.whl (65 kB)\n",
      "Installing collected packages: wheel\n",
      "Successfully installed wheel-0.43.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install -U pip setuptools wheel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting spacy\n",
      "  Downloading spacy-3.7.4-cp312-cp312-win_amd64.whl.metadata (27 kB)\n",
      "Collecting spacy-legacy<3.1.0,>=3.0.11 (from spacy)\n",
      "  Downloading spacy_legacy-3.0.12-py2.py3-none-any.whl.metadata (2.8 kB)\n",
      "Collecting spacy-loggers<2.0.0,>=1.0.0 (from spacy)\n",
      "  Downloading spacy_loggers-1.0.5-py3-none-any.whl.metadata (23 kB)\n",
      "Collecting murmurhash<1.1.0,>=0.28.0 (from spacy)\n",
      "  Downloading murmurhash-1.0.10-cp312-cp312-win_amd64.whl.metadata (2.0 kB)\n",
      "Collecting cymem<2.1.0,>=2.0.2 (from spacy)\n",
      "  Downloading cymem-2.0.8-cp312-cp312-win_amd64.whl.metadata (8.6 kB)\n",
      "Collecting preshed<3.1.0,>=3.0.2 (from spacy)\n",
      "  Downloading preshed-3.0.9-cp312-cp312-win_amd64.whl.metadata (2.2 kB)\n",
      "Collecting thinc<8.3.0,>=8.2.2 (from spacy)\n",
      "  Downloading thinc-8.2.3-cp312-cp312-win_amd64.whl.metadata (15 kB)\n",
      "Collecting wasabi<1.2.0,>=0.9.1 (from spacy)\n",
      "  Downloading wasabi-1.1.2-py3-none-any.whl.metadata (28 kB)\n",
      "Collecting srsly<3.0.0,>=2.4.3 (from spacy)\n",
      "  Downloading srsly-2.4.8-cp312-cp312-win_amd64.whl.metadata (20 kB)\n",
      "Collecting catalogue<2.1.0,>=2.0.6 (from spacy)\n",
      "  Downloading catalogue-2.0.10-py3-none-any.whl.metadata (14 kB)\n",
      "Collecting weasel<0.4.0,>=0.1.0 (from spacy)\n",
      "  Downloading weasel-0.3.4-py3-none-any.whl.metadata (4.7 kB)\n",
      "Collecting typer<0.10.0,>=0.3.0 (from spacy)\n",
      "  Downloading typer-0.9.4-py3-none-any.whl.metadata (14 kB)\n",
      "Collecting smart-open<7.0.0,>=5.2.1 (from spacy)\n",
      "  Downloading smart_open-6.4.0-py3-none-any.whl.metadata (21 kB)\n",
      "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (4.66.4)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.13.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (2.31.0)\n",
      "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (2.7.1)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (3.1.3)\n",
      "Requirement already satisfied: setuptools in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (69.5.1)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (23.2)\n",
      "Collecting langcodes<4.0.0,>=3.2.0 (from spacy)\n",
      "  Downloading langcodes-3.4.0-py3-none-any.whl.metadata (29 kB)\n",
      "Requirement already satisfied: numpy>=1.19.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from spacy) (1.26.4)\n",
      "Collecting language-data>=1.2 (from langcodes<4.0.0,>=3.2.0->spacy)\n",
      "  Downloading language_data-1.2.0-py3-none-any.whl.metadata (4.3 kB)\n",
      "Requirement already satisfied: annotated-types>=0.4.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (0.6.0)\n",
      "Requirement already satisfied: pydantic-core==2.18.2 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (2.18.2)\n",
      "Requirement already satisfied: typing-extensions>=4.6.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from pydantic!=1.8,!=1.8.1,<3.0.0,>=1.7.4->spacy) (4.11.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (2.2.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests<3.0.0,>=2.13.0->spacy) (2024.2.2)\n",
      "Collecting blis<0.8.0,>=0.7.8 (from thinc<8.3.0,>=8.2.2->spacy)\n",
      "  Downloading blis-0.7.11-cp312-cp312-win_amd64.whl.metadata (7.6 kB)\n",
      "Collecting confection<1.0.0,>=0.0.1 (from thinc<8.3.0,>=8.2.2->spacy)\n",
      "  Downloading confection-0.1.4-py3-none-any.whl.metadata (19 kB)\n",
      "Requirement already satisfied: colorama in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tqdm<5.0.0,>=4.38.0->spacy) (0.4.6)\n",
      "Requirement already satisfied: click<9.0.0,>=7.1.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from typer<0.10.0,>=0.3.0->spacy) (8.1.7)\n",
      "Collecting cloudpathlib<0.17.0,>=0.7.0 (from weasel<0.4.0,>=0.1.0->spacy)\n",
      "  Downloading cloudpathlib-0.16.0-py3-none-any.whl.metadata (14 kB)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from jinja2->spacy) (2.1.5)\n",
      "Collecting marisa-trie>=0.7.7 (from language-data>=1.2->langcodes<4.0.0,>=3.2.0->spacy)\n",
      "  Downloading marisa_trie-1.1.1-cp312-cp312-win_amd64.whl.metadata (8.9 kB)\n",
      "Downloading spacy-3.7.4-cp312-cp312-win_amd64.whl (11.7 MB)\n",
      "   ---------------------------------------- 0.0/11.7 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.1/11.7 MB 3.5 MB/s eta 0:00:04\n",
      "    --------------------------------------- 0.2/11.7 MB 2.6 MB/s eta 0:00:05\n",
      "   - -------------------------------------- 0.3/11.7 MB 2.7 MB/s eta 0:00:05\n",
      "   - -------------------------------------- 0.5/11.7 MB 3.0 MB/s eta 0:00:04\n",
      "   -- ------------------------------------- 0.7/11.7 MB 3.2 MB/s eta 0:00:04\n",
      "   --- ------------------------------------ 0.9/11.7 MB 3.4 MB/s eta 0:00:04\n",
      "   ---- ----------------------------------- 1.2/11.7 MB 3.7 MB/s eta 0:00:03\n",
      "   ---- ----------------------------------- 1.4/11.7 MB 4.0 MB/s eta 0:00:03\n",
      "   ------ --------------------------------- 1.8/11.7 MB 4.3 MB/s eta 0:00:03\n",
      "   ------- -------------------------------- 2.1/11.7 MB 4.7 MB/s eta 0:00:03\n",
      "   -------- ------------------------------- 2.3/11.7 MB 4.7 MB/s eta 0:00:02\n",
      "   -------- ------------------------------- 2.5/11.7 MB 4.7 MB/s eta 0:00:02\n",
      "   -------- ------------------------------- 2.6/11.7 MB 4.4 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 2.7/11.7 MB 4.2 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 2.7/11.7 MB 4.0 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 2.8/11.7 MB 3.9 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 2.9/11.7 MB 3.8 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 2.9/11.7 MB 3.7 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 3.0/11.7 MB 3.5 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 3.1/11.7 MB 3.4 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 3.1/11.7 MB 3.3 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 3.2/11.7 MB 3.2 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 3.2/11.7 MB 3.1 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 3.3/11.7 MB 3.1 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 3.4/11.7 MB 3.0 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 3.4/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 3.5/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 3.6/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 3.8/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 3.9/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 3.9/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 4.0/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 4.1/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 4.2/11.7 MB 2.7 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 4.3/11.7 MB 2.7 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 4.4/11.7 MB 2.7 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 4.6/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 4.8/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 4.9/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   ----------------- ---------------------- 5.1/11.7 MB 2.8 MB/s eta 0:00:03\n",
      "   ------------------ --------------------- 5.3/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------------- -------------------- 5.6/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   ------------------- -------------------- 5.7/11.7 MB 2.9 MB/s eta 0:00:03\n",
      "   -------------------- ------------------- 5.9/11.7 MB 3.0 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 6.1/11.7 MB 3.0 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 6.4/11.7 MB 3.0 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 6.7/11.7 MB 3.1 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 7.0/11.7 MB 3.2 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 7.2/11.7 MB 3.2 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 7.4/11.7 MB 3.3 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 7.7/11.7 MB 3.3 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 7.9/11.7 MB 3.3 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 7.9/11.7 MB 3.3 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 8.0/11.7 MB 3.3 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 8.0/11.7 MB 3.2 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 8.1/11.7 MB 3.2 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 8.2/11.7 MB 3.2 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 8.3/11.7 MB 3.1 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 8.4/11.7 MB 3.1 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 8.6/11.7 MB 3.1 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 8.7/11.7 MB 3.1 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 8.8/11.7 MB 3.1 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 8.8/11.7 MB 3.1 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 8.9/11.7 MB 3.1 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 9.0/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 9.1/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 9.2/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 9.3/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 9.4/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 9.6/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 9.7/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 9.8/11.7 MB 3.0 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 9.8/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 9.9/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 10.0/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 10.1/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 10.1/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 10.3/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 10.4/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 10.4/11.7 MB 2.9 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 10.5/11.7 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 10.5/11.7 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 10.6/11.7 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 10.6/11.7 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 10.6/11.7 MB 2.8 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 10.7/11.7 MB 2.7 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 10.8/11.7 MB 2.7 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 10.9/11.7 MB 2.7 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 11.0/11.7 MB 2.7 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 11.1/11.7 MB 2.7 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 11.2/11.7 MB 2.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 11.2/11.7 MB 2.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 11.3/11.7 MB 2.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 11.4/11.7 MB 2.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  11.5/11.7 MB 2.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  11.5/11.7 MB 2.5 MB/s eta 0:00:01\n",
      "   ---------------------------------------  11.6/11.7 MB 2.5 MB/s eta 0:00:01\n",
      "   ---------------------------------------  11.7/11.7 MB 2.5 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 11.7/11.7 MB 2.5 MB/s eta 0:00:00\n",
      "Downloading catalogue-2.0.10-py3-none-any.whl (17 kB)\n",
      "Downloading cymem-2.0.8-cp312-cp312-win_amd64.whl (39 kB)\n",
      "Downloading langcodes-3.4.0-py3-none-any.whl (182 kB)\n",
      "   ---------------------------------------- 0.0/182.0 kB ? eta -:--:--\n",
      "   ------ --------------------------------- 30.7/182.0 kB 1.4 MB/s eta 0:00:01\n",
      "   -------- ------------------------------ 41.0/182.0 kB 991.0 kB/s eta 0:00:01\n",
      "   -------------------- ------------------- 92.2/182.0 kB 1.1 MB/s eta 0:00:01\n",
      "   ------------------------- ------------ 122.9/182.0 kB 901.1 kB/s eta 0:00:01\n",
      "   -------------------------------------- 182.0/182.0 kB 998.8 kB/s eta 0:00:00\n",
      "Downloading murmurhash-1.0.10-cp312-cp312-win_amd64.whl (25 kB)\n",
      "Downloading preshed-3.0.9-cp312-cp312-win_amd64.whl (122 kB)\n",
      "   ---------------------------------------- 0.0/122.4 kB ? eta -:--:--\n",
      "   ------------------------------ --------- 92.2/122.4 kB 2.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 122.4/122.4 kB 1.8 MB/s eta 0:00:00\n",
      "Downloading smart_open-6.4.0-py3-none-any.whl (57 kB)\n",
      "   ---------------------------------------- 0.0/57.0 kB ? eta -:--:--\n",
      "   ---------------------------------------- 57.0/57.0 kB 1.5 MB/s eta 0:00:00\n",
      "Downloading spacy_legacy-3.0.12-py2.py3-none-any.whl (29 kB)\n",
      "Downloading spacy_loggers-1.0.5-py3-none-any.whl (22 kB)\n",
      "Downloading srsly-2.4.8-cp312-cp312-win_amd64.whl (478 kB)\n",
      "   ---------------------------------------- 0.0/478.8 kB ? eta -:--:--\n",
      "   ------ --------------------------------- 81.9/478.8 kB 2.3 MB/s eta 0:00:01\n",
      "   -------------- ------------------------- 174.1/478.8 kB 2.1 MB/s eta 0:00:01\n",
      "   --------------------- ------------------ 256.0/478.8 kB 2.0 MB/s eta 0:00:01\n",
      "   --------------------------- ------------ 327.7/478.8 kB 1.8 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 450.6/478.8 kB 2.0 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 478.8/478.8 kB 1.9 MB/s eta 0:00:00\n",
      "Downloading thinc-8.2.3-cp312-cp312-win_amd64.whl (1.4 MB)\n",
      "   ---------------------------------------- 0.0/1.4 MB ? eta -:--:--\n",
      "   - -------------------------------------- 0.1/1.4 MB 1.6 MB/s eta 0:00:01\n",
      "   --- ------------------------------------ 0.1/1.4 MB 1.7 MB/s eta 0:00:01\n",
      "   ----- ---------------------------------- 0.2/1.4 MB 1.7 MB/s eta 0:00:01\n",
      "   ------- -------------------------------- 0.3/1.4 MB 1.8 MB/s eta 0:00:01\n",
      "   ---------- ----------------------------- 0.4/1.4 MB 1.8 MB/s eta 0:00:01\n",
      "   ------------ --------------------------- 0.5/1.4 MB 1.8 MB/s eta 0:00:01\n",
      "   ---------------- ----------------------- 0.6/1.4 MB 1.9 MB/s eta 0:00:01\n",
      "   ------------------- -------------------- 0.7/1.4 MB 2.0 MB/s eta 0:00:01\n",
      "   ---------------------- ----------------- 0.8/1.4 MB 2.1 MB/s eta 0:00:01\n",
      "   ------------------------- -------------- 0.9/1.4 MB 2.1 MB/s eta 0:00:01\n",
      "   --------------------------- ------------ 1.0/1.4 MB 2.0 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 1.2/1.4 MB 2.2 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 1.3/1.4 MB 2.3 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 1.4/1.4 MB 2.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------  1.4/1.4 MB 2.2 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 1.4/1.4 MB 2.1 MB/s eta 0:00:00\n",
      "Downloading typer-0.9.4-py3-none-any.whl (45 kB)\n",
      "   ---------------------------------------- 0.0/46.0 kB ? eta -:--:--\n",
      "   -------------------------- ------------- 30.7/46.0 kB 1.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 46.0/46.0 kB 760.3 kB/s eta 0:00:00\n",
      "Downloading wasabi-1.1.2-py3-none-any.whl (27 kB)\n",
      "Downloading weasel-0.3.4-py3-none-any.whl (50 kB)\n",
      "   ---------------------------------------- 0.0/50.1 kB ? eta -:--:--\n",
      "   -------------------------------- ------- 41.0/50.1 kB 991.0 kB/s eta 0:00:01\n",
      "   ---------------------------------------- 50.1/50.1 kB 848.3 kB/s eta 0:00:00\n",
      "Downloading blis-0.7.11-cp312-cp312-win_amd64.whl (6.6 MB)\n",
      "   ---------------------------------------- 0.0/6.6 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.1/6.6 MB 1.7 MB/s eta 0:00:04\n",
      "    --------------------------------------- 0.1/6.6 MB 1.3 MB/s eta 0:00:06\n",
      "    --------------------------------------- 0.1/6.6 MB 983.0 kB/s eta 0:00:07\n",
      "   - -------------------------------------- 0.2/6.6 MB 1.1 MB/s eta 0:00:06\n",
      "   - -------------------------------------- 0.3/6.6 MB 1.2 MB/s eta 0:00:06\n",
      "   - -------------------------------------- 0.3/6.6 MB 1.2 MB/s eta 0:00:06\n",
      "   -- ------------------------------------- 0.4/6.6 MB 1.2 MB/s eta 0:00:06\n",
      "   -- ------------------------------------- 0.5/6.6 MB 1.3 MB/s eta 0:00:05\n",
      "   --- ------------------------------------ 0.5/6.6 MB 1.3 MB/s eta 0:00:05\n",
      "   --- ------------------------------------ 0.6/6.6 MB 1.3 MB/s eta 0:00:05\n",
      "   ---- ----------------------------------- 0.7/6.6 MB 1.3 MB/s eta 0:00:05\n",
      "   ---- ----------------------------------- 0.8/6.6 MB 1.4 MB/s eta 0:00:05\n",
      "   ----- ---------------------------------- 0.9/6.6 MB 1.4 MB/s eta 0:00:04\n",
      "   ----- ---------------------------------- 1.0/6.6 MB 1.5 MB/s eta 0:00:04\n",
      "   ------ --------------------------------- 1.0/6.6 MB 1.5 MB/s eta 0:00:04\n",
      "   ------ --------------------------------- 1.1/6.6 MB 1.5 MB/s eta 0:00:04\n",
      "   ------- -------------------------------- 1.2/6.6 MB 1.5 MB/s eta 0:00:04\n",
      "   ------- -------------------------------- 1.2/6.6 MB 1.5 MB/s eta 0:00:04\n",
      "   -------- ------------------------------- 1.4/6.6 MB 1.6 MB/s eta 0:00:04\n",
      "   --------- ------------------------------ 1.5/6.6 MB 1.7 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 1.7/6.6 MB 1.7 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 1.8/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 1.9/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 2.0/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 2.1/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 2.2/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 2.3/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 2.4/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 2.4/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 2.6/6.6 MB 1.9 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 2.6/6.6 MB 1.9 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 2.7/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 2.8/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ----------------- ---------------------- 2.8/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ----------------- ---------------------- 2.9/6.6 MB 1.8 MB/s eta 0:00:03\n",
      "   ------------------ --------------------- 3.0/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ------------------ --------------------- 3.1/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ------------------- -------------------- 3.2/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ------------------- -------------------- 3.2/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 3.4/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 3.5/6.6 MB 1.9 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 3.5/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 3.6/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 3.6/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 3.7/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 3.8/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 3.8/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 3.9/6.6 MB 1.8 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 3.9/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 4.0/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 4.0/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 4.1/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 4.1/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 4.1/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 4.2/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 4.3/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 4.3/6.6 MB 1.7 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 4.4/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 4.4/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 4.5/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 4.6/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 4.6/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 4.7/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 4.8/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 4.8/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 4.9/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 4.9/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ------------------------------ --------- 5.0/6.6 MB 1.6 MB/s eta 0:00:02\n",
      "   ------------------------------ --------- 5.1/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 5.2/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 5.2/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 5.3/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 5.3/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 5.4/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 5.4/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 5.5/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 5.6/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 5.6/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 5.7/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 5.7/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 5.8/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 5.9/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 5.9/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 6.0/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 6.1/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 6.2/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 6.2/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 6.3/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 6.4/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  6.5/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  6.6/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  6.6/6.6 MB 1.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 6.6/6.6 MB 1.6 MB/s eta 0:00:00\n",
      "Downloading cloudpathlib-0.16.0-py3-none-any.whl (45 kB)\n",
      "   ---------------------------------------- 0.0/45.0 kB ? eta -:--:--\n",
      "   ---------------------------------------- 45.0/45.0 kB 2.3 MB/s eta 0:00:00\n",
      "Downloading confection-0.1.4-py3-none-any.whl (35 kB)\n",
      "Downloading language_data-1.2.0-py3-none-any.whl (5.4 MB)\n",
      "   ---------------------------------------- 0.0/5.4 MB ? eta -:--:--\n",
      "    --------------------------------------- 0.1/5.4 MB 2.2 MB/s eta 0:00:03\n",
      "   - -------------------------------------- 0.2/5.4 MB 2.0 MB/s eta 0:00:03\n",
      "   - -------------------------------------- 0.2/5.4 MB 1.9 MB/s eta 0:00:03\n",
      "   -- ------------------------------------- 0.3/5.4 MB 1.7 MB/s eta 0:00:03\n",
      "   --- ------------------------------------ 0.4/5.4 MB 1.7 MB/s eta 0:00:03\n",
      "   --- ------------------------------------ 0.5/5.4 MB 1.8 MB/s eta 0:00:03\n",
      "   ---- ----------------------------------- 0.6/5.4 MB 1.8 MB/s eta 0:00:03\n",
      "   ----- ---------------------------------- 0.7/5.4 MB 1.9 MB/s eta 0:00:03\n",
      "   ----- ---------------------------------- 0.7/5.4 MB 1.8 MB/s eta 0:00:03\n",
      "   ----- ---------------------------------- 0.8/5.4 MB 1.8 MB/s eta 0:00:03\n",
      "   ------ --------------------------------- 0.9/5.4 MB 1.7 MB/s eta 0:00:03\n",
      "   ------ --------------------------------- 0.9/5.4 MB 1.7 MB/s eta 0:00:03\n",
      "   ------- -------------------------------- 1.0/5.4 MB 1.7 MB/s eta 0:00:03\n",
      "   ------- -------------------------------- 1.1/5.4 MB 1.7 MB/s eta 0:00:03\n",
      "   -------- ------------------------------- 1.1/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   -------- ------------------------------- 1.2/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 1.2/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   --------- ------------------------------ 1.3/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 1.4/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ---------- ----------------------------- 1.5/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 1.5/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ----------- ---------------------------- 1.6/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ------------ --------------------------- 1.6/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 1.8/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   ------------- -------------------------- 1.9/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 1.9/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   -------------- ------------------------- 2.0/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 2.0/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 2.1/5.4 MB 1.6 MB/s eta 0:00:03\n",
      "   --------------- ------------------------ 2.1/5.4 MB 1.5 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 2.2/5.4 MB 1.5 MB/s eta 0:00:03\n",
      "   ---------------- ----------------------- 2.2/5.4 MB 1.5 MB/s eta 0:00:03\n",
      "   ----------------- ---------------------- 2.3/5.4 MB 1.5 MB/s eta 0:00:03\n",
      "   ----------------- ---------------------- 2.4/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ------------------ --------------------- 2.5/5.4 MB 1.6 MB/s eta 0:00:02\n",
      "   ------------------- -------------------- 2.6/5.4 MB 1.6 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 2.7/5.4 MB 1.6 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 2.8/5.4 MB 1.6 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 2.8/5.4 MB 1.6 MB/s eta 0:00:02\n",
      "   -------------------- ------------------- 2.8/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 2.9/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   --------------------- ------------------ 2.9/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 3.0/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 3.1/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ---------------------- ----------------- 3.1/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 3.1/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 3.2/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ----------------------- ---------------- 3.2/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 3.3/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 3.3/5.4 MB 1.5 MB/s eta 0:00:02\n",
      "   ------------------------ --------------- 3.3/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 3.4/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 3.4/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   ------------------------- -------------- 3.5/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 3.5/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 3.6/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 3.6/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   -------------------------- ------------- 3.6/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 3.7/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   --------------------------- ------------ 3.7/5.4 MB 1.3 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 3.8/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 3.8/5.4 MB 1.4 MB/s eta 0:00:02\n",
      "   ---------------------------- ----------- 3.9/5.4 MB 1.3 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 4.0/5.4 MB 1.3 MB/s eta 0:00:02\n",
      "   ----------------------------- ---------- 4.0/5.4 MB 1.3 MB/s eta 0:00:02\n",
      "   ------------------------------ --------- 4.1/5.4 MB 1.3 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 4.1/5.4 MB 1.3 MB/s eta 0:00:01\n",
      "   ------------------------------ --------- 4.2/5.4 MB 1.3 MB/s eta 0:00:01\n",
      "   ------------------------------- -------- 4.2/5.4 MB 1.3 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 4.3/5.4 MB 1.3 MB/s eta 0:00:01\n",
      "   -------------------------------- ------- 4.4/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 4.5/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   --------------------------------- ------ 4.6/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 4.6/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ---------------------------------- ----- 4.7/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 4.7/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 4.8/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ----------------------------------- ---- 4.8/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 4.9/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 5.0/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ------------------------------------- -- 5.1/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 5.2/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 5.2/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------  5.3/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------  5.4/5.4 MB 1.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 5.4/5.4 MB 1.4 MB/s eta 0:00:00\n",
      "Downloading marisa_trie-1.1.1-cp312-cp312-win_amd64.whl (151 kB)\n",
      "   ---------------------------------------- 0.0/151.3 kB ? eta -:--:--\n",
      "   ------------------------ --------------- 92.2/151.3 kB 2.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 151.3/151.3 kB 1.8 MB/s eta 0:00:00\n",
      "Installing collected packages: cymem, wasabi, spacy-loggers, spacy-legacy, smart-open, murmurhash, marisa-trie, cloudpathlib, catalogue, blis, typer, srsly, preshed, language-data, langcodes, confection, weasel, thinc, spacy\n",
      "  Attempting uninstall: typer\n",
      "    Found existing installation: typer 0.12.3\n",
      "    Uninstalling typer-0.12.3:\n",
      "      Successfully uninstalled typer-0.12.3\n",
      "Successfully installed blis-0.7.11 catalogue-2.0.10 cloudpathlib-0.16.0 confection-0.1.4 cymem-2.0.8 langcodes-3.4.0 language-data-1.2.0 marisa-trie-1.1.1 murmurhash-1.0.10 preshed-3.0.9 smart-open-6.4.0 spacy-3.7.4 spacy-legacy-3.0.12 spacy-loggers-1.0.5 srsly-2.4.8 thinc-8.2.3 typer-0.9.4 wasabi-1.1.2 weasel-0.3.4\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "fastapi-cli 0.0.3 requires typer>=0.12.3, but you have typer 0.9.4 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "pip install -U spacy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacy.lang.en import English"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[This is a sentence., This another sentence.]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nlp = English()\n",
    "\n",
    "# Add a sentencizer pipeline, see https://spacy.io/api/sentencizer/ \n",
    "nlp.add_pipe(\"sentencizer\")\n",
    "\n",
    "# Create a document instance as an example\n",
    "doc = nlp(\"This is a sentence. This another sentence.\")\n",
    "assert len(list(doc.sents)) == 2\n",
    "\n",
    "# Access the sentences of the document\n",
    "list(doc.sents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 42/42 [00:00<00:00, 158.64it/s]\n"
     ]
    }
   ],
   "source": [
    "for item in tqdm(pages_and_texts):\n",
    "    item[\"sentences\"] = list(nlp(item[\"text\"]).sents)\n",
    "    \n",
    "    # Make sure all sentences are strings\n",
    "    item[\"sentences\"] = [str(sentence) for sentence in item[\"sentences\"]]\n",
    "    \n",
    "    # Count the sentences \n",
    "    item[\"page_sentence_count_spacy\"] = len(item[\"sentences\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'page_number': 43,\n",
       "  'page_char_count': 1809,\n",
       "  'page_word_count': 307,\n",
       "  'page_sentence_count_raw': 15,\n",
       "  'page_token_count': 452.25,\n",
       "  'text': 'VISUAL CONNECTION FIGURE 2.3 Arranged in columns and rows based on the characteristics of the elements, the periodic table provides key information about the elements and how they might interact with each other to form molecules. Most periodic tables provide a key or legend to the information they contain. How many neutrons do (K) potassium-39 and potassium-40 have, respectively? EVOLUTION CONNECTION Carbon Dating Carbon-14 (14C) is a naturally occurring radioisotope that is created in the atmosphere by cosmic rays. This is a continuous process, so more 14C is always being created. As a living organism develops, the relative level of 14C in its body is equal to the concentration of 14C in the atmosphere. When an organism dies, it is no longer ingesting 14C, so the ratio will decline. 14C decays to 14N by a process called beta decay; it gives off energy in this slow process. After approximately 5,730 years, only one-half of the starting concentration of 14C will have been converted to 14N. The time it takes for half of the original concentration of an isotope to decay to its more stable form is called its half- life. Because the half-life of 14C is long, it is used to age formerly living objects, such as fossils. Using the ratio of the 14C concentration found in an object to the amount of 14C detected in the atmosphere, the amount of the isotope that has not yet decayed can be determined. Based on this amount, the age of the fossil can be calculated to about 50,000 years (Figure 2.4). Isotopes with longer half-lives, such as potassium-40, are used to calculate the ages of older fossils. Through the use of carbon dating, scientists can reconstruct the ecology and biogeography of organisms living within the past 50,000 years. 30 2 • Chemistry of Life Access for free at openstax.org',\n",
       "  'sentences': ['VISUAL CONNECTION FIGURE 2.3 Arranged in columns and rows based on the characteristics of the elements, the periodic table provides key information about the elements and how they might interact with each other to form molecules.',\n",
       "   'Most periodic tables provide a key or legend to the information they contain.',\n",
       "   'How many neutrons do (K) potassium-39 and potassium-40 have, respectively?',\n",
       "   'EVOLUTION CONNECTION Carbon Dating Carbon-14 (14C) is a naturally occurring radioisotope that is created in the atmosphere by cosmic rays.',\n",
       "   'This is a continuous process, so more 14C is always being created.',\n",
       "   'As a living organism develops, the relative level of 14C in its body is equal to the concentration of 14C in the atmosphere.',\n",
       "   'When an organism dies, it is no longer ingesting 14C, so the ratio will decline.',\n",
       "   '14C decays to 14N by a process called beta decay; it gives off energy in this slow process.',\n",
       "   'After approximately 5,730 years, only one-half of the starting concentration of 14C will have been converted to 14N. The time it takes for half of the original concentration of an isotope to decay to its more stable form is called its half- life.',\n",
       "   'Because the half-life of 14C is long, it is used to age formerly living objects, such as fossils.',\n",
       "   'Using the ratio of the 14C concentration found in an object to the amount of 14C detected in the atmosphere, the amount of the isotope that has not yet decayed can be determined.',\n",
       "   'Based on this amount, the age of the fossil can be calculated to about 50,000 years (Figure 2.4).',\n",
       "   'Isotopes with longer half-lives, such as potassium-40, are used to calculate the ages of older fossils.',\n",
       "   'Through the use of carbon dating, scientists can reconstruct the ecology and biogeography of organisms living within the past 50,000 years.',\n",
       "   '30 2 • Chemistry of Life Access for free at openstax.org'],\n",
       "  'page_sentence_count_spacy': 15}]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "# Inspect an example\n",
    "random.sample(pages_and_texts, k=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_number</th>\n",
       "      <th>page_char_count</th>\n",
       "      <th>page_word_count</th>\n",
       "      <th>page_sentence_count_raw</th>\n",
       "      <th>page_token_count</th>\n",
       "      <th>page_sentence_count_spacy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>38.50</td>\n",
       "      <td>2626.60</td>\n",
       "      <td>418.98</td>\n",
       "      <td>22.90</td>\n",
       "      <td>656.65</td>\n",
       "      <td>22.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>12.27</td>\n",
       "      <td>945.61</td>\n",
       "      <td>154.65</td>\n",
       "      <td>10.96</td>\n",
       "      <td>236.40</td>\n",
       "      <td>9.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>18.00</td>\n",
       "      <td>284.00</td>\n",
       "      <td>48.00</td>\n",
       "      <td>4.00</td>\n",
       "      <td>71.00</td>\n",
       "      <td>4.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>28.25</td>\n",
       "      <td>1983.50</td>\n",
       "      <td>306.25</td>\n",
       "      <td>15.25</td>\n",
       "      <td>495.88</td>\n",
       "      <td>15.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>38.50</td>\n",
       "      <td>2813.00</td>\n",
       "      <td>454.50</td>\n",
       "      <td>22.00</td>\n",
       "      <td>703.25</td>\n",
       "      <td>22.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>48.75</td>\n",
       "      <td>3331.50</td>\n",
       "      <td>545.50</td>\n",
       "      <td>27.75</td>\n",
       "      <td>832.88</td>\n",
       "      <td>28.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>59.00</td>\n",
       "      <td>4190.00</td>\n",
       "      <td>706.00</td>\n",
       "      <td>61.00</td>\n",
       "      <td>1047.50</td>\n",
       "      <td>47.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       page_number  page_char_count  page_word_count  page_sentence_count_raw  \\\n",
       "count        42.00            42.00            42.00                    42.00   \n",
       "mean         38.50          2626.60           418.98                    22.90   \n",
       "std          12.27           945.61           154.65                    10.96   \n",
       "min          18.00           284.00            48.00                     4.00   \n",
       "25%          28.25          1983.50           306.25                    15.25   \n",
       "50%          38.50          2813.00           454.50                    22.00   \n",
       "75%          48.75          3331.50           545.50                    27.75   \n",
       "max          59.00          4190.00           706.00                    61.00   \n",
       "\n",
       "       page_token_count  page_sentence_count_spacy  \n",
       "count             42.00                      42.00  \n",
       "mean             656.65                      22.98  \n",
       "std              236.40                       9.70  \n",
       "min               71.00                       4.00  \n",
       "25%              495.88                      15.25  \n",
       "50%              703.25                      22.50  \n",
       "75%              832.88                      28.00  \n",
       "max             1047.50                      47.00  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(pages_and_texts)\n",
    "df.describe().round(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chunking our sentences together\n",
    "\n",
    "\n",
    "Don't overload the embedding models capacity for tokens (e.g. if an embedding model has a capacity of 384 tokens, there could be information loss if you try to embed a sequence of 400+ tokens).\n",
    "Our Average Token is 22 so if we group then we can have 220 which is less then 384"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 42/42 [00:00<?, ?it/s]\n"
     ]
    }
   ],
   "source": [
    "# Define split size to turn groups of sentences into chunks\n",
    "num_sentence_chunk_size = 10 \n",
    "\n",
    "# Create a function that recursively splits a list into desired sizes\n",
    "def split_list(input_list: list, \n",
    "               slice_size: int) -> list[list[str]]:\n",
    "    \"\"\"\n",
    "    Splits the input_list into sublists of size slice_size (or as close as possible).\n",
    "\n",
    "    For example, a list of 17 sentences would be split into two lists of [[10], [7]]\n",
    "    \"\"\"\n",
    "    return [input_list[i:i + slice_size] for i in range(0, len(input_list), slice_size)]\n",
    "\n",
    "# Loop through pages and texts and split sentences into chunks\n",
    "for item in tqdm(pages_and_texts):\n",
    "    item[\"sentence_chunks\"] = split_list(input_list=item[\"sentences\"],\n",
    "                                         slice_size=num_sentence_chunk_size)\n",
    "    item[\"num_chunks\"] = len(item[\"sentence_chunks\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'page_number': 43,\n",
       "  'page_char_count': 1809,\n",
       "  'page_word_count': 307,\n",
       "  'page_sentence_count_raw': 15,\n",
       "  'page_token_count': 452.25,\n",
       "  'text': 'VISUAL CONNECTION FIGURE 2.3 Arranged in columns and rows based on the characteristics of the elements, the periodic table provides key information about the elements and how they might interact with each other to form molecules. Most periodic tables provide a key or legend to the information they contain. How many neutrons do (K) potassium-39 and potassium-40 have, respectively? EVOLUTION CONNECTION Carbon Dating Carbon-14 (14C) is a naturally occurring radioisotope that is created in the atmosphere by cosmic rays. This is a continuous process, so more 14C is always being created. As a living organism develops, the relative level of 14C in its body is equal to the concentration of 14C in the atmosphere. When an organism dies, it is no longer ingesting 14C, so the ratio will decline. 14C decays to 14N by a process called beta decay; it gives off energy in this slow process. After approximately 5,730 years, only one-half of the starting concentration of 14C will have been converted to 14N. The time it takes for half of the original concentration of an isotope to decay to its more stable form is called its half- life. Because the half-life of 14C is long, it is used to age formerly living objects, such as fossils. Using the ratio of the 14C concentration found in an object to the amount of 14C detected in the atmosphere, the amount of the isotope that has not yet decayed can be determined. Based on this amount, the age of the fossil can be calculated to about 50,000 years (Figure 2.4). Isotopes with longer half-lives, such as potassium-40, are used to calculate the ages of older fossils. Through the use of carbon dating, scientists can reconstruct the ecology and biogeography of organisms living within the past 50,000 years. 30 2 • Chemistry of Life Access for free at openstax.org',\n",
       "  'sentences': ['VISUAL CONNECTION FIGURE 2.3 Arranged in columns and rows based on the characteristics of the elements, the periodic table provides key information about the elements and how they might interact with each other to form molecules.',\n",
       "   'Most periodic tables provide a key or legend to the information they contain.',\n",
       "   'How many neutrons do (K) potassium-39 and potassium-40 have, respectively?',\n",
       "   'EVOLUTION CONNECTION Carbon Dating Carbon-14 (14C) is a naturally occurring radioisotope that is created in the atmosphere by cosmic rays.',\n",
       "   'This is a continuous process, so more 14C is always being created.',\n",
       "   'As a living organism develops, the relative level of 14C in its body is equal to the concentration of 14C in the atmosphere.',\n",
       "   'When an organism dies, it is no longer ingesting 14C, so the ratio will decline.',\n",
       "   '14C decays to 14N by a process called beta decay; it gives off energy in this slow process.',\n",
       "   'After approximately 5,730 years, only one-half of the starting concentration of 14C will have been converted to 14N. The time it takes for half of the original concentration of an isotope to decay to its more stable form is called its half- life.',\n",
       "   'Because the half-life of 14C is long, it is used to age formerly living objects, such as fossils.',\n",
       "   'Using the ratio of the 14C concentration found in an object to the amount of 14C detected in the atmosphere, the amount of the isotope that has not yet decayed can be determined.',\n",
       "   'Based on this amount, the age of the fossil can be calculated to about 50,000 years (Figure 2.4).',\n",
       "   'Isotopes with longer half-lives, such as potassium-40, are used to calculate the ages of older fossils.',\n",
       "   'Through the use of carbon dating, scientists can reconstruct the ecology and biogeography of organisms living within the past 50,000 years.',\n",
       "   '30 2 • Chemistry of Life Access for free at openstax.org'],\n",
       "  'page_sentence_count_spacy': 15,\n",
       "  'sentence_chunks': [['VISUAL CONNECTION FIGURE 2.3 Arranged in columns and rows based on the characteristics of the elements, the periodic table provides key information about the elements and how they might interact with each other to form molecules.',\n",
       "    'Most periodic tables provide a key or legend to the information they contain.',\n",
       "    'How many neutrons do (K) potassium-39 and potassium-40 have, respectively?',\n",
       "    'EVOLUTION CONNECTION Carbon Dating Carbon-14 (14C) is a naturally occurring radioisotope that is created in the atmosphere by cosmic rays.',\n",
       "    'This is a continuous process, so more 14C is always being created.',\n",
       "    'As a living organism develops, the relative level of 14C in its body is equal to the concentration of 14C in the atmosphere.',\n",
       "    'When an organism dies, it is no longer ingesting 14C, so the ratio will decline.',\n",
       "    '14C decays to 14N by a process called beta decay; it gives off energy in this slow process.',\n",
       "    'After approximately 5,730 years, only one-half of the starting concentration of 14C will have been converted to 14N. The time it takes for half of the original concentration of an isotope to decay to its more stable form is called its half- life.',\n",
       "    'Because the half-life of 14C is long, it is used to age formerly living objects, such as fossils.'],\n",
       "   ['Using the ratio of the 14C concentration found in an object to the amount of 14C detected in the atmosphere, the amount of the isotope that has not yet decayed can be determined.',\n",
       "    'Based on this amount, the age of the fossil can be calculated to about 50,000 years (Figure 2.4).',\n",
       "    'Isotopes with longer half-lives, such as potassium-40, are used to calculate the ages of older fossils.',\n",
       "    'Through the use of carbon dating, scientists can reconstruct the ecology and biogeography of organisms living within the past 50,000 years.',\n",
       "    '30 2 • Chemistry of Life Access for free at openstax.org']],\n",
       "  'num_chunks': 2}]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random.sample(pages_and_texts, k=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_number</th>\n",
       "      <th>page_char_count</th>\n",
       "      <th>page_word_count</th>\n",
       "      <th>page_sentence_count_raw</th>\n",
       "      <th>page_token_count</th>\n",
       "      <th>page_sentence_count_spacy</th>\n",
       "      <th>num_chunks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "      <td>42.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>38.50</td>\n",
       "      <td>2626.60</td>\n",
       "      <td>418.98</td>\n",
       "      <td>22.90</td>\n",
       "      <td>656.65</td>\n",
       "      <td>22.98</td>\n",
       "      <td>2.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>12.27</td>\n",
       "      <td>945.61</td>\n",
       "      <td>154.65</td>\n",
       "      <td>10.96</td>\n",
       "      <td>236.40</td>\n",
       "      <td>9.70</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>18.00</td>\n",
       "      <td>284.00</td>\n",
       "      <td>48.00</td>\n",
       "      <td>4.00</td>\n",
       "      <td>71.00</td>\n",
       "      <td>4.00</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>28.25</td>\n",
       "      <td>1983.50</td>\n",
       "      <td>306.25</td>\n",
       "      <td>15.25</td>\n",
       "      <td>495.88</td>\n",
       "      <td>15.25</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>38.50</td>\n",
       "      <td>2813.00</td>\n",
       "      <td>454.50</td>\n",
       "      <td>22.00</td>\n",
       "      <td>703.25</td>\n",
       "      <td>22.50</td>\n",
       "      <td>3.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>48.75</td>\n",
       "      <td>3331.50</td>\n",
       "      <td>545.50</td>\n",
       "      <td>27.75</td>\n",
       "      <td>832.88</td>\n",
       "      <td>28.00</td>\n",
       "      <td>3.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>59.00</td>\n",
       "      <td>4190.00</td>\n",
       "      <td>706.00</td>\n",
       "      <td>61.00</td>\n",
       "      <td>1047.50</td>\n",
       "      <td>47.00</td>\n",
       "      <td>5.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       page_number  page_char_count  page_word_count  page_sentence_count_raw  \\\n",
       "count        42.00            42.00            42.00                    42.00   \n",
       "mean         38.50          2626.60           418.98                    22.90   \n",
       "std          12.27           945.61           154.65                    10.96   \n",
       "min          18.00           284.00            48.00                     4.00   \n",
       "25%          28.25          1983.50           306.25                    15.25   \n",
       "50%          38.50          2813.00           454.50                    22.00   \n",
       "75%          48.75          3331.50           545.50                    27.75   \n",
       "max          59.00          4190.00           706.00                    61.00   \n",
       "\n",
       "       page_token_count  page_sentence_count_spacy  num_chunks  \n",
       "count             42.00                      42.00       42.00  \n",
       "mean             656.65                      22.98        2.79  \n",
       "std              236.40                       9.70        1.00  \n",
       "min               71.00                       4.00        1.00  \n",
       "25%              495.88                      15.25        2.00  \n",
       "50%              703.25                      22.50        3.00  \n",
       "75%              832.88                      28.00        3.00  \n",
       "max             1047.50                      47.00        5.00  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(pages_and_texts)\n",
    "df.describe().round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 42/42 [00:00<00:00, 4195.90it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "117"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "# Split each chunk into its own item\n",
    "pages_and_chunks = []\n",
    "for item in tqdm(pages_and_texts):\n",
    "    for sentence_chunk in item[\"sentence_chunks\"]:\n",
    "        chunk_dict = {}\n",
    "        chunk_dict[\"page_number\"] = item[\"page_number\"]\n",
    "        \n",
    "        # Join the sentences together into a paragraph-like structure, aka a chunk (so they are a single string)\n",
    "        joined_sentence_chunk = \"\".join(sentence_chunk).replace(\"  \", \" \").strip()\n",
    "        joined_sentence_chunk = re.sub(r'\\.([A-Z])', r'. \\1', joined_sentence_chunk) # \".A\" -> \". A\" for any full-stop/capital letter combo \n",
    "        chunk_dict[\"sentence_chunk\"] = joined_sentence_chunk\n",
    "\n",
    "        # Get stats about the chunk\n",
    "        chunk_dict[\"chunk_char_count\"] = len(joined_sentence_chunk)\n",
    "        chunk_dict[\"chunk_word_count\"] = len([word for word in joined_sentence_chunk.split(\" \")])\n",
    "        chunk_dict[\"chunk_token_count\"] = len(joined_sentence_chunk) / 4 # 1 token = ~4 characters\n",
    "        \n",
    "        pages_and_chunks.append(chunk_dict)\n",
    "\n",
    "# How many chunks do we have?\n",
    "len(pages_and_chunks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'page_number': 26,\n",
       "  'sentence_chunk': 'In addition to the hierarchical taxonomic system, Linnaeus was the first to name organisms using two unique names, now called the binomial naming system. Before Linnaeus, the use of common names to refer to organisms caused confusion because there were regional differences in these common names. Binomial names consist of the genus name (which is capitalized) and the species name (all lower-case). Both names are set in italics when they are printed. Every species is given a unique binomial which is recognized the world over, so that a scientist in any location can know which organism is being referred to. For example, the North American blue jay is known uniquely as Cyanocitta cristata. Our own species is Homo sapiens. FIGURE 1.10 These images represent different domains. The scanning electron micrograph shows (a) bacterial cells belong to the domain Bacteria, while the (b) extremophiles, seen all together as colored mats in this hot spring, belong to domain Archaea. Both the (c) sunflower and (d) lion are part of domain Eukarya. (',\n",
       "  'chunk_char_count': 1046,\n",
       "  'chunk_word_count': 170,\n",
       "  'chunk_token_count': 261.5}]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# View a random sample\n",
    "random.sample(pages_and_chunks, k=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_number</th>\n",
       "      <th>chunk_char_count</th>\n",
       "      <th>chunk_word_count</th>\n",
       "      <th>chunk_token_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>117.00</td>\n",
       "      <td>117.00</td>\n",
       "      <td>117.00</td>\n",
       "      <td>117.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>39.95</td>\n",
       "      <td>941.55</td>\n",
       "      <td>149.71</td>\n",
       "      <td>235.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>11.87</td>\n",
       "      <td>498.36</td>\n",
       "      <td>77.04</td>\n",
       "      <td>124.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>18.00</td>\n",
       "      <td>53.00</td>\n",
       "      <td>8.00</td>\n",
       "      <td>13.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>31.00</td>\n",
       "      <td>640.00</td>\n",
       "      <td>102.00</td>\n",
       "      <td>160.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>41.00</td>\n",
       "      <td>1005.00</td>\n",
       "      <td>160.00</td>\n",
       "      <td>251.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>50.00</td>\n",
       "      <td>1175.00</td>\n",
       "      <td>194.00</td>\n",
       "      <td>293.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>59.00</td>\n",
       "      <td>4189.00</td>\n",
       "      <td>628.00</td>\n",
       "      <td>1047.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       page_number  chunk_char_count  chunk_word_count  chunk_token_count\n",
       "count       117.00            117.00            117.00             117.00\n",
       "mean         39.95            941.55            149.71             235.39\n",
       "std          11.87            498.36             77.04             124.59\n",
       "min          18.00             53.00              8.00              13.25\n",
       "25%          31.00            640.00            102.00             160.00\n",
       "50%          41.00           1005.00            160.00             251.25\n",
       "75%          50.00           1175.00            194.00             293.75\n",
       "max          59.00           4189.00            628.00            1047.25"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get stats about our chunks\n",
    "df = pd.DataFrame(pages_and_chunks)\n",
    "df.describe().round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk token count: 24.75 | Text: They are a type of polyunsaturated fat and 44 2 • Chemistry of Life Access for free at openstax.org\n",
      "Chunk token count: 13.25 | Text: credit: modification of work by NASA) CHAPTER OUTLINE\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Show random chunks with under 30 tokens in length\n",
    "min_token_length = 30\n",
    "for row in df[df[\"chunk_token_count\"] <= min_token_length].sample(2).iterrows():\n",
    "    print(f'Chunk token count: {row[1][\"chunk_token_count\"]} | Text: {row[1][\"sentence_chunk\"]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'page_number': 18,\n",
       "  'sentence_chunk': 'INTRODUCTION CHAPTER 1 Introduction to Biology 1.1 Themes and Concepts of Biology 1.2 The Process of Science Viewed from space, Earth (Figure 1.1) offers few clues about the diversity of life forms that reside there. The first forms of life on Earth are thought to have been microorganisms that existed for billions of years before plants and animals appeared. The mammals, birds, and flowers so familiar to us are all relatively recent, originating 130 to 200 million years ago. Humans have inhabited this planet for only the last 2.5 million years, and only in the last 300,000 years have humans started looking like we do today.1.1 Themes and Concepts of Biology LEARNING OBJECTIVES By the end of this section, you will be able to: • Identify and describe the properties of life • Describe the levels of organization among living things • List examples of different sub disciplines in biology Biology is the science that studies life. What exactly is life?This may sound like a silly question with an obvious answer, but it is not easy to define life. For example, a branch of biology called virology studies viruses, which exhibit some of the characteristics of living entities but lack others. It turns out that although viruses can attack living organisms, cause diseases, and even reproduce, FIGURE 1.1 This NASA image is a composite of several satellite-based views of Earth. To make the whole-Earth image, NASA scientists combine observations of different parts of the planet. (',\n",
       "  'chunk_char_count': 1487,\n",
       "  'chunk_word_count': 246,\n",
       "  'chunk_token_count': 371.75},\n",
       " {'page_number': 19,\n",
       "  'sentence_chunk': 'they do not meet the criteria that biologists use to define life. From its earliest beginnings, biology has wrestled with four questions: What are the shared properties that make something “alive”?How do those various living things function?When faced with the remarkable diversity of life, how do we organize the different kinds of organisms so that we can better understand them?And, finally—what biologists ultimately seek to understand—how did this diversity arise and how is it continuing?As new organisms are discovered every day, biologists continue to seek answers to these and other questions. Properties of Life All groups of living organisms share several key characteristics or functions: order, sensitivity or response to stimuli, reproduction, adaptation, growth and development, regulation/homeostasis, energy processing, and evolution. When viewed together, these eight characteristics serve to define life. Order Organisms are highly organized structures that consist of one or more cells. Even very simple, single-celled organisms are remarkably complex.',\n",
       "  'chunk_char_count': 1072,\n",
       "  'chunk_word_count': 152,\n",
       "  'chunk_token_count': 268.0}]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pages_and_chunks_over_min_token_len = df[df[\"chunk_token_count\"] > min_token_length].to_dict(orient=\"records\")\n",
    "pages_and_chunks_over_min_token_len[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Embedding our text chunks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting sentence_transformers\n",
      "  Using cached sentence_transformers-2.7.0-py3-none-any.whl.metadata (11 kB)\n",
      "Collecting transformers<5.0.0,>=4.34.0 (from sentence_transformers)\n",
      "  Using cached transformers-4.40.2-py3-none-any.whl.metadata (137 kB)\n",
      "Requirement already satisfied: tqdm in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (4.66.4)\n",
      "Requirement already satisfied: torch>=1.11.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (2.2.2)\n",
      "Requirement already satisfied: numpy in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (1.26.4)\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (1.4.2)\n",
      "Requirement already satisfied: scipy in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (1.13.0)\n",
      "Requirement already satisfied: huggingface-hub>=0.15.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (0.23.0)\n",
      "Requirement already satisfied: Pillow in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sentence_transformers) (10.3.0)\n",
      "Requirement already satisfied: filelock in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface-hub>=0.15.1->sentence_transformers) (3.13.4)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface-hub>=0.15.1->sentence_transformers) (2024.3.1)\n",
      "Requirement already satisfied: packaging>=20.9 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface-hub>=0.15.1->sentence_transformers) (23.2)\n",
      "Requirement already satisfied: pyyaml>=5.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface-hub>=0.15.1->sentence_transformers) (6.0.1)\n",
      "Requirement already satisfied: requests in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface-hub>=0.15.1->sentence_transformers) (2.31.0)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from huggingface-hub>=0.15.1->sentence_transformers) (4.11.0)\n",
      "Requirement already satisfied: sympy in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from torch>=1.11.0->sentence_transformers) (1.12)\n",
      "Requirement already satisfied: networkx in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from torch>=1.11.0->sentence_transformers) (3.3)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from torch>=1.11.0->sentence_transformers) (3.1.3)\n",
      "Requirement already satisfied: colorama in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from tqdm->sentence_transformers) (0.4.6)\n",
      "Requirement already satisfied: regex!=2019.12.17 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from transformers<5.0.0,>=4.34.0->sentence_transformers) (2024.5.10)\n",
      "Requirement already satisfied: tokenizers<0.20,>=0.19 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from transformers<5.0.0,>=4.34.0->sentence_transformers) (0.19.1)\n",
      "Requirement already satisfied: safetensors>=0.4.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from transformers<5.0.0,>=4.34.0->sentence_transformers) (0.4.3)\n",
      "Requirement already satisfied: joblib>=1.2.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scikit-learn->sentence_transformers) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from scikit-learn->sentence_transformers) (3.5.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from jinja2->torch>=1.11.0->sentence_transformers) (2.1.5)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests->huggingface-hub>=0.15.1->sentence_transformers) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests->huggingface-hub>=0.15.1->sentence_transformers) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests->huggingface-hub>=0.15.1->sentence_transformers) (2.2.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from requests->huggingface-hub>=0.15.1->sentence_transformers) (2024.2.2)\n",
      "Requirement already satisfied: mpmath>=0.19 in c:\\users\\cedai\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.12_qbz5n2kfra8p0\\localcache\\local-packages\\python312\\site-packages (from sympy->torch>=1.11.0->sentence_transformers) (1.3.0)\n",
      "Using cached sentence_transformers-2.7.0-py3-none-any.whl (171 kB)\n",
      "Using cached transformers-4.40.2-py3-none-any.whl (9.0 MB)\n",
      "Installing collected packages: transformers, sentence_transformers\n",
      "Successfully installed sentence_transformers-2.7.0 transformers-4.40.2\n"
     ]
    }
   ],
   "source": [
    "!pip install sentence_transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\CedAI\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\huggingface_hub\\file_download.py:157: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\\Users\\CedAI\\.cache\\huggingface\\hub\\models--sentence-transformers--all-mpnet-base-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.\n",
      "To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development\n",
      "  warnings.warn(message)\n",
      "C:\\Users\\CedAI\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\huggingface_hub\\file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sentence: The Sentences Transformers library provides an easy and open-source way to create embeddings.\n",
      "Embedding: [-2.07982380e-02  3.03164814e-02 -2.01218016e-02  6.86484724e-02\n",
      " -2.55255997e-02 -8.47687293e-03 -2.07233126e-04 -6.32377788e-02\n",
      "  2.81606875e-02 -3.33353579e-02  3.02634127e-02  5.30721620e-02\n",
      " -5.03526814e-02  2.62288675e-02  3.33313569e-02 -4.51577678e-02\n",
      "  3.63044888e-02 -1.37121032e-03 -1.20170955e-02  1.14947390e-02\n",
      "  5.04510701e-02  4.70856987e-02  2.11913902e-02  5.14606498e-02\n",
      " -2.03746818e-02 -3.58889513e-02 -6.67784712e-04 -2.94393711e-02\n",
      "  4.95859273e-02 -1.05639333e-02 -1.52014121e-02 -1.31760375e-03\n",
      "  4.48197238e-02  1.56023037e-02  8.60379316e-07 -1.21392997e-03\n",
      " -2.37978734e-02 -9.09389928e-04  7.34485453e-03 -2.53934204e-03\n",
      "  5.23370393e-02 -4.68043424e-02  1.66214556e-02  4.71579358e-02\n",
      " -4.15599272e-02  9.01964202e-04  3.60277891e-02  3.42214145e-02\n",
      "  9.68227386e-02  5.94829321e-02 -1.64984372e-02 -3.51249129e-02\n",
      "  5.92516223e-03 -7.07908242e-04 -2.41031013e-02  3.49740684e-02\n",
      " -2.94746794e-02  6.04268676e-03 -9.80647560e-03  2.83217803e-02\n",
      " -1.85376033e-02  3.63213234e-02  1.30292224e-02 -3.71232964e-02\n",
      "  5.27256317e-02 -1.19707184e-02 -7.18082562e-02  1.24431858e-02\n",
      " -6.70565711e-03  7.42154494e-02  1.16356220e-02 -1.74533445e-02\n",
      " -1.82405543e-02 -1.88930854e-02  2.82415077e-02  1.32828979e-02\n",
      " -3.51909511e-02  8.87303147e-04  5.79572357e-02  3.22093591e-02\n",
      " -3.48579930e-03  4.13768888e-02  1.44358287e-02 -3.28044333e-02\n",
      " -9.79087315e-03 -3.16492766e-02  4.23870869e-02 -4.70846556e-02\n",
      " -2.08937060e-02 -1.91249382e-02 -1.22627337e-02  1.01606213e-02\n",
      "  3.91922034e-02 -2.61895731e-02  1.09027829e-02  1.35723408e-02\n",
      " -5.79267256e-02 -3.21499743e-02 -5.75725315e-03 -2.43515763e-02\n",
      "  5.23416735e-02  5.46123320e-03 -2.30995920e-02  2.57173297e-03\n",
      " -6.63346648e-02  3.54125276e-02 -1.03907110e-02  2.25410182e-02\n",
      " -1.84573922e-02 -2.42007058e-02 -4.78365757e-02 -4.79225954e-03\n",
      " -5.34138419e-02  3.01790312e-02 -1.56130781e-02 -5.51476739e-02\n",
      " -3.91874276e-02  5.92153184e-02 -3.47646698e-02  9.68120992e-03\n",
      "  2.13415399e-02  2.30416916e-02  1.91713069e-02  2.77378503e-02\n",
      " -7.73506891e-03  1.04445824e-02 -2.67719645e-02 -2.40199845e-02\n",
      " -1.92290135e-02  3.91500397e-03 -2.54714377e-02  3.61942910e-02\n",
      "  5.12867235e-02 -8.41693580e-03 -3.13829780e-02  1.47483414e-02\n",
      "  2.13940311e-02 -3.84900533e-02  2.01945417e-02  1.20765790e-02\n",
      " -3.12069734e-03  7.84031395e-03  3.30336625e-03 -4.94357310e-02\n",
      "  5.83886765e-02  3.26138083e-03  4.84486669e-03 -4.50682491e-02\n",
      "  2.45682821e-02  3.55428122e-02 -5.32506146e-02  9.21152830e-02\n",
      "  2.04395130e-02 -3.36951837e-02 -6.19803518e-02 -2.11039279e-02\n",
      "  7.82359019e-02  5.11907935e-02  5.93170635e-02 -1.25109349e-04\n",
      "  4.96350080e-02 -1.55722629e-02 -3.35678365e-03  1.82016250e-02\n",
      " -2.73444243e-02 -1.08771911e-02  1.41476123e-02  1.09877735e-02\n",
      "  4.32553748e-03  8.23311284e-02 -9.85321938e-04  7.58791044e-02\n",
      "  9.44997557e-03  2.37687863e-02  1.61928162e-02  6.24993593e-02\n",
      "  4.75922301e-02 -3.92629066e-03  9.07524377e-02  4.49874736e-02\n",
      " -3.47131677e-02  2.14077067e-02 -3.35604250e-02  4.93849106e-02\n",
      "  1.08669475e-02  2.63447706e-02 -3.26089710e-02  8.00303444e-02\n",
      "  9.29766428e-03  7.16579473e-03 -2.79172510e-02 -3.06821018e-02\n",
      "  4.01060982e-03 -4.93906997e-02 -3.13769490e-03  4.00537625e-02\n",
      " -3.97855230e-02  5.48013896e-02  1.36247108e-05 -8.38373229e-02\n",
      " -1.21547673e-02  3.40949744e-02  3.22404783e-03  6.11846447e-02\n",
      "  5.60067110e-02  9.62872989e-03  2.54616383e-02 -4.64168563e-02\n",
      " -3.98900919e-02  7.68132061e-02  2.28408966e-02 -2.26568710e-02\n",
      " -1.91193689e-02 -6.53029010e-02  4.56781350e-02 -4.43655532e-03\n",
      "  1.49631752e-02 -2.15077829e-02  2.74248491e-03  1.90358367e-02\n",
      "  5.91888539e-02 -2.47569028e-02  3.66144367e-02  5.63083924e-02\n",
      " -8.86446517e-03 -1.74324457e-02 -1.03289704e-03  2.47666575e-02\n",
      "  1.30762970e-02  5.04632853e-02 -5.28495200e-03  5.92397526e-02\n",
      "  6.29906431e-02 -4.36782949e-02 -4.97830212e-02  5.56296892e-02\n",
      " -2.44854167e-02 -8.26754645e-02  2.04910580e-02 -1.06446259e-01\n",
      "  6.64840965e-03  2.97304243e-02 -2.36440301e-02 -8.84614419e-03\n",
      "  2.45557935e-03 -3.35234664e-02  7.52212405e-02 -5.89879937e-02\n",
      " -3.67808305e-02  3.41542885e-02  5.41130938e-02 -1.74904875e-02\n",
      "  1.33920619e-02  4.71682698e-02  1.46116083e-02 -2.12311056e-02\n",
      " -6.55338317e-02  1.23857539e-02  2.76074354e-02 -8.02161079e-03\n",
      " -4.59636375e-02 -8.22443701e-03  9.16955154e-03 -1.56399217e-02\n",
      "  7.54623162e-03  1.58310740e-03 -3.03958170e-02 -5.10671064e-02\n",
      "  1.96313635e-02  1.26263164e-02 -1.51737442e-03  2.02891398e-02\n",
      "  1.37817292e-02  1.49110425e-02  2.50766762e-02 -3.62870619e-02\n",
      "  1.08085573e-02  2.74134520e-03  1.81510765e-02  5.39872125e-02\n",
      " -4.74542081e-02 -4.28731218e-02 -2.89914217e-02  2.13235412e-02\n",
      " -3.85161452e-02  6.31922409e-02 -5.77975884e-02  3.77888302e-03\n",
      " -2.54394244e-02 -1.77249574e-04  9.08244867e-03  1.59095395e-02\n",
      "  4.11799811e-02 -3.94367725e-02 -9.64435004e-03  1.30791366e-02\n",
      "  6.87962547e-02  4.32192534e-02  7.53993052e-04  6.77741542e-02\n",
      "  4.93705496e-02 -3.47822881e-03 -1.06054908e-02  6.72495551e-03\n",
      " -1.39062032e-02  4.88276668e-02 -1.05735231e-02  3.50223761e-03\n",
      "  2.90224864e-03  2.40044333e-02  1.20271910e-02 -2.09797323e-02\n",
      " -2.39111837e-02  3.26579399e-02 -1.01325649e-03 -5.92756644e-03\n",
      " -7.40533322e-03  3.63146747e-03 -2.26698648e-02 -2.21241433e-02\n",
      "  3.86996567e-02  1.72321647e-02  3.85920219e-02 -5.04711419e-02\n",
      " -3.42145115e-02 -4.00444120e-02 -3.57910804e-02 -4.62561101e-02\n",
      "  6.70232475e-02 -4.61648591e-03 -3.29676410e-03  2.08444241e-02\n",
      " -5.14256908e-03 -5.00849858e-02  2.22503990e-02  4.66933213e-02\n",
      "  1.36208758e-02  1.77530199e-02  4.28077625e-03 -2.79332083e-02\n",
      " -1.93420649e-02 -3.87859903e-02 -3.09554935e-02 -6.64133951e-02\n",
      " -1.13434456e-02  1.64266992e-02  1.77629348e-02 -2.28225649e-03\n",
      " -3.30087431e-02 -1.36262504e-03 -2.17934120e-02 -2.67508253e-02\n",
      " -1.26375668e-02  1.61867787e-03 -4.95672412e-02  7.85445720e-02\n",
      "  4.10962664e-02  9.65920091e-03 -1.14643490e-02  1.68858271e-03\n",
      "  5.37663065e-02  2.05532974e-03 -4.11200076e-02  1.46329878e-02\n",
      " -3.75564210e-02 -3.35689932e-02  5.19255316e-03 -6.33088797e-02\n",
      "  3.32963504e-02  8.76120292e-03  1.33860274e-03 -3.95749789e-03\n",
      " -1.61677860e-02  8.26746672e-02  4.75944839e-02 -3.43055241e-02\n",
      "  2.50881240e-02 -3.50977108e-02  3.68657038e-02  4.12648870e-03\n",
      "  4.16018367e-02 -1.35181695e-01 -4.76338230e-02 -1.20025733e-02\n",
      " -3.48891728e-02  3.25454995e-02 -2.93576205e-03 -4.85056080e-03\n",
      " -1.04223728e-01  2.78610419e-02  1.41570298e-02  3.94395404e-02\n",
      " -3.88806947e-02 -1.42463567e-02 -5.19984476e-02  8.92733410e-03\n",
      " -1.99770816e-02 -2.51724981e-02 -3.41300108e-02  1.93041656e-02\n",
      " -5.20208068e-02 -6.72000423e-02 -9.46363993e-03 -1.25585531e-03\n",
      " -5.66048622e-02  2.62098331e-02  9.91586968e-03  4.38286625e-02\n",
      "  2.26644194e-03 -3.11895777e-02 -6.25467300e-02 -3.87792960e-02\n",
      " -6.83939084e-02  4.93721776e-02  5.85507564e-02 -4.08729501e-02\n",
      " -1.98638290e-02 -2.12634262e-02  4.98037152e-02 -4.51748818e-02\n",
      " -2.37141717e-02  2.32674945e-02  1.00594811e-01  9.87118762e-03\n",
      " -1.38014741e-02 -5.21041267e-02  9.08212271e-03  1.72427651e-02\n",
      "  5.91431521e-02  2.62336675e-02 -7.04639126e-03 -1.50032053e-02\n",
      " -3.76661541e-03  6.28256705e-03 -5.23981340e-02 -4.96638753e-02\n",
      "  3.06611080e-02 -3.33649572e-03  2.34911330e-02 -8.58829990e-02\n",
      " -4.62449789e-02  5.59701025e-02  3.09089141e-04  2.01729089e-02\n",
      " -2.98053026e-03  1.76645946e-02  1.54669899e-02 -7.41718411e-02\n",
      "  7.34993769e-03 -1.05015589e-02  2.45247297e-02  1.36879096e-02\n",
      " -1.17803952e-02  4.51545082e-02  3.29039358e-02 -3.50388116e-03\n",
      " -2.71314476e-02 -5.27363829e-02 -4.60164137e-02  2.22848728e-02\n",
      "  2.62271762e-02  5.56152035e-03  1.45789124e-02 -2.97145732e-02\n",
      "  3.57042812e-02  2.22533792e-02  3.89616936e-02 -7.92634785e-02\n",
      " -9.01093241e-03  2.19013132e-02 -5.49054984e-03  8.69956426e-03\n",
      "  4.33030836e-02 -2.12631449e-02  1.13291964e-02 -6.33699000e-02\n",
      "  3.63723375e-02  2.67442558e-02 -6.64252192e-02  1.70499273e-02\n",
      " -2.79690735e-02  2.36351625e-03 -1.81952789e-02  1.52955288e-02\n",
      " -8.50431342e-03  1.16647566e-02 -9.75921974e-02 -2.92093363e-02\n",
      " -5.42547479e-02  3.61234397e-02  3.25116925e-02  8.26977566e-03\n",
      " -2.96510087e-04  1.11555504e-02 -3.85188200e-02  2.36161295e-02\n",
      "  9.85923037e-03  5.73998466e-02  4.86061275e-02 -1.37579115e-02\n",
      " -6.19217660e-03  1.11973090e-02 -3.37174907e-02 -1.10515645e-02\n",
      " -7.08333626e-02 -1.01816189e-02 -3.66010740e-02 -1.55561045e-02\n",
      " -2.13110689e-02 -1.02760978e-02 -4.35733795e-02  5.55186458e-02\n",
      " -3.76548320e-02  5.29252067e-02 -3.45223919e-02 -2.43007462e-03\n",
      "  7.25553334e-02  4.45072912e-03  4.71417829e-02 -9.43883043e-03\n",
      " -1.98978763e-02  5.71899675e-02  8.60540941e-02 -5.25057688e-02\n",
      " -1.39549794e-02  1.17373196e-02  1.33974329e-02 -4.73052822e-02\n",
      " -5.41673899e-02  4.62725312e-02 -2.58969907e-02  1.51415523e-02\n",
      "  3.38944979e-02 -3.78262368e-03 -5.76043427e-02 -1.60082113e-02\n",
      "  2.42738687e-02  3.37360799e-02 -1.96821336e-02 -2.53464244e-02\n",
      " -4.75617386e-02 -5.68755493e-02 -2.28194185e-02  3.83187607e-02\n",
      " -1.78330671e-02  1.35964211e-02  7.85977289e-04  9.74010490e-03\n",
      "  3.34298797e-02 -2.60133836e-02 -7.38568371e-03  3.56451198e-02\n",
      " -2.68532690e-02 -7.53624737e-02 -2.66983621e-02 -4.46457383e-33\n",
      " -3.31646726e-02  1.41703496e-02 -3.92909385e-02 -3.46319005e-02\n",
      " -5.88669674e-03 -1.18211899e-02  1.53951254e-02  1.18473629e-02\n",
      "  1.07757216e-02  3.62140238e-02  7.87952263e-03 -2.31845789e-02\n",
      "  1.07623292e-02  1.72346048e-02  9.54203540e-04  2.83640381e-02\n",
      "  2.37419382e-02 -1.48057733e-02  1.24199910e-03  3.52351856e-03\n",
      "  2.33735722e-02  5.58308065e-02  5.38327917e-02 -3.74079123e-02\n",
      " -2.11805589e-02  1.52734283e-04 -7.27793435e-03  5.50559070e-03\n",
      "  3.05823497e-02  4.54633273e-02 -3.35786827e-02  3.16142552e-02\n",
      " -2.56394572e-03  3.96355540e-02 -1.47573715e-02  5.67167737e-02\n",
      " -5.62787540e-02 -5.04598394e-03  3.56154703e-02 -2.76198499e-02\n",
      " -2.32292004e-02 -4.63292375e-02 -3.70919667e-02 -4.23188992e-02\n",
      "  3.70306708e-02  7.88718835e-03  3.85176092e-02  1.74774218e-03\n",
      "  5.62762935e-03  6.18107617e-03 -6.90268874e-02 -9.42963734e-03\n",
      " -7.74671743e-03  1.68349929e-02  1.22767696e-02  2.26406418e-02\n",
      "  1.21009024e-02  1.11743575e-02  1.21539012e-02 -1.16862049e-02\n",
      " -4.41612490e-02  2.30047163e-02  2.20672600e-02 -5.87504320e-02\n",
      " -3.96428406e-02  6.83135390e-02 -3.29948030e-02 -3.66774648e-02\n",
      " -3.53654847e-02  1.76185183e-02  6.95639383e-03  5.92692979e-02\n",
      "  4.12156358e-02  7.98109546e-02 -5.36562456e-03  1.14238905e-02\n",
      " -2.96388585e-02 -1.15411747e-02  2.22812761e-02  7.93186668e-03\n",
      "  2.60355901e-02  1.28212431e-02  1.71345454e-02 -6.90185698e-03\n",
      " -1.07603315e-02  1.35714281e-02 -9.90788802e-04 -6.16075695e-02\n",
      "  4.40514050e-02 -8.26595817e-04 -2.78340820e-02 -1.23619288e-02\n",
      "  1.34629281e-02 -3.85745391e-02  1.08701934e-03  2.18712371e-02\n",
      " -3.32398899e-02  1.84615869e-02 -5.10114105e-03  3.74664925e-02\n",
      " -3.67540866e-03 -2.19246391e-02 -4.96481452e-03 -9.59841069e-03\n",
      "  2.33591292e-02  1.04876896e-02  4.38721962e-02 -1.51424268e-02\n",
      " -6.30309433e-02  8.23258795e-03 -1.09130852e-02 -4.06409465e-02\n",
      " -6.21690899e-02  2.21326221e-02 -2.71434337e-02  4.05539200e-02\n",
      " -8.09452403e-03 -1.76412752e-03  3.01526617e-02 -5.42263780e-03\n",
      " -4.69821766e-02 -1.73768736e-02  4.11630459e-02  3.20634581e-02\n",
      " -2.22943872e-02 -1.58162117e-02 -4.50720750e-02  5.69486171e-02\n",
      "  4.71595712e-02 -5.78058548e-02  1.32475132e-02 -4.71289409e-03\n",
      "  1.66824478e-07  4.81090359e-02  5.03627993e-02  5.45263477e-02\n",
      "  2.07568444e-02 -1.19080665e-02 -6.37491141e-03  5.26364334e-03\n",
      "  7.21949339e-02 -2.21762974e-02  2.20103208e-02 -9.90467262e-04\n",
      " -1.37163782e-02  6.89205760e-03  2.46912725e-02 -1.39462039e-01\n",
      "  2.56980583e-03 -4.64826338e-02 -4.04967591e-02 -6.08555712e-02\n",
      " -1.53212827e-02  1.36129901e-01  9.45034772e-02  4.25741673e-02\n",
      "  4.67131063e-02 -2.30677724e-02 -1.20966071e-02  3.86672728e-02\n",
      "  2.11647036e-03 -2.51473505e-02 -1.15076266e-02 -3.46506462e-02\n",
      " -2.29533911e-02 -6.33849716e-03 -3.05175800e-02 -1.56236654e-02\n",
      "  1.39513640e-02  3.27461632e-04  2.00325483e-03  4.15106118e-03\n",
      " -2.22924650e-02 -3.62589844e-02 -2.36579664e-02 -1.87817123e-02\n",
      " -1.96289122e-02  4.52125855e-02 -8.12568441e-02 -2.14568973e-02\n",
      " -4.41543348e-02 -2.68475693e-02  2.01974306e-02  2.82992050e-03\n",
      " -1.95011571e-02 -3.45330723e-02  2.26913746e-02  3.78325991e-02\n",
      " -1.02544092e-02 -2.19759461e-03 -8.96744356e-02 -4.50030901e-02\n",
      "  8.09704140e-03 -2.05805507e-02 -2.02998742e-02 -2.09922288e-02\n",
      " -1.79404691e-02  5.81897497e-02 -7.63653126e-03  1.50847482e-02\n",
      "  1.78279755e-34  4.86179367e-02  4.22228500e-02  4.71595638e-02\n",
      "  5.89047186e-02  3.99785526e-02 -5.27071357e-02  1.56905279e-02\n",
      " -5.25091542e-04  1.13652069e-02 -6.56410307e-02 -2.20849309e-02]\n",
      "\n",
      "Sentence: Sentences can be embedded one by one or as a list of strings.\n",
      "Embedding: [ 4.31717932e-02 -5.38702197e-02 -3.78044434e-02  4.27235775e-02\n",
      " -2.35408898e-02  3.44860777e-02  2.89586801e-02  1.92817335e-03\n",
      "  2.41733156e-02 -3.17012705e-02  7.32856020e-02  1.25590172e-02\n",
      "  3.64620686e-02 -2.05251761e-02  2.81973965e-02 -6.87329024e-02\n",
      "  4.22231406e-02  9.31759947e-04  3.54035348e-02  1.41787557e-02\n",
      "  7.83996563e-03  2.31178999e-02 -4.84735472e-03  1.07174041e-02\n",
      "  4.39489027e-03  5.47800865e-03 -3.80338766e-02 -3.05490126e-03\n",
      "  5.72229596e-03 -6.78213984e-02 -4.88007441e-02 -1.45032471e-02\n",
      "  6.68012304e-03 -7.17480034e-02  1.64644814e-06  1.07564488e-02\n",
      " -3.60922776e-02 -2.37057190e-02 -5.22791483e-02  3.46110240e-02\n",
      " -5.42179868e-03  1.62611622e-02  1.96564905e-02  2.25395951e-02\n",
      " -2.25997670e-03  4.06341814e-02  8.17157403e-02  2.48179846e-02\n",
      "  5.31884544e-02  7.82715008e-02 -1.91813726e-02 -1.94087140e-02\n",
      " -2.62805391e-02 -2.44082920e-02  5.49405366e-02  1.90318711e-02\n",
      "  1.60811339e-02 -2.68895570e-02 -8.24690424e-03  7.33444840e-02\n",
      "  1.00123286e-02  2.93315649e-02  3.42859561e-03 -2.13270113e-02\n",
      " -1.62442937e-03 -5.56254108e-03 -7.64879659e-02 -5.85450456e-02\n",
      " -2.82272212e-02  7.51845958e-03  7.11225644e-02  1.95454923e-03\n",
      "  5.45929605e-03  3.22320568e-03  5.12799993e-02 -3.54106054e-02\n",
      " -5.03608398e-02  4.70519327e-02  5.15478244e-03  1.52287316e-02\n",
      " -1.06680552e-02  3.16298380e-02 -9.09039099e-03 -4.01327461e-02\n",
      " -4.35235761e-02 -1.94969270e-02  1.65604874e-02 -4.71168905e-02\n",
      " -3.92091721e-02 -3.07756793e-02 -2.94167455e-02 -4.20826674e-02\n",
      "  2.27070786e-03 -2.78329495e-02  1.69421546e-02  7.74499308e-03\n",
      " -5.23741953e-02 -4.50040027e-02  3.83605249e-02 -4.90786657e-02\n",
      "  5.06619103e-02  1.01615479e-02 -1.25021664e-02 -4.64556133e-03\n",
      " -1.54539961e-02  1.58862434e-02  1.18370419e-02 -3.59232761e-02\n",
      " -7.76225701e-02  3.43358852e-02 -2.14709975e-02 -6.86098710e-02\n",
      " -5.46236485e-02  7.83901736e-02 -3.00702602e-02 -3.37550342e-02\n",
      " -4.04998846e-02  4.80515473e-02  9.53903235e-03  2.31399257e-02\n",
      " -8.16115290e-02 -6.51697116e-03  1.54213011e-02  7.04257712e-02\n",
      " -1.25069041e-02 -2.48266719e-02 -1.71328913e-02  6.13113353e-03\n",
      "  5.44412658e-02 -1.40566211e-02 -6.24513719e-03  3.65788117e-02\n",
      "  7.36230388e-02 -6.05681632e-03 -3.61630507e-02 -1.42203597e-03\n",
      "  4.43166047e-02 -3.14511289e-03  3.18767466e-02 -1.30948033e-02\n",
      " -3.69524471e-02 -4.98032896e-03  1.30019151e-03 -2.05213651e-02\n",
      "  2.06277482e-02  5.93871018e-03 -3.07165296e-03 -3.97513174e-02\n",
      "  4.29890156e-02  6.49802759e-02 -6.76022172e-02  5.41655608e-02\n",
      "  1.52562419e-03 -3.72908786e-02 -4.02427278e-02 -2.28772350e-02\n",
      "  1.31769747e-01  4.87876730e-03  1.39470790e-02  4.92435619e-02\n",
      "  2.49219369e-02 -8.76092725e-03 -5.38767688e-03 -2.65595634e-02\n",
      " -1.19766314e-02 -2.32006889e-02 -2.67433878e-02  5.66912768e-03\n",
      "  2.21722033e-02  4.67294194e-02 -5.78487851e-02  8.22120905e-02\n",
      " -3.36834602e-03  8.09647962e-02  1.41423400e-02  1.02393061e-01\n",
      " -5.76833636e-03 -1.15877483e-02  4.90584970e-02  5.87829798e-02\n",
      "  6.50030673e-02  4.74621691e-02 -2.89464220e-02 -1.76583638e-03\n",
      "  3.32561359e-02  2.91198380e-02  6.03811331e-02  3.73511779e-04\n",
      "  1.06576141e-02 -5.96284717e-02 -7.28601813e-02  2.95080040e-02\n",
      "  9.54463053e-03 -2.71542426e-02 -5.63305095e-02  9.66721331e-04\n",
      " -4.77729514e-02  4.67576645e-02  4.87255119e-03 -6.57519549e-02\n",
      " -1.42248468e-02  3.99873033e-02 -1.09797204e-02  7.68941492e-02\n",
      " -4.00004275e-02  2.96826307e-02  2.81303357e-02 -5.55424877e-02\n",
      "  6.31263480e-03  5.00450172e-02  1.89884249e-02  5.38683757e-02\n",
      " -1.95981599e-02  1.08600957e-02  1.64150447e-02  1.44135114e-02\n",
      "  1.71448812e-02  2.17624810e-02 -4.98862937e-02  1.56105142e-02\n",
      "  4.83736442e-03  1.87053401e-02 -3.18541913e-03  2.66863033e-02\n",
      "  5.55552654e-02 -4.88005728e-02 -3.02929208e-02  2.52110437e-02\n",
      "  1.07264789e-02  1.88270397e-02 -1.50687881e-02  3.43831964e-02\n",
      "  4.15124856e-02  1.37788802e-02 -5.54849021e-02  1.43848797e-02\n",
      " -5.88139258e-02 -6.01676926e-02  2.69856248e-02 -5.46130091e-02\n",
      "  8.14635120e-03 -1.17758336e-02  1.57442745e-02  1.43904414e-03\n",
      " -2.64554657e-02 -4.48876619e-02  4.39732485e-02 -1.06157466e-04\n",
      " -2.25905776e-02  3.00296713e-02  1.97440609e-02  7.44074583e-03\n",
      " -1.93789750e-02  8.09798483e-03  4.34859879e-02 -1.08624656e-04\n",
      " -3.77225876e-02  2.67196223e-02 -4.63157296e-02 -1.53403066e-03\n",
      "  8.05313885e-03 -4.30900790e-02 -2.13848669e-02  1.20185493e-02\n",
      "  8.41403287e-03  2.48271949e-03 -3.09566557e-02 -9.05277804e-02\n",
      " -4.76693958e-02  1.22606261e-02 -1.36466855e-02 -2.63655279e-02\n",
      " -7.65550090e-03  8.72378144e-03  2.65724901e-02  8.40030145e-04\n",
      " -5.55927679e-03 -9.29537416e-03  3.19337435e-02  5.94646744e-02\n",
      "  1.83205623e-02 -7.56546557e-02 -5.59389070e-02 -1.20870909e-02\n",
      " -3.16260904e-02  3.62187214e-02  7.53610022e-03 -6.15654066e-02\n",
      " -2.30459012e-02 -3.51668149e-03  1.23331957e-02 -9.67636984e-03\n",
      "  4.96861823e-02 -8.42256472e-02  1.52396280e-02 -1.82445180e-02\n",
      "  7.70462826e-02  9.28717554e-02  4.03724536e-02  1.11732617e-01\n",
      " -1.03270616e-02 -2.54558176e-02  2.13154089e-02 -1.16185972e-03\n",
      "  2.82601430e-03  5.06967045e-02 -3.13697234e-02 -8.14271253e-03\n",
      "  1.38387419e-02  4.66889627e-02  5.09670898e-02  3.77153978e-02\n",
      " -2.94988528e-02  3.60631831e-02 -2.61166086e-03  2.72283418e-04\n",
      " -6.71807006e-02 -6.54025972e-02 -3.43590640e-02  1.91067588e-02\n",
      "  4.13294248e-02 -1.10970605e-02  4.51951735e-02 -5.93565181e-02\n",
      "  1.06963506e-02 -1.82230026e-02 -5.65814264e-02  1.20386807e-02\n",
      "  4.44773734e-02  1.87050030e-02  1.63809992e-02  5.51150590e-02\n",
      " -2.23332308e-02  2.12861542e-02 -1.20339980e-02  3.26753035e-02\n",
      "  1.47004602e-02 -8.16679653e-03  1.12904888e-02 -3.00620496e-02\n",
      " -2.34345887e-02 -2.68646199e-02 -1.28720899e-03 -7.67189488e-02\n",
      "  2.22607632e-03 -5.89476246e-03  2.63103675e-02  2.07129726e-03\n",
      " -6.91151842e-02 -1.43792052e-02  2.68788133e-02 -3.51539738e-02\n",
      " -2.69612279e-02  2.54716794e-03 -6.48881793e-02  3.18728127e-02\n",
      "  1.70126799e-02 -4.54004221e-02 -1.80616193e-02 -1.61116160e-02\n",
      "  5.70773818e-02 -2.78272806e-03 -6.45585209e-02  7.86597878e-02\n",
      "  2.29075179e-02  6.81841467e-03 -9.11737606e-03 -2.27726419e-02\n",
      " -4.76526655e-02  4.88430895e-02 -2.09891796e-02 -2.43693776e-02\n",
      " -5.01209125e-03  6.70253560e-02  6.91368897e-03  2.25843024e-02\n",
      "  2.51125302e-02 -6.92506228e-03  8.59398767e-03  2.38977261e-02\n",
      "  3.29738371e-02 -1.05310559e-01  1.22094750e-02 -1.22263553e-02\n",
      " -5.73768988e-02  1.84311718e-02  2.97158584e-02 -6.09429218e-02\n",
      " -6.55256063e-02  3.55713069e-02  5.64320525e-03  3.34644900e-03\n",
      " -3.59686911e-02 -8.83415155e-03 -6.97895065e-02  6.89779446e-02\n",
      " -4.88212518e-03  2.23995317e-02 -3.16053480e-02 -7.41182547e-03\n",
      "  3.19351032e-02 -5.18788472e-02  2.11601313e-02 -5.03340289e-02\n",
      "  9.10576340e-03  2.13354379e-02  1.66838132e-02  3.49020176e-02\n",
      " -6.38500080e-02 -6.75625168e-03 -1.27404984e-02 -4.63366397e-02\n",
      " -1.14779985e-02  2.08778605e-02  2.44822055e-02  3.66462395e-03\n",
      " -2.86096497e-03  2.29390636e-02  2.13745907e-02 -3.48901153e-02\n",
      " -3.00387684e-02  4.78870459e-02  5.83370738e-02 -9.70499124e-03\n",
      "  1.38234440e-02 -3.27485651e-02 -8.11471255e-04  9.54227336e-03\n",
      "  1.20401271e-02  1.97230149e-02 -4.74868400e-04 -1.39226215e-02\n",
      " -5.21070324e-02 -1.75592676e-02 -5.41698523e-02 -1.17970314e-02\n",
      " -1.71030574e-02 -3.50195281e-02  3.38661522e-02 -6.76587224e-02\n",
      " -2.27606837e-02  1.95607003e-02  5.50249331e-02  1.22029092e-02\n",
      " -1.75168249e-03  7.22446106e-03  1.16350008e-02 -1.61908008e-02\n",
      " -3.37754861e-02  3.22626978e-02 -2.03813743e-02 -2.33859457e-02\n",
      " -1.29991798e-02 -1.66799035e-02  1.03070727e-02 -1.46030085e-02\n",
      " -7.79071078e-02 -8.25812295e-02 -3.38809714e-02  3.81114073e-02\n",
      "  7.86007289e-03  2.41454989e-02 -2.75715142e-02  1.30867278e-02\n",
      " -7.88589939e-03  1.78651102e-02  5.37322946e-02 -3.01823486e-02\n",
      "  1.69455428e-02  1.19571630e-02  3.52576899e-04  4.90209274e-02\n",
      " -8.57210252e-03  1.71268440e-03  4.83876327e-03 -4.10081260e-02\n",
      " -4.68120798e-02 -2.32555578e-03 -5.16776256e-02  3.10030654e-02\n",
      "  1.60961710e-02 -1.00803263e-02 -3.72484233e-03 -3.53388153e-02\n",
      "  2.95961499e-02  2.89096814e-02 -7.59911835e-02 -5.02981097e-02\n",
      " -2.11783685e-02  3.20462361e-02 -3.84538844e-02  2.45103203e-02\n",
      " -2.04188190e-02  6.02108985e-03 -9.81940050e-03  3.74778025e-02\n",
      "  3.40838917e-02  1.28864385e-02  5.67342490e-02 -8.09703916e-02\n",
      " -8.93608853e-03  1.33352662e-02 -2.51565892e-02  2.58414261e-03\n",
      " -6.51803017e-02  1.34400064e-02 -2.04682294e-02  6.53378153e-03\n",
      "  4.56975773e-03  1.99271291e-02 -6.07340522e-02  1.40691567e-02\n",
      " -5.75333722e-02  9.79802571e-03  3.55393142e-02 -2.45283525e-02\n",
      " -4.73311683e-03 -2.77493000e-02  2.34282576e-02 -8.76070990e-05\n",
      "  7.30440114e-03  1.42028732e-02  4.92807254e-02 -3.16540375e-02\n",
      " -1.34901861e-02  3.08487844e-02  2.80402005e-02 -4.33068909e-02\n",
      " -4.42284122e-02  3.80739085e-02  9.47956723e-05 -4.34896685e-02\n",
      "  1.43868318e-02  2.44328962e-03 -4.84073646e-02  1.08955512e-02\n",
      " -9.87490173e-03  4.59295362e-02  3.96379493e-02 -2.60117091e-02\n",
      "  2.48134527e-02 -5.37149534e-02  5.62824719e-02  8.81360006e-03\n",
      "  5.25077060e-02 -1.47371134e-02 -1.74380802e-02  3.45084034e-02\n",
      "  3.75523679e-02 -4.70167547e-02 -1.94911305e-02  3.82631868e-02\n",
      " -5.67595847e-02 -1.78609311e-03  2.33404059e-02 -5.88216198e-33\n",
      " -4.87188101e-02 -2.76265759e-02 -3.38240378e-02  2.66188104e-02\n",
      " -3.39276865e-02 -8.49195104e-03 -1.91250294e-02  3.00252233e-02\n",
      "  3.40781994e-02  5.11157438e-02 -1.92479938e-02  2.85642371e-02\n",
      "  3.66040207e-02  1.68858673e-02  4.77258191e-02  1.23801976e-02\n",
      "  2.14844476e-02  4.93626227e-04  1.21273566e-02 -5.82144111e-02\n",
      "  1.62954051e-02 -7.14260805e-03  4.80091795e-02  2.51190793e-02\n",
      "  4.60097566e-02 -2.29840409e-02 -2.05696728e-02 -3.22231348e-03\n",
      "  4.00091410e-02  3.52310538e-02 -3.43153700e-02  2.75633647e-03\n",
      " -1.25138611e-02  1.97686236e-02  5.53487660e-03  1.03744566e-01\n",
      "  5.77611895e-03 -5.65426685e-02  4.19558808e-02 -3.78830060e-02\n",
      " -3.93443294e-02 -6.24309704e-02 -2.24398472e-03 -5.46549000e-02\n",
      "  4.56134900e-02 -5.69244847e-03  3.38916928e-02 -1.44447656e-02\n",
      "  2.72109639e-03  1.11191841e-02 -5.00661284e-02 -1.61127392e-02\n",
      "  1.72824156e-03  6.88877925e-02  1.16493087e-02  2.83171292e-02\n",
      "  6.97189383e-03  2.68372130e-02 -7.72084156e-03  2.16828585e-02\n",
      "  1.15182698e-02  8.72832313e-02 -6.27272902e-03 -6.44473433e-02\n",
      " -1.58234015e-02  4.03268002e-02 -1.69728603e-02 -1.61188859e-02\n",
      " -3.75576988e-02  7.02938437e-02 -3.30486484e-02  4.66324054e-02\n",
      "  1.18028000e-02  6.51075244e-02 -1.16979703e-02 -8.28343444e-03\n",
      " -5.46904914e-02 -2.00226866e-02  8.42690817e-04 -8.19524191e-03\n",
      "  2.08356604e-02  1.37454867e-02 -1.29924295e-03 -3.94575335e-02\n",
      " -2.00184640e-02 -1.53721338e-02  1.17271813e-02 -4.40111049e-02\n",
      "  5.39267696e-02 -2.33010836e-02 -2.24210918e-02 -3.65216564e-03\n",
      "  2.92213447e-02  7.56441429e-03 -2.90923137e-02  4.01517898e-02\n",
      " -2.00853441e-02 -1.79864350e-03 -1.26235886e-02  2.51076967e-02\n",
      " -4.69285510e-02 -3.08554564e-02 -3.63352097e-04  6.01790519e-03\n",
      "  3.97508964e-02  1.38548110e-02  2.49774400e-02  1.76976305e-02\n",
      " -9.31573212e-02 -9.83686745e-03  8.44923500e-03 -1.95391234e-02\n",
      " -3.26569304e-02  5.13739837e-03  5.80930850e-03  2.08536610e-02\n",
      " -5.97833563e-03  5.86810969e-02 -1.49496607e-02 -5.72965778e-02\n",
      " -5.98238036e-03  1.95207831e-03  2.72976072e-03  6.06995774e-03\n",
      " -2.00525392e-02 -1.31688034e-02 -4.06228825e-02  5.68997189e-02\n",
      "  4.44968417e-02 -1.24308197e-02  1.96967591e-02  3.80979218e-02\n",
      "  2.30237632e-07  1.10574570e-02  4.79512960e-02  6.18298426e-02\n",
      "  4.40278538e-02  6.17664680e-03  2.58290651e-03  3.38913836e-02\n",
      " -5.32939751e-03 -2.59283669e-02 -1.26143508e-02  2.46495418e-02\n",
      " -1.68774149e-03  1.17902865e-03  2.40443144e-02 -9.77309644e-02\n",
      "  1.97367519e-02 -5.52920401e-02 -6.17424622e-02 -4.87151593e-02\n",
      "  1.11085235e-03  1.18732050e-01  8.13258290e-02  3.32448892e-02\n",
      "  4.38327305e-02 -2.49559507e-02 -3.59626524e-02  1.66319385e-02\n",
      "  5.93767595e-03 -1.43972049e-02  4.46709478e-03 -6.01986311e-02\n",
      " -5.65911531e-02 -8.21539015e-03  5.83059806e-03 -1.69482231e-02\n",
      "  9.58633702e-03  1.46733886e-02  5.05845472e-02  3.06891371e-02\n",
      "  6.60468489e-02 -2.56551690e-02 -2.78858058e-02 -3.19173485e-02\n",
      " -3.39236297e-02  1.49903055e-02 -3.03336419e-02 -6.06493931e-03\n",
      " -4.81769675e-03  1.72137171e-02 -8.23371019e-03  1.55547913e-02\n",
      "  2.69106776e-02  5.44306403e-03 -1.06899561e-02 -7.82138295e-03\n",
      " -4.44506928e-02  2.55874228e-02 -5.74760474e-02 -2.05443185e-02\n",
      " -3.07849515e-02 -1.57855060e-02 -7.07539869e-03 -4.21312042e-02\n",
      "  3.79933938e-02  6.27765059e-02 -7.67791644e-03 -3.18353511e-02\n",
      "  1.99277746e-34  1.04834531e-02 -3.39326560e-02  3.93821336e-02\n",
      "  5.53065613e-02  9.42168385e-03  1.09728053e-02 -4.91939597e-02\n",
      "  2.95024291e-02 -8.85373075e-03 -5.96248396e-02 -2.37825699e-02]\n",
      "\n",
      "Sentence: Embeddings are one of the most powerful concepts in machine learning!\n",
      "Embedding: [-2.98611540e-02 -1.37522323e-02 -4.75402176e-02  2.72126570e-02\n",
      "  3.40053514e-02  3.16465981e-02  4.26963791e-02  3.29800532e-03\n",
      "  4.35718149e-02  2.53836978e-02  3.02528627e-02  3.21130864e-02\n",
      " -3.99912484e-02  1.28761707e-02  6.70219362e-02 -7.92899579e-02\n",
      "  4.68772352e-02  2.40266602e-02 -2.07997877e-02 -1.07433097e-02\n",
      " -1.19410949e-02 -5.39291203e-02  4.21055257e-02  2.23587938e-02\n",
      " -2.98949610e-02  8.35980009e-03  1.58385076e-02 -4.80236486e-02\n",
      "  1.88439421e-03 -1.67521797e-02 -2.15628725e-02 -3.88488062e-02\n",
      "  3.06273941e-02  4.20526005e-02  1.69483337e-06 -1.86928902e-02\n",
      " -1.24558611e-02  1.32129043e-02 -4.89039421e-02  1.34746004e-02\n",
      "  2.28873249e-02  8.81785527e-03  8.64927843e-03 -2.00950094e-02\n",
      " -3.15217525e-02 -2.53433064e-02  7.57318214e-02  3.62447202e-02\n",
      "  1.25290323e-02  3.09694968e-02  4.50759614e-03 -3.50042470e-02\n",
      " -4.42456309e-04 -9.76648834e-03  6.04545549e-02  4.03472260e-02\n",
      "  1.10735223e-02  6.56199548e-03 -5.84597606e-03  3.79787688e-03\n",
      " -4.46914881e-02  1.76405143e-02  2.45916862e-02 -3.60043673e-03\n",
      "  1.02473341e-01  3.73758636e-02  6.13313355e-03 -2.24677082e-02\n",
      "  1.46482559e-02  5.00536598e-02 -2.29907166e-02  1.12924855e-02\n",
      " -3.10552157e-02 -1.49509478e-02 -2.53130938e-03  3.20945457e-02\n",
      " -4.67055887e-02 -4.85886969e-02  2.98306365e-02  6.44215494e-02\n",
      " -3.12612914e-02  3.57406884e-02  4.16526720e-02 -5.52517399e-02\n",
      " -8.74624308e-03 -2.18630359e-02 -1.12744058e-02 -2.14436371e-02\n",
      " -1.32823838e-02 -2.04865858e-02 -1.00576431e-02  3.54763716e-02\n",
      " -7.47608952e-03 -3.70188057e-02  5.77893294e-02 -2.18168888e-02\n",
      "  4.36226744e-03  2.04380471e-02  3.36814784e-02 -4.92801033e-02\n",
      "  4.82793376e-02 -1.81003194e-03 -1.05118696e-02  4.13323380e-02\n",
      " -6.79834038e-02  1.75715946e-02 -4.43412513e-02  9.90841724e-03\n",
      " -3.81808840e-02  1.10827824e-02 -5.07280082e-02 -2.17451341e-02\n",
      " -1.03836432e-02  4.60333154e-02  1.55862430e-02 -4.21366841e-02\n",
      " -2.72146482e-02  3.22819091e-02 -4.24739458e-02  2.71207411e-02\n",
      " -7.41061419e-02  4.20107171e-02  2.02438049e-02  7.31811449e-02\n",
      " -8.97699222e-03 -2.31160652e-02 -3.93559001e-02 -1.46008870e-02\n",
      " -3.30910496e-02  1.12239728e-02  2.58570816e-03 -4.36851382e-03\n",
      "  1.85855441e-02  2.69934647e-02 -1.67215373e-02  3.69569585e-02\n",
      "  4.44489084e-02 -2.21723430e-02  6.72963215e-03  1.22935502e-02\n",
      "  1.71758514e-02 -2.36478797e-03  3.72263379e-02 -2.22870391e-02\n",
      "  2.94603202e-02 -2.33691446e-02  5.38467243e-03 -3.06581985e-02\n",
      " -2.38919929e-02 -2.63614580e-02 -2.01788880e-02  1.11245595e-01\n",
      " -1.99836325e-02 -3.54030505e-02  3.84143740e-02  2.53069364e-02\n",
      "  1.99551471e-02  5.53518124e-02 -1.99332666e-02 -2.16721976e-03\n",
      "  4.91092652e-02 -4.03530709e-02 -1.16977151e-02 -5.33113219e-02\n",
      "  8.29591695e-03 -5.08251749e-02 -2.65504196e-02 -1.53243048e-02\n",
      "  5.78817353e-03  2.46578502e-03 -3.44449952e-02 -1.85125961e-03\n",
      " -3.95730808e-02 -2.71690078e-02  4.93568443e-02  8.38369057e-02\n",
      "  5.43492287e-02  8.22261274e-02  1.23893926e-02 -4.79796575e-03\n",
      "  7.77415989e-04  2.98486874e-02 -1.85584258e-02  5.98795451e-02\n",
      " -6.82794582e-03  9.78147495e-04  2.85485201e-02 -7.64621701e-03\n",
      " -1.86619982e-02 -2.69287601e-02 -2.90332530e-02 -1.37871616e-02\n",
      " -2.57596583e-03 -2.20173765e-02 -1.70821454e-02 -3.81842777e-02\n",
      "  2.21505556e-02 -3.59233953e-02 -1.19439876e-02 -3.18307765e-02\n",
      " -4.80800718e-02  9.77633521e-03 -1.04856992e-03  4.15370911e-02\n",
      " -1.10974209e-02 -4.72830608e-02  1.90984476e-02 -5.31177036e-02\n",
      "  2.11325008e-02 -2.53068889e-03  5.61054610e-02 -1.33795543e-02\n",
      " -5.95864654e-03 -1.20307915e-02  4.63929772e-02 -2.81908680e-02\n",
      "  2.25355066e-02 -2.50473013e-03 -3.52452882e-02  2.55495347e-02\n",
      "  9.10396129e-03  3.25216330e-03  2.55970843e-03 -1.25624323e-02\n",
      " -3.51496674e-02 -4.28946465e-02 -2.32335529e-03  2.41020899e-02\n",
      " -5.16836671e-03  1.68740004e-02  5.52644301e-03  2.36791987e-02\n",
      "  5.65164685e-02 -3.47868837e-02 -6.34518042e-02 -7.45626492e-03\n",
      " -1.78448185e-02  5.35897464e-02  2.67291851e-02 -8.74199197e-02\n",
      "  1.04196286e-02 -4.13978618e-04 -3.04445368e-03  9.14243981e-03\n",
      "  2.91530285e-02 -5.81831485e-02  6.83463290e-02 -4.08618413e-02\n",
      " -9.09783226e-03 -3.40770409e-02  3.52411307e-02 -1.02628162e-02\n",
      " -5.72450226e-04 -2.73450441e-03  1.59635451e-02  4.49067401e-03\n",
      " -2.09051780e-02  3.02769560e-02  2.46119201e-02 -1.44067779e-02\n",
      "  1.73270069e-02  1.99041585e-03  4.23051938e-02 -2.39176359e-02\n",
      " -3.25547345e-02 -1.45940101e-02  3.95100564e-02 -6.04650006e-02\n",
      " -3.02065555e-02  1.67189874e-02 -2.26816889e-02 -2.61954814e-02\n",
      " -5.51319532e-02  1.44908335e-02 -1.99246220e-02  3.99749540e-03\n",
      "  3.12610194e-02 -4.90727536e-02 -9.49759793e-04  5.39497435e-02\n",
      " -9.10082087e-03 -2.69486047e-02 -3.63158509e-02 -1.38436714e-02\n",
      " -4.45622057e-02  5.49359173e-02  2.17594346e-03  2.23440886e-03\n",
      " -5.23012038e-03 -1.47894006e-02  3.60592045e-02  1.45263579e-02\n",
      "  8.39194190e-03 -6.10360280e-02 -7.89954420e-03 -2.98293447e-03\n",
      "  3.56565928e-03  8.33991915e-02 -2.61214972e-02  8.06722045e-02\n",
      "  3.63053894e-03  1.69974659e-02  2.58604158e-02  1.09443231e-03\n",
      " -4.57063615e-02  5.55678792e-02  2.00643893e-02  4.76661623e-02\n",
      " -4.91053015e-02 -1.86081380e-02  3.34404893e-02 -2.57310439e-02\n",
      " -3.16365645e-03  7.21444190e-02 -1.61519460e-02 -1.33933146e-02\n",
      " -6.06294274e-02 -2.82187220e-02 -8.91929585e-03 -2.71174102e-03\n",
      "  8.04916769e-03 -4.95209470e-02  7.89434686e-02  2.76428182e-02\n",
      " -5.42579871e-03 -3.06725455e-03 -4.11826670e-02  1.39172012e-02\n",
      "  3.04252934e-02  1.02856290e-02  1.06679024e-02 -5.56554422e-02\n",
      " -1.75083447e-02  2.03868542e-02  8.43308587e-03  3.82471308e-02\n",
      " -3.89100052e-02 -1.61303822e-02  3.18059921e-02 -7.32969865e-02\n",
      " -1.76502559e-02 -4.79873940e-02 -5.55042140e-02 -5.00750728e-03\n",
      "  4.46838239e-04  3.57333235e-02 -8.24668794e-04 -3.34324203e-02\n",
      " -3.32416631e-02 -2.46460326e-02  2.15332247e-02  3.90861137e-03\n",
      "  2.53471322e-02  6.02994766e-03 -7.81608373e-03  1.23765804e-02\n",
      " -1.71039402e-02  2.68102568e-02  2.83671659e-03 -1.27643254e-02\n",
      "  1.00510888e-01  1.03580467e-02 -3.55143249e-02  1.56616271e-02\n",
      " -9.85949188e-02  4.58441861e-02 -3.15230712e-02 -2.35781446e-02\n",
      " -2.78350580e-02 -7.75919543e-05 -2.82363184e-02 -1.92918703e-02\n",
      "  1.87390093e-02  5.71940914e-02  2.56911963e-02 -3.20030227e-02\n",
      "  1.99074093e-02 -3.15820314e-02 -4.02061976e-02  5.77630289e-02\n",
      "  1.72974151e-02 -5.37014157e-02 -1.25325499e-02 -1.45484367e-02\n",
      " -5.76174855e-02  1.09727802e-02 -2.04728320e-02  2.85541154e-02\n",
      " -5.04399575e-02  4.36991155e-02  1.75711084e-02 -1.02344435e-02\n",
      " -9.69771966e-02 -2.99995095e-02 -2.86679063e-02  2.24936660e-02\n",
      " -1.68121178e-02 -1.43673345e-02 -8.79589282e-03 -1.69043858e-02\n",
      "  2.41558515e-02 -6.53192773e-02 -4.10799906e-02 -2.34059468e-02\n",
      " -6.76065087e-02 -1.55691030e-02  3.62357907e-02  7.83160254e-02\n",
      " -4.97516990e-02 -7.08548352e-02 -5.01178876e-02 -8.56449304e-04\n",
      "  5.44900540e-03  4.34697000e-03  9.88053009e-02 -2.16415208e-02\n",
      " -1.87750943e-02  1.15069682e-02  2.63995826e-02  1.65236164e-02\n",
      " -2.24057995e-02 -4.31826897e-02  1.31803930e-01 -2.97034830e-02\n",
      "  2.65934318e-02 -1.38888396e-02 -1.67003833e-02  3.44146229e-02\n",
      " -8.94359313e-03  6.16001152e-02 -3.42303254e-02  2.46417988e-03\n",
      " -8.14089272e-03  5.80325760e-02  5.24208173e-02 -1.53281195e-02\n",
      "  4.01382446e-02  1.51407914e-02 -3.01468908e-03 -4.97021563e-02\n",
      " -4.24300041e-03  5.77288792e-02  3.17874067e-02  4.74008173e-02\n",
      "  2.95218267e-02 -1.50122903e-02 -2.47945655e-02 -7.11502358e-02\n",
      "  2.06847563e-02  3.11487596e-02 -5.86069096e-03  1.62786972e-02\n",
      " -3.93682420e-02  5.46505563e-02  3.26594561e-02 -1.87021513e-02\n",
      " -9.79863405e-02  4.33767028e-03 -5.58157526e-02 -1.34620816e-02\n",
      "  2.88453642e-02  1.58748161e-02 -3.32564600e-02  1.44416827e-03\n",
      " -5.51107824e-02  8.24674293e-02  2.38845535e-02 -2.04837713e-02\n",
      " -4.78589628e-03  3.78722958e-02 -4.87562791e-02  3.44646759e-02\n",
      "  1.10358065e-02  1.12450607e-02  1.33262575e-02 -3.46375592e-02\n",
      " -6.92220628e-02  7.30128866e-03 -6.57013804e-03  1.73203647e-02\n",
      "  5.23056416e-03  4.48132269e-02  3.89852971e-02 -1.99274123e-02\n",
      " -1.80920642e-02  3.25937606e-02 -2.02026349e-02  4.86182078e-04\n",
      " -8.88753589e-03 -1.91348325e-02  2.50685569e-02  4.74020429e-02\n",
      "  2.18656985e-03 -1.69988144e-02  3.62671688e-02  3.46246175e-03\n",
      "  4.21938859e-03  8.04172307e-02  3.10627781e-02 -1.04944163e-03\n",
      " -3.55466716e-02  4.34836857e-02 -3.06218583e-02 -3.03192325e-02\n",
      " -4.13175002e-02 -1.05257872e-02 -2.35242844e-02 -1.86771620e-02\n",
      "  4.42936504e-03  5.45056351e-02 -6.05017617e-02  2.48421710e-02\n",
      " -3.36966105e-02 -4.54169102e-02 -2.63173319e-02  6.98053604e-03\n",
      "  6.92871660e-02 -2.04490907e-02 -1.96813364e-02 -9.72559303e-03\n",
      " -1.21564707e-02  7.89344218e-03  1.84744364e-03 -6.93651140e-02\n",
      "  2.43357420e-02  4.00609262e-02  3.44013013e-02 -2.84282006e-02\n",
      " -1.09431241e-02  1.38742346e-02 -4.40587196e-03  1.19344890e-03\n",
      " -8.81164819e-02  1.15931015e-02 -2.56351158e-02  5.57525977e-02\n",
      "  1.26946166e-01  5.39565869e-02 -1.41436616e-02  1.27196051e-02\n",
      " -1.32236090e-02 -5.94484396e-02  2.86704805e-02  2.57284846e-02\n",
      " -8.33776500e-03  8.17395397e-04  5.93052059e-03  3.29110026e-02\n",
      "  4.12751995e-02 -5.77960489e-03 -1.71124507e-02  1.06227612e-02\n",
      " -2.19601188e-02 -4.97206785e-02  2.53765676e-02 -5.60257469e-33\n",
      " -1.35397762e-02 -3.77959013e-02 -2.67923134e-03 -3.69688350e-04\n",
      " -1.98268034e-02  5.47049521e-03  6.02681655e-03  1.93068981e-02\n",
      "  3.87986237e-03  2.97698602e-02 -1.88228711e-02  2.20037089e-03\n",
      "  8.17020051e-03  1.61965154e-02  3.17528732e-02 -6.83421222e-03\n",
      "  2.19252408e-02  4.37722629e-04  2.96859033e-02 -2.62557156e-02\n",
      "  6.49381056e-03  3.56025212e-02  1.58610509e-03 -4.76583727e-02\n",
      " -5.26238680e-02  3.78274396e-02  3.54341865e-02 -3.10347937e-02\n",
      "  7.96405040e-03  5.48469275e-02 -3.56443189e-02  9.10129771e-03\n",
      " -9.45987180e-03 -4.63287011e-02 -1.63906273e-02  6.32453188e-02\n",
      " -1.38588436e-02 -5.95724657e-02 -1.57990754e-02  2.01887470e-02\n",
      " -1.98293440e-02 -3.49211358e-02  2.27937829e-02 -5.91621585e-02\n",
      "  4.18853872e-02  1.20745995e-03  5.19158505e-02 -1.88436024e-02\n",
      " -3.12101617e-02  2.34933067e-02 -7.41029978e-02 -2.76626030e-04\n",
      " -1.51719851e-02  6.11714050e-02  1.25065118e-01 -1.28459232e-02\n",
      " -1.12671368e-02  1.51749374e-03 -8.09153318e-02  1.12689324e-02\n",
      " -1.97573993e-02  2.74268407e-02  9.40988213e-03 -9.58756916e-03\n",
      "  2.54850667e-02  6.81658834e-02 -1.83452908e-02 -1.00963905e-01\n",
      " -9.45250783e-03 -5.27006388e-03  1.98684130e-02  9.80847701e-02\n",
      "  3.15632969e-02  5.30422777e-02  3.75122987e-02 -6.64208978e-02\n",
      " -5.92880100e-02 -1.57073587e-02  1.76609233e-02 -5.81073500e-02\n",
      "  2.23231055e-02  1.29869655e-02 -3.30260880e-02  9.96864168e-04\n",
      " -9.87090077e-03 -3.12955119e-02  2.28524162e-03 -4.91250008e-02\n",
      "  1.47693455e-02 -1.83367524e-02 -4.16305214e-02  3.76897343e-02\n",
      "  3.35410014e-02 -7.97111318e-02  4.01300266e-02  1.59071740e-02\n",
      "  5.06086089e-03  4.28808592e-02  2.29759719e-02 -4.13351394e-02\n",
      " -3.10502909e-02 -5.26405275e-02 -4.95404862e-02 -2.94256527e-02\n",
      "  5.94923906e-02 -2.59802993e-02  3.02497130e-02  8.80412105e-03\n",
      " -4.84466814e-02 -2.00851355e-02  9.82174464e-03 -7.89194033e-02\n",
      "  4.52885916e-03 -9.34896618e-03  9.23311897e-03 -3.17362584e-02\n",
      "  2.10833009e-02  6.37122104e-03  3.36347856e-02  3.83613780e-02\n",
      " -4.55527902e-02  1.08118646e-03 -9.83322971e-03  7.70189334e-03\n",
      " -2.87617669e-02 -1.74959581e-02 -4.27810894e-03  2.81287339e-02\n",
      "  4.97339256e-02 -7.45570883e-02 -1.07008684e-02 -7.66056404e-03\n",
      "  2.33968805e-07  1.52482847e-02  8.39614570e-02  3.67242955e-02\n",
      " -3.69249210e-02  3.64752188e-02  4.26421948e-02 -4.39827237e-03\n",
      "  1.78133342e-02 -2.67076902e-02 -7.13897031e-03  5.59976064e-02\n",
      "  3.13965976e-02  2.13439134e-03  3.90370861e-02 -8.78528580e-02\n",
      " -2.21660305e-02 -2.47735754e-02 -1.18190460e-02 -7.89703522e-03\n",
      " -2.08857208e-02  4.30556089e-02  1.07643485e-01  4.40618806e-02\n",
      "  1.47963297e-02  2.44862624e-02 -3.86268422e-02  1.80743560e-02\n",
      " -1.47838600e-03  7.74167031e-02 -4.19565625e-02 -3.80529240e-02\n",
      "  3.61258015e-02  1.59039337e-03  1.95323937e-02 -2.00080313e-02\n",
      "  4.22539040e-02  3.06111891e-02 -3.53700249e-03  5.93106961e-03\n",
      " -2.23223194e-02 -2.07131319e-02 -3.62915429e-03  1.74654070e-02\n",
      " -4.08758670e-02  5.91594689e-02 -5.89099154e-02 -3.96754369e-02\n",
      " -3.33529152e-02  1.02161942e-02  6.97289547e-03  7.70390108e-02\n",
      " -1.86911523e-02 -1.82568394e-02 -2.42319256e-02 -3.40700964e-03\n",
      " -3.60554792e-02  4.33389246e-02 -3.48602682e-02  5.27768061e-02\n",
      "  2.89710276e-02 -4.98462655e-02 -1.94749087e-02  1.16398381e-02\n",
      " -3.04614659e-02  8.04637671e-02  6.56247959e-02 -2.84533072e-02\n",
      "  1.81615243e-34 -4.19154484e-03 -2.57882699e-02  5.17319962e-02\n",
      "  4.94420938e-02  1.32476222e-02 -4.21993732e-02 -1.12458626e-02\n",
      " -2.61520278e-02  5.51130548e-02  2.20025014e-02 -2.51170527e-02]\n",
      "\n",
      "Sentence: Learn to use embeddings well and you'll be well on your way to being an AI engineer.\n",
      "Embedding: [-2.20731031e-02  2.08950527e-02 -6.03005737e-02  8.43948685e-03\n",
      "  4.37650830e-02  1.55070620e-02  4.99907285e-02 -3.03232297e-02\n",
      "  4.94784079e-02  2.35511530e-02  3.29350382e-02  1.53877586e-02\n",
      " -6.68355078e-02  1.11002855e-01  6.92676529e-02 -2.31889170e-02\n",
      "  3.79103199e-02 -4.94145043e-03 -1.57801062e-02 -3.45476046e-02\n",
      " -2.65053082e-02 -2.47879438e-02 -1.86141394e-02  3.00361775e-02\n",
      " -2.81186551e-02 -8.75137653e-03 -3.30770179e-03 -2.06116457e-02\n",
      "  1.03315525e-02 -1.51483566e-02 -3.48330848e-02 -2.63248067e-02\n",
      "  2.06907894e-02  3.79109606e-02  1.81912844e-06 -2.44292011e-03\n",
      " -1.80561643e-03  5.61762787e-03 -2.79869884e-02  1.54703120e-02\n",
      "  3.06456797e-02  3.72601300e-02 -1.55611532e-02  2.54414305e-02\n",
      " -6.42072037e-02  3.16352956e-02  6.63442314e-02  3.80970538e-02\n",
      "  5.57844937e-02  5.31660318e-02 -9.69293900e-03 -3.61424573e-02\n",
      "  3.72434929e-02 -4.67827544e-03  5.14575616e-02  1.00057842e-02\n",
      "  4.90290113e-03  1.41562950e-02  4.95099649e-02  3.32948659e-03\n",
      " -3.21100764e-02  4.42387462e-02  3.27415839e-02 -7.90622458e-03\n",
      "  1.07809760e-01  7.32944459e-02  3.36702429e-02 -4.28345613e-02\n",
      "  1.05966358e-02  2.05654074e-02 -2.02670284e-02  1.04964105e-02\n",
      " -1.97615586e-02 -2.89186173e-05 -2.61862371e-02 -1.85174160e-02\n",
      " -3.44269313e-02 -4.08620648e-02  2.32571829e-02  2.14195848e-02\n",
      "  1.31321335e-02 -3.27211246e-02 -1.91425961e-02 -2.86572129e-02\n",
      " -1.16859265e-02  1.21910656e-02  1.05248820e-02 -3.39584760e-02\n",
      "  3.08910361e-03 -4.44888659e-02  2.65104696e-02  1.09536611e-02\n",
      "  2.51450613e-02 -6.48831669e-03  4.54362435e-03 -2.02784836e-02\n",
      " -1.03216004e-02  2.06589941e-02 -1.65313706e-02 -2.45612208e-02\n",
      "  5.47549389e-02  2.68261675e-02  2.95510236e-02  3.86755131e-02\n",
      " -7.76720941e-02  3.80055867e-02 -2.98364274e-02  7.96886086e-02\n",
      " -3.00942697e-02  7.57844001e-03 -6.89826980e-02 -2.92667076e-02\n",
      " -2.35580280e-02  3.48198451e-02  2.52938624e-02 -4.53817137e-02\n",
      " -1.57939047e-02  4.39031236e-02 -4.04335670e-02  8.32530297e-03\n",
      " -2.84664556e-02  4.94933389e-02  2.41277423e-02  3.02192569e-02\n",
      " -4.99590486e-02 -5.94532900e-02 -3.70175950e-02  1.30330259e-02\n",
      " -3.36469188e-02  3.45589928e-02 -1.44523084e-02  2.57640500e-02\n",
      "  4.61182185e-03  2.21551638e-02 -4.93461313e-03  9.66005102e-02\n",
      " -2.72447779e-03  5.65359311e-04 -3.24248821e-02  1.31681599e-02\n",
      "  4.41607349e-02 -7.03057228e-03  6.84260875e-02 -2.28166599e-02\n",
      " -2.81035085e-03 -4.23883051e-02 -1.33632021e-02 -5.96738309e-02\n",
      " -6.96124230e-03 -2.31900997e-02 -3.78850810e-02  9.80186760e-02\n",
      " -2.21728776e-02 -2.30062902e-02  3.22815031e-02  8.21806677e-03\n",
      " -7.04120379e-03  4.84079309e-02  4.23291661e-02 -2.59715784e-03\n",
      "  1.20521116e-04  1.67414192e-02  2.91197523e-02 -1.28740883e-02\n",
      " -2.41077933e-02 -3.29320654e-02 -3.50293610e-03 -3.19321975e-02\n",
      " -2.64171772e-02  2.30404250e-02  1.11637199e-02 -9.95995849e-03\n",
      " -1.75901111e-02 -2.00278778e-03  1.21594677e-02  4.67823371e-02\n",
      "  5.20882122e-02  7.21171945e-02  1.94978211e-02  1.15072569e-02\n",
      "  5.94160846e-03 -1.47834318e-02 -2.87724007e-02  6.72666728e-02\n",
      " -1.68758705e-02  5.25877625e-03 -3.39739248e-02  5.24596311e-02\n",
      " -2.59793065e-02 -4.41379957e-02  1.47451565e-03 -1.06599228e-02\n",
      " -1.51859652e-02 -1.55881001e-03  1.81505028e-02 -4.85410914e-02\n",
      "  3.67901451e-03 -6.59313127e-02 -1.49418414e-02 -3.23529802e-02\n",
      " -2.79949158e-02  1.71856843e-02 -7.87081476e-03  4.65692207e-02\n",
      "  1.47123523e-02 -7.40439147e-02 -6.52104393e-02 -5.22735007e-02\n",
      " -1.82343647e-02  5.20859174e-02  3.06304321e-02 -2.36037560e-02\n",
      "  2.42384523e-02 -1.83940213e-02 -4.83012944e-03 -2.13386286e-02\n",
      "  1.56584140e-02  9.87344142e-03 -4.25561443e-02  6.00458262e-03\n",
      " -3.14497389e-03  4.51519294e-03 -1.52778265e-03  1.13731530e-02\n",
      " -6.96353689e-02 -3.37258130e-02  1.33406902e-02  4.87286597e-03\n",
      " -7.81486556e-03  4.78049330e-02 -1.59711689e-02  3.14606316e-02\n",
      "  5.15920669e-02 -4.05121781e-02 -5.06460816e-02  9.99934971e-03\n",
      " -2.00729948e-02  4.21552286e-02  3.03182919e-02 -1.00431219e-01\n",
      " -4.12020162e-02  3.43990773e-02  3.29209454e-02  1.07407570e-03\n",
      "  3.70962396e-02 -6.94237128e-02  6.52393252e-02  8.31662398e-03\n",
      "  1.68036465e-02 -2.60706041e-02  8.14501662e-03 -1.48009304e-02\n",
      " -2.65670791e-02  3.29322293e-02 -7.37514719e-03  7.23976083e-03\n",
      " -2.69330107e-02  1.71755478e-02 -2.28083245e-02 -4.75342385e-03\n",
      "  2.88568959e-02  1.30786080e-04  5.44129014e-02 -1.43378936e-02\n",
      "  1.89891960e-02 -1.32732149e-02  4.01177891e-02 -7.29275197e-02\n",
      " -2.41210721e-02  3.16217281e-02 -1.68014467e-02  8.47543869e-03\n",
      " -5.23940772e-02 -1.43882809e-02 -1.46156447e-02  6.39910018e-03\n",
      "  2.15113591e-02 -5.18960431e-02 -4.30576541e-02  2.34075766e-02\n",
      "  2.30272207e-03 -2.48434264e-02 -4.38243300e-02 -2.16570478e-02\n",
      " -6.91595599e-02  1.76770389e-02  2.12066770e-02 -2.20294278e-02\n",
      " -1.06772976e-02  9.57422983e-03  2.21988875e-02  5.51471114e-02\n",
      "  1.03683164e-02 -8.14824402e-02 -7.94706214e-03 -1.85866505e-02\n",
      "  1.20494403e-02  7.51402974e-02 -1.41215529e-02  8.83924663e-02\n",
      "  3.12628672e-02  8.12258944e-03 -2.29444169e-02  3.96010801e-02\n",
      " -2.00168006e-02  9.16027650e-02 -2.06838828e-02  5.84991239e-02\n",
      " -4.32368778e-02 -4.74743405e-03 -9.51888692e-03  5.42947603e-03\n",
      "  1.19169781e-04  6.15376495e-02 -1.68788515e-03 -4.66320701e-02\n",
      " -2.01405194e-02  1.32406102e-02  9.70682688e-03  2.73847505e-02\n",
      "  3.06639764e-02  6.71585556e-03  8.71220976e-02 -1.97370234e-03\n",
      "  8.18298105e-03  5.44358604e-03 -5.76205775e-02  1.34820631e-02\n",
      "  5.06293308e-03 -2.10568979e-02  1.25936856e-02 -5.49783697e-03\n",
      " -1.44645562e-02 -2.92567275e-02  5.53298891e-02 -2.60099135e-02\n",
      " -2.82453932e-03 -2.30901688e-02  8.89706332e-03 -2.61565465e-02\n",
      "  9.08626884e-04 -6.16203807e-02 -7.56419748e-02 -1.05931740e-02\n",
      " -1.19563667e-02  6.71458542e-02 -1.96234938e-02 -5.00934087e-02\n",
      " -3.91228981e-02 -3.07008084e-02  7.18906671e-02  9.29247309e-03\n",
      " -6.34385180e-03  7.86995108e-04 -1.36484122e-02  2.87189130e-02\n",
      "  4.01224680e-02  1.28037073e-02  1.77381374e-02 -4.75977641e-03\n",
      "  5.47173433e-02  1.10809703e-03 -2.25790311e-02 -2.80286768e-03\n",
      " -1.13696054e-01  2.55904589e-02  4.00429562e-04 -4.39811572e-02\n",
      "  1.36319557e-02 -1.54137397e-02 -4.99015301e-02 -2.32889857e-02\n",
      " -1.62989099e-03  3.95835340e-02  1.89040564e-02 -3.02703176e-02\n",
      "  2.71438118e-02  1.05678255e-03 -4.21069711e-02  3.71961184e-02\n",
      "  3.54258493e-02 -6.98274523e-02 -2.20937543e-02 -4.01496403e-02\n",
      " -1.90164484e-02 -2.69835405e-02 -1.51213091e-02  3.33365388e-02\n",
      " -9.74889472e-02  1.73102263e-02  6.21017301e-03 -2.59426283e-03\n",
      " -1.10152930e-01 -6.10186234e-02 -1.36549659e-02 -1.35035217e-02\n",
      " -6.72573373e-02 -4.05121082e-03 -6.64983504e-03  3.86566296e-03\n",
      "  9.43470094e-03 -3.86636890e-02 -1.93593018e-02  1.34934196e-02\n",
      " -4.58106548e-02  6.06737062e-02  6.06380105e-02  4.85596582e-02\n",
      " -4.56088372e-02 -5.71936443e-02 -1.55094508e-02  3.40962820e-02\n",
      "  9.48148838e-04 -9.94349830e-03  2.84657478e-02 -3.29023860e-02\n",
      " -2.83211209e-02  3.19907628e-02  2.61298809e-02 -2.74054948e-02\n",
      " -1.36353280e-02  7.47715216e-03  1.19430304e-01 -4.45807166e-02\n",
      "  1.07671125e-02 -8.69424120e-02 -2.19550952e-02  1.83874927e-02\n",
      " -1.06521463e-02 -1.89243164e-02 -3.06513775e-02 -3.04701924e-02\n",
      " -3.22214700e-02  4.12150994e-02  8.95752106e-03 -2.73179971e-02\n",
      "  9.39992908e-03 -9.57092620e-04 -1.94010064e-02 -4.92623225e-02\n",
      " -9.18888394e-03  4.66894247e-02  5.41892648e-02  2.21609510e-02\n",
      " -2.86352094e-02  5.20295613e-02  2.47589145e-02 -7.14268908e-02\n",
      " -1.26206400e-02  7.35520339e-03  2.13784464e-02  2.93514077e-02\n",
      " -2.57651359e-02  5.20562604e-02 -2.74891984e-02 -3.10242716e-02\n",
      " -9.02879611e-02  6.10371344e-02 -5.22609763e-02  2.13111620e-02\n",
      "  4.41735089e-02  3.23370919e-02  1.75647810e-02 -2.39520241e-02\n",
      " -2.69709323e-02  5.11277989e-02  2.69065574e-02 -4.51932847e-02\n",
      "  2.52654497e-03  2.44933851e-02 -2.89541110e-02  2.79992856e-02\n",
      " -1.36022680e-02 -4.32368852e-02  1.85830314e-02  7.63770513e-05\n",
      "  2.43514776e-03 -3.73322191e-03 -1.72279738e-02  1.01292785e-02\n",
      "  1.98437236e-02 -2.60018148e-02 -3.40176700e-03  1.09125590e-02\n",
      " -4.16363552e-02  3.37031819e-02 -2.81635448e-02  1.79126225e-02\n",
      " -4.53095064e-02 -1.09819658e-02 -2.20823847e-03  1.99102815e-02\n",
      "  3.56371738e-02 -3.11799347e-02  3.78752127e-02 -1.41409431e-02\n",
      " -2.16907002e-02  2.73019746e-02  3.69820371e-03  6.35388121e-02\n",
      "  1.22669209e-02 -6.02275692e-03 -7.60758435e-03 -1.86565276e-02\n",
      " -5.64718712e-03 -2.20049103e-03 -1.31825237e-02  1.67724546e-02\n",
      " -3.77264768e-02  2.97895204e-02 -5.01569659e-02  4.89087887e-02\n",
      " -6.07444346e-02 -8.39418471e-02 -5.09001575e-02  1.81768164e-02\n",
      "  6.66732565e-02 -3.30037414e-03 -2.82401126e-03 -5.35406992e-02\n",
      "  3.90341394e-02  2.19852682e-02  3.13555412e-02 -3.36527117e-02\n",
      "  1.96913499e-02  1.67883858e-02  5.04003130e-02  3.08058062e-03\n",
      " -7.24742364e-04  4.42906879e-02 -4.12958860e-03  4.29328606e-02\n",
      " -6.62551820e-02  1.16055354e-03 -2.81716697e-02  1.56886112e-02\n",
      "  9.78133157e-02  5.53594455e-02 -1.39379166e-02  2.12307051e-02\n",
      " -1.30956117e-02 -6.82027787e-02 -8.09052959e-04  4.99292277e-02\n",
      " -2.69265529e-02 -2.97803804e-02  3.84462029e-02  1.97353829e-02\n",
      "  3.37088183e-02  1.65873747e-02  5.77316107e-03 -3.04897558e-02\n",
      " -1.52511587e-02 -3.56159471e-02 -8.69220216e-03 -5.42296760e-33\n",
      "  3.24370596e-03 -3.46329845e-02  3.58933173e-02  1.83771215e-02\n",
      " -2.17504762e-02 -3.26411724e-02  2.88396119e-03  1.50463674e-02\n",
      " -1.75259542e-03 -1.99419018e-02 -6.10349979e-03  2.23847236e-02\n",
      " -8.78941035e-04  2.48684920e-02  3.39736417e-02  2.75592860e-02\n",
      "  3.37792300e-02  3.98565084e-02  2.55545564e-02  1.83042437e-02\n",
      " -2.92878319e-02  5.18082036e-03  8.37854692e-04 -3.66560519e-02\n",
      " -3.46733071e-02  3.82687338e-02  5.50821982e-03 -4.35187854e-02\n",
      "  2.44077481e-02  3.54167521e-02 -2.13442538e-02  2.86623668e-02\n",
      " -2.65408307e-04  3.73409651e-02 -8.68166890e-03  3.04786279e-03\n",
      " -2.71682478e-02 -3.85088287e-02 -6.12388551e-02 -2.00839597e-03\n",
      " -1.22080445e-02 -8.67198110e-02  3.75362486e-03 -1.77707430e-02\n",
      "  8.32482427e-03 -1.69167407e-02  7.02404082e-02  3.32234129e-02\n",
      "  4.34313193e-02  1.47016924e-02 -1.25546932e-01  1.50866564e-02\n",
      " -5.43164425e-02 -1.79151597e-03  4.99600582e-02 -1.53786251e-02\n",
      "  3.32683139e-02 -3.07708886e-02 -1.83896422e-02  9.45800729e-03\n",
      " -4.60291281e-02 -2.03871657e-03  2.62428522e-02 -5.00788875e-02\n",
      "  2.01835949e-02  6.08983263e-02 -2.01180577e-02 -2.60053631e-02\n",
      "  1.05925361e-02 -3.31153385e-02  1.62595250e-02  7.77862370e-02\n",
      " -1.90731674e-03 -5.62889222e-03  1.43716186e-02 -4.06833254e-02\n",
      " -5.14971316e-02  1.66211539e-04 -3.33050336e-03  1.44688813e-02\n",
      "  4.24889586e-04  3.04453187e-02 -1.83636341e-02  1.51192432e-03\n",
      "  2.99861599e-02 -3.68002281e-02  8.35624523e-03 -3.31025422e-02\n",
      "  2.66911611e-02  5.47829876e-03 -1.80523917e-02  2.42577940e-02\n",
      "  5.72708249e-03 -5.93372062e-02  1.04358420e-01 -9.87924263e-03\n",
      " -1.36105977e-02  5.79998828e-02  2.50108466e-02  2.89337169e-02\n",
      " -3.20520923e-02 -3.40233147e-02 -3.41698565e-02 -2.76982002e-02\n",
      "  6.47003055e-02  1.50798066e-02 -1.61925219e-02  3.03266346e-02\n",
      " -2.67187804e-02 -3.67774591e-02 -2.27845572e-02 -5.36433943e-02\n",
      "  1.90499835e-02 -3.42503116e-02  1.32688330e-02 -5.41316066e-03\n",
      "  7.49741215e-03 -7.36983551e-04 -3.08569279e-02  3.82288136e-02\n",
      " -2.08311304e-02 -3.43154855e-02  5.60239702e-03  1.44999269e-02\n",
      " -3.76363993e-02 -5.11783138e-02 -3.51075083e-02  1.71867777e-02\n",
      "  1.50721176e-02 -9.62026715e-02 -1.53544880e-02  1.58376191e-02\n",
      "  2.42940899e-07 -5.88801829e-03  7.68796504e-02  5.86063303e-02\n",
      "  2.21232977e-02 -2.36690789e-02  5.25273792e-02  1.48661574e-02\n",
      "  7.34175881e-03 -4.98924730e-03  4.37413827e-02 -1.28331883e-02\n",
      "  3.37342098e-02 -1.10815018e-02 -1.33937970e-02 -7.80064017e-02\n",
      " -1.36330789e-02  1.94749255e-02  1.91748980e-03 -3.00251935e-02\n",
      "  1.02723170e-04  9.54532996e-02  1.19653992e-01  3.73372212e-02\n",
      "  4.25128872e-03  2.05130037e-02 -3.85414697e-02 -1.90614443e-02\n",
      "  5.88793308e-02  6.81264400e-02 -3.12595554e-02 -6.50440678e-02\n",
      "  2.48044208e-02  3.90102854e-04  7.54762441e-02 -3.46075855e-02\n",
      "  1.32949268e-02  4.14005481e-02  3.07568759e-02  5.50353760e-03\n",
      " -1.53086276e-03  2.75993571e-02  6.46030856e-03  1.05398288e-02\n",
      " -3.09298523e-02  4.60232273e-02 -3.64920571e-02 -1.39539819e-02\n",
      " -3.53720672e-02  7.97806599e-04  1.40632587e-02  1.80258360e-02\n",
      " -1.43368505e-02  2.19215825e-03 -3.96873429e-02 -1.17282281e-02\n",
      " -4.45219912e-02  8.05770978e-03 -4.04861942e-02  3.56148705e-02\n",
      "  5.12852818e-02 -6.64038733e-02 -5.32594919e-02  8.92899092e-03\n",
      "  1.56423915e-02  1.02110982e-01  8.10768176e-03 -4.03858768e-03\n",
      "  2.02352605e-34 -1.38293868e-02 -1.17623834e-02  1.51005890e-02\n",
      "  8.25895891e-02  2.39228476e-02 -1.10378005e-02  3.65658151e-03\n",
      " -7.44785555e-03  2.94555090e-02  3.53000569e-03 -6.10421933e-02]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Requires !pip install sentence-transformers\n",
    "from sentence_transformers import SentenceTransformer\n",
    "embedding_model = SentenceTransformer(model_name_or_path=\"all-mpnet-base-v2\", \n",
    "                                      device=\"cpu\") # choose the device to load the model to (note: GPU will often be *much* faster than CPU)\n",
    "\n",
    "# Create a list of sentences to turn into numbers\n",
    "sentences = [\n",
    "    \"The Sentences Transformers library provides an easy and open-source way to create embeddings.\",\n",
    "    \"Sentences can be embedded one by one or as a list of strings.\",\n",
    "    \"Embeddings are one of the most powerful concepts in machine learning!\",\n",
    "    \"Learn to use embeddings well and you'll be well on your way to being an AI engineer.\"\n",
    "]\n",
    "\n",
    "# Sentences are encoded/embedded by calling model.encode()\n",
    "embeddings = embedding_model.encode(sentences)\n",
    "embeddings_dict = dict(zip(sentences, embeddings))\n",
    "\n",
    "# See the embeddings\n",
    "for sentence, embedding in embeddings_dict.items():\n",
    "    print(\"Sentence:\", sentence)\n",
    "    print(\"Embedding:\", embedding)\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: %%time is a cell magic, but the cell body is empty. Did you mean the line magic %time (single %)?\n"
     ]
    }
   ],
   "source": [
    "# %%time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import torch\n",
    "# torch.cuda.is_available()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "Torch not compiled with CUDA enabled",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "File \u001b[1;32m<timed eval>:2\u001b[0m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\nn\\modules\\module.py:1152\u001b[0m, in \u001b[0;36mModule.to\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1148\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m t\u001b[38;5;241m.\u001b[39mto(device, dtype \u001b[38;5;28;01mif\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_floating_point() \u001b[38;5;129;01mor\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_complex() \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   1149\u001b[0m                     non_blocking, memory_format\u001b[38;5;241m=\u001b[39mconvert_to_format)\n\u001b[0;32m   1150\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m t\u001b[38;5;241m.\u001b[39mto(device, dtype \u001b[38;5;28;01mif\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_floating_point() \u001b[38;5;129;01mor\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_complex() \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m, non_blocking)\n\u001b[1;32m-> 1152\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconvert\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\nn\\modules\\module.py:802\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    800\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    801\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 802\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    804\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    805\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    806\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    807\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    812\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    813\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\nn\\modules\\module.py:802\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    800\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    801\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 802\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    804\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    805\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    806\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    807\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    812\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    813\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "    \u001b[1;31m[... skipping similar frames: Module._apply at line 802 (1 times)]\u001b[0m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\nn\\modules\\module.py:802\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    800\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[0;32m    801\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[1;32m--> 802\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    804\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[0;32m    805\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[0;32m    806\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[0;32m    807\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    812\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[0;32m    813\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\nn\\modules\\module.py:825\u001b[0m, in \u001b[0;36mModule._apply\u001b[1;34m(self, fn, recurse)\u001b[0m\n\u001b[0;32m    821\u001b[0m \u001b[38;5;66;03m# Tensors stored in modules are graph leaves, and we don't want to\u001b[39;00m\n\u001b[0;32m    822\u001b[0m \u001b[38;5;66;03m# track autograd history of `param_applied`, so we have to use\u001b[39;00m\n\u001b[0;32m    823\u001b[0m \u001b[38;5;66;03m# `with torch.no_grad():`\u001b[39;00m\n\u001b[0;32m    824\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[1;32m--> 825\u001b[0m     param_applied \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparam\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    826\u001b[0m should_use_set_data \u001b[38;5;241m=\u001b[39m compute_should_use_set_data(param, param_applied)\n\u001b[0;32m    827\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m should_use_set_data:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\nn\\modules\\module.py:1150\u001b[0m, in \u001b[0;36mModule.to.<locals>.convert\u001b[1;34m(t)\u001b[0m\n\u001b[0;32m   1147\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m convert_to_format \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m t\u001b[38;5;241m.\u001b[39mdim() \u001b[38;5;129;01min\u001b[39;00m (\u001b[38;5;241m4\u001b[39m, \u001b[38;5;241m5\u001b[39m):\n\u001b[0;32m   1148\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m t\u001b[38;5;241m.\u001b[39mto(device, dtype \u001b[38;5;28;01mif\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_floating_point() \u001b[38;5;129;01mor\u001b[39;00m t\u001b[38;5;241m.\u001b[39mis_complex() \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   1149\u001b[0m                 non_blocking, memory_format\u001b[38;5;241m=\u001b[39mconvert_to_format)\n\u001b[1;32m-> 1150\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_floating_point\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mis_complex\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnon_blocking\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\torch\\cuda\\__init__.py:293\u001b[0m, in \u001b[0;36m_lazy_init\u001b[1;34m()\u001b[0m\n\u001b[0;32m    288\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[0;32m    289\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot re-initialize CUDA in forked subprocess. To use CUDA with \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    290\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmultiprocessing, you must use the \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mspawn\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m start method\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    291\u001b[0m     )\n\u001b[0;32m    292\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(torch\u001b[38;5;241m.\u001b[39m_C, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_cuda_getDeviceCount\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m--> 293\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTorch not compiled with CUDA enabled\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    294\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m _cudart \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    295\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAssertionError\u001b[39;00m(\n\u001b[0;32m    296\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlibcudart functions unavailable. It looks like you have a broken build?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    297\u001b[0m     )\n",
      "\u001b[1;31mAssertionError\u001b[0m: Torch not compiled with CUDA enabled"
     ]
    }
   ],
   "source": [
    "# %%time\n",
    "\n",
    "# # Send the model to the GPU\n",
    "# embedding_model.to(\"cuda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 115/115 [00:22<00:00,  5.17it/s]\n"
     ]
    }
   ],
   "source": [
    "# Create embeddings one by one on the GPU\n",
    "for item in tqdm(pages_and_chunks_over_min_token_len):\n",
    "    item[\"embedding\"] = embedding_model.encode(item[\"sentence_chunk\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Turn text chunks into a single list\n",
    "text_chunks = [item[\"sentence_chunk\"] for item in pages_and_chunks_over_min_token_len]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: total: 1min 33s\n",
      "Wall time: 23.7 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "tensor([[ 0.0343, -0.0542, -0.0034,  ...,  0.0369,  0.0411, -0.0364],\n",
       "        [-0.0068, -0.0410,  0.0085,  ...,  0.0316,  0.0607, -0.0237],\n",
       "        [-0.0050, -0.1260, -0.0062,  ...,  0.0068,  0.0593, -0.0206],\n",
       "        ...,\n",
       "        [ 0.0296,  0.0022,  0.0380,  ...,  0.0203,  0.0186, -0.0001],\n",
       "        [ 0.0303, -0.0798,  0.0306,  ..., -0.0595,  0.0317, -0.0389],\n",
       "        [ 0.0075, -0.0537,  0.0127,  ..., -0.0398,  0.0403, -0.0321]])"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Embed all texts in batches\n",
    "text_chunk_embeddings = embedding_model.encode(text_chunks,\n",
    "                                               batch_size=32, # you can use different batch sizes here for speed/performance, I found 32 works well for this use case\n",
    "                                               convert_to_tensor=True) # optional to return embeddings as tensor instead of array\n",
    "\n",
    "text_chunk_embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save embeddings to file\n",
    "text_chunks_and_embeddings_df = pd.DataFrame(pages_and_chunks_over_min_token_len)\n",
    "embeddings_df_save_path = \"text_chunks_and_embeddings_df.csv\"\n",
    "text_chunks_and_embeddings_df.to_csv(embeddings_df_save_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_number</th>\n",
       "      <th>sentence_chunk</th>\n",
       "      <th>chunk_char_count</th>\n",
       "      <th>chunk_word_count</th>\n",
       "      <th>chunk_token_count</th>\n",
       "      <th>embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>INTRODUCTION CHAPTER 1 Introduction to Biology...</td>\n",
       "      <td>1487</td>\n",
       "      <td>246</td>\n",
       "      <td>371.75</td>\n",
       "      <td>[ 3.43400836e-02 -5.42259663e-02 -3.44239292e-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>they do not meet the criteria that biologists ...</td>\n",
       "      <td>1072</td>\n",
       "      <td>152</td>\n",
       "      <td>268.00</td>\n",
       "      <td>[-6.76683383e-03 -4.10040952e-02  8.46304279e-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>Inside each cell, atoms make up molecules. The...</td>\n",
       "      <td>1156</td>\n",
       "      <td>180</td>\n",
       "      <td>289.00</td>\n",
       "      <td>[-5.01606520e-03 -1.25997588e-01 -6.17972575e-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20</td>\n",
       "      <td>FIGURE 1.3 The leaves of this sensitive plant ...</td>\n",
       "      <td>1346</td>\n",
       "      <td>211</td>\n",
       "      <td>336.50</td>\n",
       "      <td>[-6.06122576e-02 -8.55128393e-02 -1.79607477e-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20</td>\n",
       "      <td>Adaptations enhance the reproductive potential...</td>\n",
       "      <td>640</td>\n",
       "      <td>94</td>\n",
       "      <td>160.00</td>\n",
       "      <td>[-7.58055039e-03 -4.15218994e-02 -3.92948538e-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   page_number                                     sentence_chunk  \\\n",
       "0           18  INTRODUCTION CHAPTER 1 Introduction to Biology...   \n",
       "1           19  they do not meet the criteria that biologists ...   \n",
       "2           19  Inside each cell, atoms make up molecules. The...   \n",
       "3           20  FIGURE 1.3 The leaves of this sensitive plant ...   \n",
       "4           20  Adaptations enhance the reproductive potential...   \n",
       "\n",
       "   chunk_char_count  chunk_word_count  chunk_token_count  \\\n",
       "0              1487               246             371.75   \n",
       "1              1072               152             268.00   \n",
       "2              1156               180             289.00   \n",
       "3              1346               211             336.50   \n",
       "4               640                94             160.00   \n",
       "\n",
       "                                           embedding  \n",
       "0  [ 3.43400836e-02 -5.42259663e-02 -3.44239292e-...  \n",
       "1  [-6.76683383e-03 -4.10040952e-02  8.46304279e-...  \n",
       "2  [-5.01606520e-03 -1.25997588e-01 -6.17972575e-...  \n",
       "3  [-6.06122576e-02 -8.55128393e-02 -1.79607477e-...  \n",
       "4  [-7.58055039e-03 -4.15218994e-02 -3.92948538e-...  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import saved file and view\n",
    "text_chunks_and_embedding_df_load = pd.read_csv(embeddings_df_save_path)\n",
    "text_chunks_and_embedding_df_load.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
